{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# eDNA conversion of practice data\n",
    "\n",
    "**Resources:**\n",
    "- https://docs.gbif-uat.org/publishing-sequence-derived-data/1.0/en/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## This is a test change ##\n",
    "### This is another test change ###"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Imports\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import random\n",
    "\n",
    "from datetime import datetime # for handling dates\n",
    "import pytz # for handling time zones"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(280440, 11)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ASV</th>\n",
       "      <th>FilterID</th>\n",
       "      <th>Sequence_ID</th>\n",
       "      <th>Reads</th>\n",
       "      <th>Kingdom</th>\n",
       "      <th>Phylum</th>\n",
       "      <th>Class</th>\n",
       "      <th>Order</th>\n",
       "      <th>Family</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>05114c01_12_edna_1</td>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>14825</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>unassigned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>05114c01_12_edna_2</td>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>16094</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>unassigned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>05114c01_12_edna_3</td>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>22459</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>unassigned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>11216c01_12_edna_1</td>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>19312</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>unassigned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>11216c01_12_edna_2</td>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>16491</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>unassigned</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 ASV            FilterID  \\\n",
       "0  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  05114c01_12_edna_1   \n",
       "1  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  05114c01_12_edna_2   \n",
       "2  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  05114c01_12_edna_3   \n",
       "3  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  11216c01_12_edna_1   \n",
       "4  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  11216c01_12_edna_2   \n",
       "\n",
       "            Sequence_ID  Reads    Kingdom      Phylum        Class      Order  \\\n",
       "0  05114c01_12_edna_1_S  14825  Eukaryota  Arthropoda  Hexanauplia  Calanoida   \n",
       "1  05114c01_12_edna_2_S  16094  Eukaryota  Arthropoda  Hexanauplia  Calanoida   \n",
       "2  05114c01_12_edna_3_S  22459  Eukaryota  Arthropoda  Hexanauplia  Calanoida   \n",
       "3  11216c01_12_edna_1_S  19312  Eukaryota  Arthropoda  Hexanauplia  Calanoida   \n",
       "4  11216c01_12_edna_2_S  16491  Eukaryota  Arthropoda  Hexanauplia  Calanoida   \n",
       "\n",
       "          Family        Genus     Species  \n",
       "0  Paracalanidae  Paracalanus  unassigned  \n",
       "1  Paracalanidae  Paracalanus  unassigned  \n",
       "2  Paracalanidae  Paracalanus  unassigned  \n",
       "3  Paracalanidae  Paracalanus  unassigned  \n",
       "4  Paracalanidae  Paracalanus  unassigned  "
      ]
     },
     "execution_count": 188,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Plate data\n",
    "\n",
    "plate = pd.read_csv('Plate_S_ASV_OBIS_data.csv')\n",
    "print(plate.shape)\n",
    "plate.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Note** that there are 60 unique FilterIDs and 60 unique Sequence_IDs in this table. Is that usual, to have # FilterID = # Sequence_ID? **IDs are a combination of what information?** Sequence_ID appears to be just the FilterID with underscore and S added."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(60, 2)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>FilterID</th>\n",
       "      <th>Sequence_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1</td>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2</td>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3</td>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1</td>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2</td>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             FilterID           Sequence_ID\n",
       "0  05114c01_12_edna_1  05114c01_12_edna_1_S\n",
       "1  05114c01_12_edna_2  05114c01_12_edna_2_S\n",
       "2  05114c01_12_edna_3  05114c01_12_edna_3_S\n",
       "3  11216c01_12_edna_1  11216c01_12_edna_1_S\n",
       "4  11216c01_12_edna_2  11216c01_12_edna_2_S"
      ]
     },
     "execution_count": 189,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Sequence_ID is just FilterID + '_S'\n",
    "\n",
    "temp = plate[['FilterID', 'Sequence_ID']].copy()\n",
    "temp.drop_duplicates(inplace=True)\n",
    "print(temp.shape)\n",
    "temp.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**FilterID seems to be composed of:**\n",
    "- SAMPLING_cruise from plate_meta\n",
    "- c + SAMPLING_station_number (or just SAMPLING_station, lower case and zero-padded?)\n",
    "- _ + SAMPLING_bottle\n",
    "- _ + edna\n",
    "- _ + 1, 2 or 3 (replicate?)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(4711, 79)\n",
      "Index(['ASV', 'Kingdom', 'Phylum', 'Class', 'Order', 'Family', 'Genus',\n",
      "       'Species', '14213c01_12_edna_1', '14213c01_12_edna_2',\n",
      "       '14213c01_12_edna_3', '22013c01_12_edna_1', '22013c01_12_edna_2',\n",
      "       '22013c01_12_edna_3', 'CN13Dc01_12_edna_1', 'CN13Dc01_12_edna_2',\n",
      "       'CN13Dc01_12_edna_3', '05114c01_12_edna_1', '05114c01_12_edna_2',\n",
      "       '05114c01_12_edna_3', '14714c01_12_edna_1', '14714c01_12_edna_2',\n",
      "       '14714c01_12_edna_3', '19114c01_12_edna_1', '19114c01_12_edna_2',\n",
      "       '19114c01_12_edna_3', '30214c01_12_edna_1', '30214c01_12_edna_2',\n",
      "       '30214c01_12_edna_3', '32414c01_12_edna_1', '32414c01_12_edna_2',\n",
      "       '32414c01_12_edna_3', '12015c01_12_edna_1', '12015c01_12_edna_2',\n",
      "       '12015c01_12_edna_3', '18815c01_12_edna_1', '18815c01_12_edna_2',\n",
      "       '18815c01_12_edna_3', 'EB_20161116', 'pcrblank_1', '28215c01_12_edna_1',\n",
      "       '28215c01_12_edna_2', '28215c01_12_edna_3', '34915c01_12_edna_1',\n",
      "       '34915c01_12_edna_2', '34915c01_12_edna_3', '11216c01_12_edna_1',\n",
      "       '11216c01_12_edna_2', '11216c01_12_edna_3', '24416c01_12_edna_1',\n",
      "       '24416c01_12_2', '24416c01_12_edna_3', 'CANON16c14_12_edna_1',\n",
      "       'CANON16c14_12_edna_2', 'CANON16c14_12_edna_3', '34916c01_12_edna_1',\n",
      "       '34916c01_12_edna_2', '34916c01_12_edna_3', 'JD10706C1_0m_1',\n",
      "       'JD10706C1_0m_2', 'JD10706C1_0m_3', 'JD15806C1_0m_1', 'JD15806C1_0m_2',\n",
      "       'JD15806C1_0m_3', 'JD27006C1_0m_1', 'JD27006C1_0m_2', 'JD27006C1_0m_3',\n",
      "       'JD33306C1_0m_1', 'JD33306C1_0m_2', 'JD33306C1_0m_3', 'pcrblank2',\n",
      "       'EB_20161121', 'EB_20161228', 'EB_20170117', 'pcrblank3',\n",
      "       'CB_CANON160925_1', 'CB_CANON160925_2', 'CB_CANON160925_3', 'ArtComm2'],\n",
      "      dtype='object')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ASV</th>\n",
       "      <th>Kingdom</th>\n",
       "      <th>Phylum</th>\n",
       "      <th>Class</th>\n",
       "      <th>Order</th>\n",
       "      <th>Family</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Species</th>\n",
       "      <th>14213c01_12_edna_1</th>\n",
       "      <th>14213c01_12_edna_2</th>\n",
       "      <th>...</th>\n",
       "      <th>JD33306C1_0m_3</th>\n",
       "      <th>pcrblank2</th>\n",
       "      <th>EB_20161121</th>\n",
       "      <th>EB_20161228</th>\n",
       "      <th>EB_20170117</th>\n",
       "      <th>pcrblank3</th>\n",
       "      <th>CB_CANON160925_1</th>\n",
       "      <th>CB_CANON160925_2</th>\n",
       "      <th>CB_CANON160925_3</th>\n",
       "      <th>ArtComm2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ASV_1</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>9</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>956</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1312</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>197</td>\n",
       "      <td>3</td>\n",
       "      <td>210</td>\n",
       "      <td>44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ASV_2</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Calanidae</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>4</td>\n",
       "      <td>7</td>\n",
       "      <td>...</td>\n",
       "      <td>358</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>535</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>118</td>\n",
       "      <td>162</td>\n",
       "      <td>139</td>\n",
       "      <td>5976</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ASV_3</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>Gymnodiniales</td>\n",
       "      <td>Gymnodiniaceae</td>\n",
       "      <td>Akashiwo</td>\n",
       "      <td>Akashiwo sanguinea</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>795</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>35</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ASV_4</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>Gymnodiniales</td>\n",
       "      <td>Gymnodiniaceae</td>\n",
       "      <td>Cochlodinium</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>82065</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>836</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ASV_5</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>148</td>\n",
       "      <td>147</td>\n",
       "      <td>...</td>\n",
       "      <td>14688</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>36</td>\n",
       "      <td>401</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>11</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 79 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     ASV    Kingdom      Phylum        Class          Order          Family  \\\n",
       "0  ASV_1  Eukaryota  Arthropoda  Hexanauplia      Calanoida   Paracalanidae   \n",
       "1  ASV_2  Eukaryota  Arthropoda  Hexanauplia      Calanoida       Calanidae   \n",
       "2  ASV_3  Eukaryota     unknown  Dinophyceae  Gymnodiniales  Gymnodiniaceae   \n",
       "3  ASV_4  Eukaryota     unknown  Dinophyceae  Gymnodiniales  Gymnodiniaceae   \n",
       "4  ASV_5  Eukaryota     unknown  Dinophyceae     unassigned      unassigned   \n",
       "\n",
       "          Genus             Species  14213c01_12_edna_1  14213c01_12_edna_2  \\\n",
       "0   Paracalanus          unassigned                   9                   4   \n",
       "1    unassigned          unassigned                   4                   7   \n",
       "2      Akashiwo  Akashiwo sanguinea                   2                   3   \n",
       "3  Cochlodinium          unassigned                   2                   3   \n",
       "4    unassigned          unassigned                 148                 147   \n",
       "\n",
       "   ...  JD33306C1_0m_3  pcrblank2  EB_20161121  EB_20161228  EB_20170117  \\\n",
       "0  ...             956          0            0         1312            1   \n",
       "1  ...             358          0            0          535            0   \n",
       "2  ...             795          0            0            0           35   \n",
       "3  ...           82065          0            0            0          836   \n",
       "4  ...           14688          0            0           36          401   \n",
       "\n",
       "   pcrblank3  CB_CANON160925_1  CB_CANON160925_2  CB_CANON160925_3  ArtComm2  \n",
       "0          0               197                 3               210        44  \n",
       "1          0               118               162               139      5976  \n",
       "2          0                 1                 0                 1         8  \n",
       "3          0                 1                 0                 0         7  \n",
       "4          0                 0                 4                11         7  \n",
       "\n",
       "[5 rows x 79 columns]"
      ]
     },
     "execution_count": 190,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## ASV taxa table\n",
    "\n",
    "taxa = pd.read_csv('Filtered_ASV_taxa_table_all.csv')\n",
    "print(taxa.shape)\n",
    "print(taxa.columns)\n",
    "taxa.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It looks like this shows the number of reads for each ASV detected for each FilterID (columns), plus the matched taxa information for each ASV, if available. Blank cells have been filled with the word 'unassigned'. \n",
    "\n",
    "There are some column names that do not correspond to FilterIDs. They are the taxonomy columns (of course), in addition to the following:\n",
    "- EB_20161116\n",
    "- pcrblank_1\n",
    "- pcrblank2\n",
    "- EB_20161121\n",
    "- EB_20161228\n",
    "- EB_20170117\n",
    "- pcrblank3\n",
    "- CB_CANON160925_1\n",
    "- CB_CANON160925_2\n",
    "- CB_CANON160925_3\n",
    "- ArtComm2\n",
    "\n",
    "Are some of these controls? \n",
    "\n",
    "```python\n",
    "for col in taxa.columns:\n",
    "    if col not in plate['FilterID'].unique():\n",
    "        print(col)\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(60, 52)\n",
      "Index(['SequenceID', 'sample_name', 'order', 'tag_sequence',\n",
      "       'primer_sequence_F', 'primer_sequence_R', 'library_tag_combo',\n",
      "       'library', 'date_PCR', 'sample_type', 'locus', 'tag_number', 'R1', 'R2',\n",
      "       'SAMPLING_cruise', 'SAMPLING_station_number', 'SAMPLING_bottle',\n",
      "       'depth', 'SAMPLING_station', 'SAMPLING_project', 'SAMPLING_platform',\n",
      "       'SAMPLING_platform_type', 'SAMPLING_dec_lat', 'SAMPLING_dec_lon',\n",
      "       'temp', 'salinity', 'chlorophyll', 'pressure_dbar', 'nitrate',\n",
      "       'diss_oxygen', 'SAMPLING_real_depth', 'SAMPLING_transmiss_%',\n",
      "       'SAMPLING_sig_t', 'SAMPLING_fluor', 'SAMPLING_date_time', 'description',\n",
      "       'SAMPLING_PI', 'SAMPLING_institute', 'env_biome', 'env_feature',\n",
      "       'env_material', 'samp_collection_device', 'project_name',\n",
      "       'samp_vol_we_dna_ext', 'samp_filter_size_ext', 'samp_filter_ext_type',\n",
      "       'samp_store_temp', 'seq_meth', 'sequencing_facility', 'geo_loc_name',\n",
      "       'investigation_type', 'FilterID'],\n",
      "      dtype='object')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SequenceID</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>order</th>\n",
       "      <th>tag_sequence</th>\n",
       "      <th>primer_sequence_F</th>\n",
       "      <th>primer_sequence_R</th>\n",
       "      <th>library_tag_combo</th>\n",
       "      <th>library</th>\n",
       "      <th>date_PCR</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>...</th>\n",
       "      <th>project_name</th>\n",
       "      <th>samp_vol_we_dna_ext</th>\n",
       "      <th>samp_filter_size_ext</th>\n",
       "      <th>samp_filter_ext_type</th>\n",
       "      <th>samp_store_temp</th>\n",
       "      <th>seq_meth</th>\n",
       "      <th>sequencing_facility</th>\n",
       "      <th>geo_loc_name</th>\n",
       "      <th>investigation_type</th>\n",
       "      <th>FilterID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>14213c01_12_edna_1_S</td>\n",
       "      <td>14213c01_12_edna_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ACGAGACTGATT</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>S1_ACGAGACTGATT</td>\n",
       "      <td>S1</td>\n",
       "      <td>20170124</td>\n",
       "      <td>environmental</td>\n",
       "      <td>...</td>\n",
       "      <td>MBON</td>\n",
       "      <td>100ml</td>\n",
       "      <td>0.22um</td>\n",
       "      <td>Poretics</td>\n",
       "      <td>-80C</td>\n",
       "      <td>NGS Illumina Miseq</td>\n",
       "      <td>Stanford</td>\n",
       "      <td>USA:California:Monterey Bay</td>\n",
       "      <td>eukaryote</td>\n",
       "      <td>14213c01_12_edna_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>14213c01_12_edna_2_S</td>\n",
       "      <td>14213c01_12_edna_2</td>\n",
       "      <td>2</td>\n",
       "      <td>GAATACCAAGTC</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>S2_GAATACCAAGTC</td>\n",
       "      <td>S2</td>\n",
       "      <td>20170124</td>\n",
       "      <td>environmental</td>\n",
       "      <td>...</td>\n",
       "      <td>MBON</td>\n",
       "      <td>100ml</td>\n",
       "      <td>0.22um</td>\n",
       "      <td>Poretics</td>\n",
       "      <td>-80C</td>\n",
       "      <td>NGS Illumina Miseq</td>\n",
       "      <td>Stanford</td>\n",
       "      <td>USA:California:Monterey Bay</td>\n",
       "      <td>eukaryote</td>\n",
       "      <td>14213c01_12_edna_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>14213c01_12_edna_3_S</td>\n",
       "      <td>14213c01_12_edna_3</td>\n",
       "      <td>3</td>\n",
       "      <td>CGAGGGAAAGTC</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>S3_CGAGGGAAAGTC</td>\n",
       "      <td>S3</td>\n",
       "      <td>20170124</td>\n",
       "      <td>environmental</td>\n",
       "      <td>...</td>\n",
       "      <td>MBON</td>\n",
       "      <td>100ml</td>\n",
       "      <td>0.22um</td>\n",
       "      <td>Poretics</td>\n",
       "      <td>-80C</td>\n",
       "      <td>NGS Illumina Miseq</td>\n",
       "      <td>Stanford</td>\n",
       "      <td>USA:California:Monterey Bay</td>\n",
       "      <td>eukaryote</td>\n",
       "      <td>14213c01_12_edna_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>22013c01_12_edna_1_S</td>\n",
       "      <td>22013c01_12_edna_1</td>\n",
       "      <td>4</td>\n",
       "      <td>GAACACTTTGGA</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>S4_GAACACTTTGGA</td>\n",
       "      <td>S4</td>\n",
       "      <td>20170124</td>\n",
       "      <td>environmental</td>\n",
       "      <td>...</td>\n",
       "      <td>MBON</td>\n",
       "      <td>100ml</td>\n",
       "      <td>0.22um</td>\n",
       "      <td>Poretics</td>\n",
       "      <td>-80C</td>\n",
       "      <td>NGS Illumina Miseq</td>\n",
       "      <td>Stanford</td>\n",
       "      <td>USA:California:Monterey Bay</td>\n",
       "      <td>eukaryote</td>\n",
       "      <td>22013c01_12_edna_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>22013c01_12_edna_2_S</td>\n",
       "      <td>22013c01_12_edna_2</td>\n",
       "      <td>5</td>\n",
       "      <td>ACTCACAGGAAT</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>S5_ACTCACAGGAAT</td>\n",
       "      <td>S5</td>\n",
       "      <td>20170124</td>\n",
       "      <td>environmental</td>\n",
       "      <td>...</td>\n",
       "      <td>MBON</td>\n",
       "      <td>100ml</td>\n",
       "      <td>0.22um</td>\n",
       "      <td>Poretics</td>\n",
       "      <td>-80C</td>\n",
       "      <td>NGS Illumina Miseq</td>\n",
       "      <td>Stanford</td>\n",
       "      <td>USA:California:Monterey Bay</td>\n",
       "      <td>eukaryote</td>\n",
       "      <td>22013c01_12_edna_2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 52 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             SequenceID         sample_name  order  tag_sequence  \\\n",
       "0  14213c01_12_edna_1_S  14213c01_12_edna_1      1  ACGAGACTGATT   \n",
       "1  14213c01_12_edna_2_S  14213c01_12_edna_2      2  GAATACCAAGTC   \n",
       "2  14213c01_12_edna_3_S  14213c01_12_edna_3      3  CGAGGGAAAGTC   \n",
       "3  22013c01_12_edna_1_S  22013c01_12_edna_1      4  GAACACTTTGGA   \n",
       "4  22013c01_12_edna_2_S  22013c01_12_edna_2      5  ACTCACAGGAAT   \n",
       "\n",
       "  primer_sequence_F         primer_sequence_R library_tag_combo library  \\\n",
       "0  GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC   S1_ACGAGACTGATT      S1   \n",
       "1  GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC   S2_GAATACCAAGTC      S2   \n",
       "2  GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC   S3_CGAGGGAAAGTC      S3   \n",
       "3  GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC   S4_GAACACTTTGGA      S4   \n",
       "4  GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC   S5_ACTCACAGGAAT      S5   \n",
       "\n",
       "   date_PCR    sample_type  ... project_name  samp_vol_we_dna_ext  \\\n",
       "0  20170124  environmental  ...         MBON                100ml   \n",
       "1  20170124  environmental  ...         MBON                100ml   \n",
       "2  20170124  environmental  ...         MBON                100ml   \n",
       "3  20170124  environmental  ...         MBON                100ml   \n",
       "4  20170124  environmental  ...         MBON                100ml   \n",
       "\n",
       "  samp_filter_size_ext samp_filter_ext_type samp_store_temp  \\\n",
       "0               0.22um             Poretics            -80C   \n",
       "1               0.22um             Poretics            -80C   \n",
       "2               0.22um             Poretics            -80C   \n",
       "3               0.22um             Poretics            -80C   \n",
       "4               0.22um             Poretics            -80C   \n",
       "\n",
       "             seq_meth  sequencing_facility                 geo_loc_name  \\\n",
       "0  NGS Illumina Miseq             Stanford  USA:California:Monterey Bay   \n",
       "1  NGS Illumina Miseq             Stanford  USA:California:Monterey Bay   \n",
       "2  NGS Illumina Miseq             Stanford  USA:California:Monterey Bay   \n",
       "3  NGS Illumina Miseq             Stanford  USA:California:Monterey Bay   \n",
       "4  NGS Illumina Miseq             Stanford  USA:California:Monterey Bay   \n",
       "\n",
       "  investigation_type            FilterID  \n",
       "0          eukaryote  14213c01_12_edna_1  \n",
       "1          eukaryote  14213c01_12_edna_2  \n",
       "2          eukaryote  14213c01_12_edna_3  \n",
       "3          eukaryote  22013c01_12_edna_1  \n",
       "4          eukaryote  22013c01_12_edna_2  \n",
       "\n",
       "[5 rows x 52 columns]"
      ]
     },
     "execution_count": 191,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Plate metadata\n",
    "\n",
    "plate_meta = pd.read_csv('Plate_S_meta_OBIS_data.csv')\n",
    "print(plate_meta.shape)\n",
    "print(plate_meta.columns)\n",
    "plate_meta.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This shows some metadata associated with each SequenceID and FilterID. This metadata **seems mostly redundant with that in MB_20200317_1326_18S_analysis_metadata.csv**, so for now I'll just work with it.\n",
    "\n",
    "## Conversion Plan\n",
    "\n",
    "According to GBIF's new guide to publishing sequence-derived biodiversity data, my occurrence file should include the following:\n",
    "\n",
    "<span style=\"color:blue\">eventID</span> - Highly recommended, **SequenceID in plate and plate_meta** <br>\n",
    "<span style=\"color:blue\">eventDate</span> - Required, **SAMPLING_date_time in plate_meta** <br>\n",
    "<span style=\"color:blue\">decimalLatitude</span> - Highly recommended, **SAMPLING_dec_lat in plate_meta** <br>\n",
    "<span style=\"color:blue\">decimalLongitude</span> - Highly recommended, **SAMPLING_dec_lon in plate_meta**  <br>\n",
    "<span style=\"color:blue\">env_broad_scale</span> - Recommended, equivalent to env_biome in MIxS, the major environmental system your sample or specimen came from, use [subclasses of ENVO´s biome class](http://www.ontobee.org/ontology/ENVO?iri=http://purl.obolibrary.org/obo/ENVO_00000428), **env_biome in plate_meta** <br>\n",
    "<span style=\"color:blue\">env_local_scale</span> - Recommended, equivalent to env_feature in MIxS, the entity or entities which are in your sample or specimen´s local vicinity and which you believe have significant causal influences on your sample or specimen, use terms that are present in ENVO and which are of smaller spatial grain than your entry for env_broad_scale, **env_feature in plate_meta** <br>\n",
    "<span style=\"color:blue\">env_medium</span> - Recommended, equivalent to env_material in MIxS, environmental material that immediately surrounded your sample or specimen prior to sampling, use [subclasses of ENVO´s environmental material class](http://www.ontobee.org/ontology/ENVO?iri=http://purl.obolibrary.org/obo/ENVO_00010483), **env_material in plate_meta**<br>\n",
    "<span style=\"color:blue\">sop</span> - Recommended, standard operating procedures used in assembly and/or annotation of metagenomes or a reference to a well documented protocol (e.g. using protocols.io). **Does something like this exist? If not, what information might be important to include here?** <br>\n",
    "<span style=\"color:blue\">lib_layout</span> - Recommended, equivalent to lib_const_meth in MIxS, whether to expect single, paired, or other configuration of reads. **Is this relevant here?** <br>\n",
    "<span style=\"color:blue\">target_gene</span> - Highly recommended, targeted gene or marker name for marker-based studies (e.g. 16S rRNA), **locus? Does there need to be more info included here?** <br>\n",
    "<span style=\"color:blue\">target_subfragment</span> - Highly recommended, name of subfragment of a gene or marker (e.g. V6). **Is this relevant here? Should the tag_sequence fit in somewhere?** <br>\n",
    "<span style=\"color:blue\">pcr_primer_name_forward</span> - Highly recommended, name of forward primer. **Do these primers have names?** <br>\n",
    "<span style=\"color:blue\">pcr_primer_forward</span> - Highly recommended, sequence of the forward primer, **primer_sequence_F in plate_meta** <br>\n",
    "<span style=\"color:blue\">pcr_primer_name_reverse</span> - Highly recommended, name of reverse primer <br>\n",
    "<span style=\"color:blue\">pcr_primer_revers</span> - Highly recommended, sequence of the reverse primer, **primer_sequence_R in plate_meta** <br>\n",
    "<span style=\"color:blue\">pcr_primer_reference</span> - Highly recommended, reference for primers (e.g. a DOI to a paper). **Is there a reference for primers?** <br>\n",
    "<span style=\"color:blue\">DNA_sequence</span> - Highly recommended, the actual DNA sequence of the ASV. TaxonID is highly recommended if DNA_sequence is not provided, **ASV in plate** <br>\n",
    "<span style=\"color:blue\">scientificName</span> - Required, Latin name of the closest known taxon or an OTU identifier from BOLD or UNITE, **a combination of Genus and Species from plate, or the lowest available taxon. This will have to be looked into more, especially for cases where no traditional taxonomic classification is available.** <br>\n",
    "<span style=\"color:blue\">kingdom</span> - Highly recommended, **Kingdom in plate** <br>\n",
    "<span style=\"color:blue\">phylum</span> - Recommended, **Phylum in plate** <br>\n",
    "<span style=\"color:blue\">class</span> - Recommended, **Class in plate** <br>\n",
    "<span style=\"color:blue\">order</span> - Recommended, **Order in plate** <br>\n",
    "<span style=\"color:blue\">family</span> - Recommended, **Family in plate** <br>\n",
    "<span style=\"color:blue\">genus</span> - Recommended, **Genus in plate** <br>\n",
    "<span style=\"color:blue\">basisOfRecord</span> - Required, MaterialSample <br>\n",
    "<span style=\"color:blue\">materialSampleID</span> - Highly recommended, an identifier for the MaterialSample, use the biosample ID if one was obtained from a nucleotide archive otherwise construct a globally unique identifier. **Is this Sequence_ID again? FilterID? Something else?** <br>\n",
    "<span style=\"color:blue\">identificationRemarks</span> - recommended, specification of taxonomic identification process ideally including data on applied algorithm and reference database as well as on level of confidence in the resulting identification. **Is this information available somewhere?** <br>\n",
    "<span style=\"color:blue\">identificationReferences</span> - recommended, link to protocol or code. **Is this information available somewhere?** <br>\n",
    "<span style=\"color:blue\">organismQuantity</span> - Highly recommended, number of reads, **Reads in plate** <br>\n",
    "<span style=\"color:blue\">organismQuantityType</span> - Highly recommended, DNA sequence reads <br>\n",
    "<span style=\"color:blue\">sampleSizeValue</span> - Highly recommended, total number of reads in the sample for calculating the relative abundance of sequence variants. **Is it accurate to just sum all the reads by Sequence_ID?** <br>\n",
    "<span style=\"color:blue\">sampleSizeUnit</span> - Highly recommended, DNA sequence reads <br>\n",
    "<span style=\"color:blue\">associatedSequences</span> - recommended, list of identifiers linking to archived (raw) sequence reads. **Are these sequences already archived?** <br>\n",
    "\n",
    "## Conversion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 235,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(280440, 1)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID\n",
       "0  05114c01_12_edna_1_S\n",
       "1  05114c01_12_edna_2_S\n",
       "2  05114c01_12_edna_3_S\n",
       "3  11216c01_12_edna_1_S\n",
       "4  11216c01_12_edna_2_S"
      ]
     },
     "execution_count": 235,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## eventID\n",
    "\n",
    "occ = pd.DataFrame({'eventID':plate['Sequence_ID']})\n",
    "print(occ.shape)\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 236,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>pcr_primer_forward</th>\n",
       "      <th>pcr_primer_reverse</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20 15:33</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20 15:33</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20 15:33</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21 14:39</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21 14:39</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID         eventDate  decimalLatitude  decimalLongitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20 15:33          36.7958          -121.848   \n",
       "1  05114c01_12_edna_2_S  2014-02-20 15:33          36.7958          -121.848   \n",
       "2  05114c01_12_edna_3_S  2014-02-20 15:33          36.7958          -121.848   \n",
       "3  11216c01_12_edna_1_S  2016-04-21 14:39          36.7962          -121.846   \n",
       "4  11216c01_12_edna_2_S  2016-04-21 14:39          36.7962          -121.846   \n",
       "\n",
       "                 env_broad_scale                env_local_scale  \\\n",
       "0  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   \n",
       "1  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   \n",
       "2  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   \n",
       "3  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   \n",
       "4  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  pcr_primer_forward        pcr_primer_reverse  \n",
       "0   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  \n",
       "1   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  \n",
       "2   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  \n",
       "3   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  \n",
       "4   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  "
      ]
     },
     "execution_count": 236,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Merge with plate_meta to obtain eventDate, decimal Lat and Lon\n",
    "\n",
    "occ = occ.merge(plate_meta[['SequenceID', 'SAMPLING_date_time', 'SAMPLING_dec_lat', 'SAMPLING_dec_lon',\n",
    "                            'env_biome', 'env_feature', 'env_material', 'locus', 'primer_sequence_F',\n",
    "                            'primer_sequence_R']], how='left', left_on='eventID', right_on='SequenceID')\n",
    "occ.drop(columns='SequenceID', inplace=True)\n",
    "occ.columns = ['eventID', 'eventDate', 'decimalLatitude', 'decimalLongitude', 'env_broad_scale', 'env_local_scale', 'env_medium',\n",
    "               'target_gene', 'pcr_primer_forward', 'pcr_primer_reverse']\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Need to add sop? lib_layout? target_subfragment? primer names? primer references?**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 237,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>pcr_primer_forward</th>\n",
       "      <th>pcr_primer_reverse</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude                env_broad_scale  \\\n",
       "0          -121.848  marine biome  (ENVO:00000447)   \n",
       "1          -121.848  marine biome  (ENVO:00000447)   \n",
       "2          -121.848  marine biome  (ENVO:00000447)   \n",
       "3          -121.846  marine biome  (ENVO:00000447)   \n",
       "4          -121.846  marine biome  (ENVO:00000447)   \n",
       "\n",
       "                 env_local_scale  \\\n",
       "0  coastal water (ENVO:00001250)   \n",
       "1  coastal water (ENVO:00001250)   \n",
       "2  coastal water (ENVO:00001250)   \n",
       "3  coastal water (ENVO:00001250)   \n",
       "4  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  pcr_primer_forward        pcr_primer_reverse  \n",
       "0   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  \n",
       "1   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  \n",
       "2   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  \n",
       "3   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  \n",
       "4   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  "
      ]
     },
     "execution_count": 237,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Format eventDate\n",
    "\n",
    "pst = pytz.timezone('America/Los_Angeles')\n",
    "eventDate = [pst.localize(datetime.strptime(dt, '%Y-%m-%d %H:%M')).isoformat() for dt in occ['eventDate']]\n",
    "occ['eventDate'] = eventDate\n",
    "\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 238,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>pcr_primer_forward</th>\n",
       "      <th>pcr_primer_reverse</th>\n",
       "      <th>DNA_sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude                env_broad_scale  \\\n",
       "0          -121.848  marine biome  (ENVO:00000447)   \n",
       "1          -121.848  marine biome  (ENVO:00000447)   \n",
       "2          -121.848  marine biome  (ENVO:00000447)   \n",
       "3          -121.846  marine biome  (ENVO:00000447)   \n",
       "4          -121.846  marine biome  (ENVO:00000447)   \n",
       "\n",
       "                 env_local_scale  \\\n",
       "0  coastal water (ENVO:00001250)   \n",
       "1  coastal water (ENVO:00001250)   \n",
       "2  coastal water (ENVO:00001250)   \n",
       "3  coastal water (ENVO:00001250)   \n",
       "4  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  pcr_primer_forward        pcr_primer_reverse  \\\n",
       "0   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC   \n",
       "1   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC   \n",
       "2   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC   \n",
       "3   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC   \n",
       "4   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC   \n",
       "\n",
       "                                        DNA_sequence  \n",
       "0  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  \n",
       "1  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  \n",
       "2  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  \n",
       "3  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  \n",
       "4  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  "
      ]
     },
     "execution_count": 238,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Add DNA_sequence\n",
    "\n",
    "occ['DNA_sequence'] = plate['ASV']\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Before adding taxonomic information, I'll have to clean up the Genus and Species columns in the plate df. **Weird values include:**\n",
    "- 'unassigned' has been used when there's no Genus data; also when there's no Species data\n",
    "- 'g_', **I assume** also signifies no Genus; 's_' seems to be the equivalent in Species\n",
    "- 'unknown' also has been used\n",
    "- 'no_hit' in both Genus and Species\n",
    "- 'Herdmania &lt;dinoflagellates>' - The Species entry for this seems to give both Genus and Species (Herdmania litoralis)\n",
    "    - Side note: **How can a Phylum be unknown, but then Class, Order, Family, etc. be known?**\n",
    "- 'Halofilum &lt;green algae>' - Similarly, Species is Halofilum ramosum\n",
    "- 'Candida &lt;clade Candida/Lodderomyces clade>' - This has Species unassigned, Family = Debaryomycetaceae\n",
    "\n",
    "Only in Species:\n",
    "- 'uncultured marine eukaryote'\n",
    "- 'eukaryote clone OLI11007'\n",
    "- 'Dinophyceae sp. UDMS0803'\n",
    "- 'uncultured marine picoeukaryote'\n",
    "- 'Chaetoceros sp. UNC1415'\n",
    "- 'bacterium'\n",
    "\n",
    "**Note** there are loads of terms like those given above in the Species column. **What do they mean and where do they come from?** Maybe some of those numbers correspond to BOLD or UNITE OTU ID's. **Also note** that the Species column often contains both Genus and Species - the proper scientific name - if available. I'm not sure if this always is the case, but maybe I can just use the Species column for scientificName rather than combining Genus and Species."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 239,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>pcr_primer_forward</th>\n",
       "      <th>pcr_primer_reverse</th>\n",
       "      <th>DNA_sequence</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude                env_broad_scale  \\\n",
       "0          -121.848  marine biome  (ENVO:00000447)   \n",
       "1          -121.848  marine biome  (ENVO:00000447)   \n",
       "2          -121.848  marine biome  (ENVO:00000447)   \n",
       "3          -121.846  marine biome  (ENVO:00000447)   \n",
       "4          -121.846  marine biome  (ENVO:00000447)   \n",
       "\n",
       "                 env_local_scale  \\\n",
       "0  coastal water (ENVO:00001250)   \n",
       "1  coastal water (ENVO:00001250)   \n",
       "2  coastal water (ENVO:00001250)   \n",
       "3  coastal water (ENVO:00001250)   \n",
       "4  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  pcr_primer_forward        pcr_primer_reverse  \\\n",
       "0   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC   \n",
       "1   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC   \n",
       "2   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC   \n",
       "3   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC   \n",
       "4   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC   \n",
       "\n",
       "                                        DNA_sequence scientificName  \\\n",
       "0  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...     unassigned   \n",
       "1  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...     unassigned   \n",
       "2  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...     unassigned   \n",
       "3  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...     unassigned   \n",
       "4  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...     unassigned   \n",
       "\n",
       "     kingdom      phylum        class      order         family        genus  \n",
       "0  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus  \n",
       "1  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus  \n",
       "2  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus  \n",
       "3  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus  \n",
       "4  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus  "
      ]
     },
     "execution_count": 239,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## scientificName, taxonomic info\n",
    "\n",
    "occ['scientificName'] = plate['Species']\n",
    "occ['kingdom'] = plate['Kingdom']\n",
    "occ['phylum'] = plate['Phylum']\n",
    "occ['class'] = plate['Class']\n",
    "occ['order'] = plate['Order']\n",
    "occ['family'] = plate['Family']\n",
    "occ['genus'] = plate['Genus']\n",
    "\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**So we don't have to do the WoRMS thing for eDNA data?**\n",
    "\n",
    "**Note** that:\n",
    "- 63600 (out of a total 280440 records) have scientificName = 'unassigned'\n",
    "- 163380 have scientificName = 's_'\n",
    "- 0 have scientificName = 'unknown'\n",
    "- 12540 have scientificName = 'no_hit'\n",
    "\n",
    "It seems like for the unassigned and s_ species, other taxonomic levels of classification (e.g. Family) are known (although I haven't checked this in all cases). When the species is 'no_hit', though, it seems like it's 'no_hit' across the board.\n",
    "\n",
    "**Let me replace these values with NaNs to make things easier to work with.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 240,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Clean scientificName and taxonomy columns\n",
    "\n",
    "cols = ['scientificName', 'kingdom', 'phylum', 'class', 'order', 'family', 'genus']\n",
    "occ[cols] = occ[cols].replace({'unassigned':np.nan,\n",
    "                              's_':np.nan,\n",
    "                              'g_':np.nan,\n",
    "                              'unknown':np.nan,\n",
    "                              'no_hit':np.nan})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Ok, so now we can see that:\n",
    "- 9300 records have full taxonomic info (all columns are not NaN)\n",
    "- A total of 239520 do not have a species designation (i.e. scientificName = NaN). This is as expected based on the above numbers.\n",
    "- Of these, 9840 are only missing species\n",
    "- 120 are only missing genus\n",
    "- 300 are only missing family\n",
    "- 1020 are only missing order\n",
    "- 780 are only missing class\n",
    "- 10860 are only missing phylum\n",
    "- 0 are only missing kingdom\n",
    "- This leaves 216600 records with more than one field missing\n",
    "\n",
    "**How do I handle these? Based on the standard information, it seems like I should fill NaNs in the species column with the lowest known taxonomic rank. If all the rows are NaN, the record cannot be submitted to OBIS. *But*, before deleting these, I need to look into the non-Linnaean name options.**\n",
    "\n",
    "**Note** that 33360 rows have ALL NaNs in ALL taxonomic columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 243,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Fill missing values in the species column with the lowest available taxonomic rank\n",
    "\n",
    "occ['scientificName'] = occ['scientificName'].combine_first(occ['genus'])\n",
    "occ['scientificName'] = occ['scientificName'].combine_first(occ['family'])\n",
    "occ['scientificName'] = occ['scientificName'].combine_first(occ['order'])\n",
    "occ['scientificName'] = occ['scientificName'].combine_first(occ['class'])\n",
    "occ['scientificName'] = occ['scientificName'].combine_first(occ['phylum'])\n",
    "occ['scientificName'] = occ['scientificName'].combine_first(occ['kingdom'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 245,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Paracalanus', 'Chlorophyta', 'Eukaryota', 'Florenciellales',\n",
       "       'Amoebophryaceae', 'Cercozoa', 'Chaetoceros diadema',\n",
       "       'uncultured marine eukaryote', 'Protoperidinium',\n",
       "       'Protoperidinium divergens', 'Dinophyceae',\n",
       "       'Cyclotella meneghiniana', 'Prymnesium', 'Ensiculifera imariensis',\n",
       "       'eukaryote clone OLI11007', 'Amoebophrya', nan, 'Gonyaulacaceae',\n",
       "       'Chrysophyceae', 'Scrippsiella', 'Prymnesiaceae',\n",
       "       'Dinophyceae sp. UDMS0803', 'Pelargonium', 'Hemistasiidae',\n",
       "       'Thysanoessa', 'Thoracosphaeraceae', 'Diplopsaliaceae',\n",
       "       'Hemiselmis', 'Euphausiidae', 'Gymnodiniaceae', 'Ciliophora',\n",
       "       'Karlodinium', 'Basidiomycota', 'Dictyocha fibula', 'Characeae',\n",
       "       'Picomonas', 'Polykrikos geminatum', 'Spirotrichea',\n",
       "       'uncultured marine picoeukaryote', 'Strombidiidae',\n",
       "       'Strombidium caudispina', 'Enteropneusta', 'Peridiniales',\n",
       "       'Kareniaceae', 'Halodinium verrucatum', 'Phaeocystaceae',\n",
       "       'Gyrodinium', 'Gonyaulax spinifera', 'Chaetoceros sp. UNC1415',\n",
       "       'Euduboscquella cachoni', 'Urechis', 'Symbiomonas scintillans',\n",
       "       'Mamiellaceae', 'Syndiniales', 'Haptophyta', 'Telonema',\n",
       "       'Thalassiosirales', 'Thaumatomastigidae', 'Alexandrium',\n",
       "       'Heterocapsaceae', 'Chaetothyriales', 'bacterium',\n",
       "       'Blastodiniales', 'Poaceae', 'Duboscquella',\n",
       "       'Gymnodinium litoralis', 'Prasinococcaceae', 'Rhizosoleniaceae',\n",
       "       'Tripos tenuis', 'Katablepharis remigera', 'Syndiniaceae',\n",
       "       'Bicosoecidae', 'Haliphthoraceae', 'Warnowia', 'Cryptosporidiidae',\n",
       "       'Strongylocentrotus purpuratus', 'Takayama cf. pulchellum',\n",
       "       'Proterythropsis', 'Prorocentrum triestinum', 'Protoperidiniaceae',\n",
       "       'Neogloboquadrina', 'Chaetoceros cinctus', 'Eurotiales',\n",
       "       'Rhabdonellidae', 'Ensiculifera', 'Prorodontidae',\n",
       "       'Pterosperma cristatum', 'Protoperidinium conicum', 'Tontoniidae',\n",
       "       'Pseudo-nitzschia', 'Spiniferites ramosus', 'Protaspidae',\n",
       "       'Chattonella', 'Pentapharsodinium', 'Haliphthoros', 'Bryophryidae',\n",
       "       'Cercomonadidae', 'eukaryote clone OLI11072', 'Rhogostomidae',\n",
       "       'Pseudo-nitzschia fraudulenta', 'Peridiniaceae', 'Prionospio',\n",
       "       'phototrophic eukaryote', 'Stephanopyxidaceae',\n",
       "       'Pirsonia guinardiae', 'Cochlodinium', 'Spirotontonia',\n",
       "       'Fragilariophyceae', 'Protoceratium reticulatum', 'Pteridomonas',\n",
       "       'Pedinellales', 'Pirsoniales', 'Actinocyclus curvatulus',\n",
       "       'Bathycoccus prasinos', 'Acantharea', 'Cyrtolophosididae',\n",
       "       'Protogastrostyla', 'Oomycetes', 'Corycaeidae',\n",
       "       'Solenicola setigera', 'Crustomastigaceae', 'Naviculales',\n",
       "       'Corethron inerme', 'Protoperidinium leonis',\n",
       "       'Bicosoeca vacillans', 'Stephanopyxis nipponica',\n",
       "       'Acantharian sp. 6201', 'Vannellidae', 'Echinoidea',\n",
       "       'Pelagophyceae sp. CCMP3390', 'Diplonemidae', 'Fragilariopsis',\n",
       "       'Bolidomonas', 'Thraustochytriaceae', 'Chaetoceros',\n",
       "       'Siphonophorae', 'Choanoflagellida sp. SL163', 'Thaumatomonadida',\n",
       "       'Ebria tripartita', 'Gyrodinium rubrum', 'Chrysochromulina',\n",
       "       'Stephanoecidae', 'Haptolina', 'Tetraselmis sp. KB-CR05',\n",
       "       'Telonemida', 'Cryptophyta', 'Blastodinium spinulosum',\n",
       "       'Pentapharsodinium tyrrhenicum', 'Marynidae', 'Thalassiosiraceae',\n",
       "       'Prorodontida', 'Cafeteria', 'Tintinnida', 'Pelagostrobilidium',\n",
       "       'Minorisa minuta', 'Spongodiscidae', 'Thecofilosea',\n",
       "       'Leptocylindrus', 'Bacillariaceae', 'Paraphysomonas',\n",
       "       'Tovelliaceae', 'Decapoda', 'Strobilidiidae', 'Cymatocylis',\n",
       "       'Eimeriidae', 'Teleaulax amphioxeia', 'Ditylum brightwellii',\n",
       "       'Pleurosigmataceae', 'Hartmannellidae', 'Geminigeraceae',\n",
       "       'Pseudocalanus', 'Pterocystidae', 'Theleodinium calcisporum',\n",
       "       'Goniodomataceae', 'Rapaza viridis', 'Appendicularia',\n",
       "       'Dinophysis', 'Dasybranchus', 'Diplonemea', 'Cryothecomonas',\n",
       "       'Amphidomataceae', 'Aureococcus anophagefferens',\n",
       "       'Dissodinium pseudolunula', 'Chrysochromulina strobilus',\n",
       "       'Strombidium sp. SNB99-2', 'Triparmaceae', 'Lagenidiales',\n",
       "       'Temora', 'Strombidium chlorophilum', 'Dinophysis acuminata',\n",
       "       'Thalassiosira', 'Trebouxia', 'Pfiesteria', 'Tintinnidae',\n",
       "       'Salpingoecidae', 'Phaeocystis', 'Mactridae',\n",
       "       'Phalacroma oxytoxoides', 'Hemistasia phaeocysticola', 'Mataza',\n",
       "       'Kryptoperidiniaceae', 'Fragilariaceae',\n",
       "       'Cyclotrichium cyclokaryon', 'Salpingella', 'Sinocalanus',\n",
       "       'Amylax', 'Paragymnodinium', 'Actinopteri', 'Eutintinnus pectinis',\n",
       "       'Strombidinopsidae', 'Pyramimonas gelidicola', 'Cercomonadida',\n",
       "       'Naviculaceae', 'Pythiaceae', 'Coccoid pelagophyte CCMP1395',\n",
       "       'Myzocytiopsidaceae', 'Euduboscquella', 'Attheyaceae',\n",
       "       'Ascomycota', 'Noelaerhabdaceae', 'Strombidium',\n",
       "       'Strombidinopsis acuminata', 'Paragymnodinium shiwhaense',\n",
       "       'Katablepharidaceae', 'Acanthamoeba mauritaniensis',\n",
       "       'Amphidiniopsidaceae', 'Pseudo-nitzschia australis', 'Mammalia',\n",
       "       'Chlamydomonas parkeae', 'Protoperidinium pellucidum',\n",
       "       'Duboscquellidae', 'Cymbomonas tetramitiformis',\n",
       "       'Chaetoceros tenuissimus', 'Minorisa', 'Poecilostomatoida',\n",
       "       'Phyllodocida', 'Amphibelone anomala', 'Paramoebidae',\n",
       "       'Oblongichytrium', 'Blastodinium mangini', 'Chromulinaceae',\n",
       "       'Protaspis', 'Protoperidinium monovelum', 'Cymatosiraceae',\n",
       "       'Calanoida', 'Ascomorpha ovalis', 'Bacillariophyta',\n",
       "       'Quadricilia rotundata', 'Pelagococcus subviridis',\n",
       "       'Pilea microphylla', 'Dothideales', 'Ichthyophonida',\n",
       "       'Thecate dinoflagellate UDTSW0701', 'Prorocentrum tsawwassenense',\n",
       "       'Gymnodinium', 'Plesiotrichopidae', 'Veneroida', 'Hemidiscaceae',\n",
       "       'Protoraphidaceae', 'Skeletonema', 'Capitellidae',\n",
       "       'Mycosphaerellaceae', 'Leegaardiella',\n",
       "       'Crustomastix sp. MBIC10709', 'Chaetoceros sp. UNC1414',\n",
       "       'Chaetoceros sp. NIES-3971', 'Haslea', 'Protodinium simplex',\n",
       "       'Strombidinopsis sp. NSMS0601', 'Solenicola',\n",
       "       'Protoperidinium thulesense', 'Pseudo-nitzschia heimii',\n",
       "       'Pachycordyle michaeli', 'Pyrenomonadaceae',\n",
       "       'Cryptophyta sp. CR-MAL03', 'Pyrocystaceae', 'Akashiwo sanguinea',\n",
       "       'Gonyaulax polygramma', 'Lessardia elongata',\n",
       "       'Oltmannsiellopsis sp. CCMP1240', 'Protoperidinium crassipes',\n",
       "       'Balanus nubilus', 'Polyporaceae', 'Myrtaceae',\n",
       "       'Protoperidinium thorianum', 'Pyramimonadales', 'Lepetodrilidae',\n",
       "       'Chaetoceros sp. TG07-C28', 'Fragilidium duplocampanaeforme',\n",
       "       'Pyramimonas parkeae', 'Polycystinea', 'Lineidae sp. 3DS',\n",
       "       'Laboea', 'Crustomastix', 'Acanthoeca sp. SIOpierAcanth1',\n",
       "       'Malassezia globosa', 'Hemiselmis rufescens', 'Phyllodocidae',\n",
       "       'Heteronemertea', 'Idas argenteus', 'Prorocentraceae',\n",
       "       'Minorisa sp. SRT710', 'Eucampia', 'Coscinodiscophyceae',\n",
       "       'Cinetochilidae', 'Sporidiobolales', 'Paulinellidae',\n",
       "       'Rhopalodiaceae', 'Marimonadida', 'Colpodellaceae',\n",
       "       'Ptychocylididae',\n",
       "       'Uncultured marine alveolate Group II DH145-EKD20', 'Sapindales',\n",
       "       'Calanidae', 'Gymnodiniales', 'Euglyphidae', 'Stokesiidae',\n",
       "       'Ciliate sp. NCMS0601', 'Sagitta sp. DP-2006', 'Ventrifissuridae',\n",
       "       'Eutintinnus cf. apertus', 'Dolichomastix', 'Actinoptychus',\n",
       "       'Nectriaceae', 'Acanthamoeba', 'Tiarina',\n",
       "       'Trachyrhizium urniformis', 'Ustilaginaceae',\n",
       "       'Blastodinium galatheanum', 'Incisomonas marina', 'Chytriodinium',\n",
       "       'eukaryote clone OLI11305', 'Rhizosolenia', 'Proboscia',\n",
       "       'Gesiella jameensis', 'Hemiselmidaceae', 'Peridinium',\n",
       "       'Pseudobodo', 'Goniomonas', 'Thalassiosira concaviuscula',\n",
       "       'Sabellariidae', 'Cryothecomonas sp. APCC MC5-1Cryo',\n",
       "       'Warnowia sp. BSL-2009a', 'Strombidium sp. JG-2011a', 'Ansanella',\n",
       "       'Pedinomonas', 'Fragilariopsis doliolus', 'Evadne nordmanni',\n",
       "       'Nephtys', 'Malassezia', 'Chytriodinium sp. Atlantic',\n",
       "       'Melanthiaceae', 'Chrysolepidomonadaceae', 'Calliacantha natans',\n",
       "       'Ventrifissura sp. SRT224', 'Oikopleura', 'Agaricomycetes',\n",
       "       'Adeleidae', 'Tintinnopsis', 'Pedinella sp. squamata', 'Hiatella',\n",
       "       'Chrysochromulina leadbeateri', 'Caecitellus parvulus',\n",
       "       'Calyptogena magnifica', 'Apiaceae', 'Anomiidae', 'Urostylida',\n",
       "       'Capnodiales', 'Spirostrombidium', 'Leucocryptos marina',\n",
       "       'Saccharomycetales', 'Plagiacanthidae', 'Streptophyta',\n",
       "       'Chaetoceros curvisetus', 'Myxophyllum magnum', 'Incisomonas',\n",
       "       'Sarcocystidae', 'Tetraselmis', 'Acanthoecidae', 'Mediophyceae',\n",
       "       'Pyramimonas', 'Xystonellidae', 'Eleutherascus lectardii',\n",
       "       'Ceratoperidiniaeceae', 'Mitrocomidae', 'Fragilidium',\n",
       "       'Cryptomycota', 'Synchaeta', 'Cryothecomonas aestivalis',\n",
       "       'Bentheuphausia amblyops', 'Mamiellophyceae',\n",
       "       'Pyramimonas australis', 'Eutintinnus', 'Pterosperma',\n",
       "       'Chlorophyta symbiont of Spongaster tetra', 'Oncaea', 'Fabaceae',\n",
       "       'Heterocapsa', 'Strombidinopsis', 'Mataza hastifera',\n",
       "       'Asterionellopsis glacialis', 'eukaryote clone OLI11055', 'Beroe',\n",
       "       'Spongillida', 'Paraphysomonadaceae',\n",
       "       'Oithona sp. Roadstead of Brest', 'Azadinium caudatum',\n",
       "       'Pseudopedinella elastica', 'Haptophyceae sp. EE-2014',\n",
       "       'Wallemia ichthyophaga', 'Chaetoceros sp. Na26B1', 'Bivalvia',\n",
       "       'Akashiwo', 'Cryptoperidiniopsis', 'Rimostrombidium',\n",
       "       'Ascoseirales', 'Herdmania litoralis', 'eukaryote clone OLI11261',\n",
       "       'Protoperidinium elegans', 'Penicillium', 'Sellaphora',\n",
       "       'Karlodinium veneficum', 'Caecitellus', 'Paulsenella vonstoschii',\n",
       "       'Pleosporaceae', 'Sclerotinia', 'Abollifer globosa',\n",
       "       'Brockmanniella brockmannii', 'Goniomonadaceae', 'Lineidae',\n",
       "       'Eucoccidiorida', 'Apusomonadidae', 'Cirratulidae', 'Russulales',\n",
       "       'Bicosta minor', 'Nannochloropsis', 'Varistrombidium',\n",
       "       'Desmochloris halophila', 'Chlorarachniophyceae',\n",
       "       'Selaginellaceae', 'Puccinia', 'Chroomonadaceae', 'Micromonas',\n",
       "       'Heterocapsa rotundata', 'Guinardia flaccida', 'Tremellomycetes',\n",
       "       'Hexasterias', 'Loxocephalidae', 'Rhizophydiales', 'Epicoccum',\n",
       "       'Vannella', 'Micromonas pusilla', 'Azadinium concinnum',\n",
       "       'Lynnella semiglobulosa', 'Prymnesiales', 'Chytriodiniaceae',\n",
       "       'Clytia gracilis', 'Protoperidinium punctulatum',\n",
       "       'Pseudocollinia beringensis', 'Halosphaera sp. M1670',\n",
       "       'prasinophyte sp. MBIC10622', 'Chaetocerotaceae', 'Hesionidae',\n",
       "       'Hypocreales', 'Azadinium',\n",
       "       'Thaumatomastix sp. CC002-Boundary Bay',\n",
       "       'Hemiselmis cryptochromatica', 'Glenodiniaceae',\n",
       "       'Dolichomastigaceae', 'Pectinaria granulata',\n",
       "       'Crypthecodinium sp. CAAE-CL2',\n",
       "       'Euduboscquella sp. ex Favella arcuata', 'Globigerinita uvula',\n",
       "       'Tintinnidium mucicola', 'Colpodea', 'Spumellaria', 'Cryptocaryon',\n",
       "       'Tremellales', 'Physciaceae', 'Hexanauplia',\n",
       "       'Bacteriastrum mediterraneum', 'Chlamydophryidae',\n",
       "       'Azadinium polongum', 'Eudubosquellidae',\n",
       "       'Proterythropsis sp. BSL-2009a', 'Pucciniaceae',\n",
       "       'Microstromatales', 'Tripos', 'Laminariales', 'Chaetoceros brevis',\n",
       "       'Diptera', 'Ulva flexuosa', 'Bacteriastrum',\n",
       "       'Amastigomonas sp. 2 Bamfield', 'Bacteria',\n",
       "       'Hexasterias problematica', 'Nassulidae', 'Ellisellidae',\n",
       "       'Spionidae', 'Paralagenidium', 'Chrysochromulina acantha',\n",
       "       'Stephanopyxis turris', 'Goussia', 'Ceratocorys horrida',\n",
       "       'Leptocylindrus convexus', 'Stoeckeria algicida', 'Sagittidae',\n",
       "       'Chrysochromulina andersonii', 'Thalassiosira aestivalis',\n",
       "       'Coscinodiscaceae', 'Nereididae', 'Cephalaspidea', 'Protodinium',\n",
       "       'Labyrinthulochytrium arktikum', 'Candida', 'Bicosoecida',\n",
       "       'Pteridomonas danica', 'Bigelowiella natans', 'Apicomplexa',\n",
       "       'Rhodomonas abbreviata', 'Micrura chlorapardalis', 'Myoida',\n",
       "       'Hyphozyma variabilis', 'Ebriacea', 'Biddulphia tridens',\n",
       "       'Paradinium poucheti', 'Tergestiella adriatica', 'Stephanoeca',\n",
       "       'Choanoflagellata', 'Opheliidae', 'Halosphaera',\n",
       "       'Chaetoceros debilis', 'Trebouxiophyceae', 'Colepidae',\n",
       "       'Amastigomonas', 'Mya arenaria', 'Polydora', 'Foettingeriidae',\n",
       "       'Trimyemidae', 'Protoperidinium pentagonum', 'Pheopolykrikos',\n",
       "       'Scytomonadidae', 'Amphora sp. 38', 'Teleaulax',\n",
       "       'Tilletiopsis sp. B-TB2005', 'Prorocentrum', 'Suessiaceae',\n",
       "       'Phaffomycetaceae', 'Chaetoceros radicans',\n",
       "       'Tintinnopsis sp. JG-2011a', 'Chromulinales', 'Lecanoromycetes',\n",
       "       'Pseudosolenia calcar-avis', 'Hemidiscus cuneiformis',\n",
       "       'Islandinium tricingulatum', 'Udeniomyces megalosporus',\n",
       "       'Takayama', 'Rhizopus stolonifer', 'Gonyaulacales',\n",
       "       'Chrysochromulinaceae', 'Laackmanniella prolongata', 'Bodonidae',\n",
       "       'Pleosporales', 'Chrysomeridaceae', 'Amphosoma', 'Cyclotella',\n",
       "       'Perkinsidae', 'Gonyaulax', 'Halopsis', 'Prosphaerosyllis xarifae',\n",
       "       'Agaricales', 'Warnowiaceae', 'Trizona', 'Funariaceae', 'Kondoa',\n",
       "       'Pfiesteriaceae', 'Lobata', 'Ophiurida',\n",
       "       'Pelagodinium sp. JD-2012', 'Falcomonas daucoides',\n",
       "       'Lophodiniaceae', 'Hemistasia', 'Tetramitus', 'Segregatospumella',\n",
       "       'Dinophysiaceae', 'Oblongichytrium sp. BAFCcult 3519',\n",
       "       'Pelagiidae', 'Klebsormidiaceae', 'Ectocarpus', 'Chaunacanthida',\n",
       "       'Cyclosalpa polae', 'Lysilla sp. THS-2012',\n",
       "       'Coscinodiscus sp. GGM-2004', 'Massisteria sp. diva', 'Synuraceae',\n",
       "       'Bacillariophyceae', 'Sphaerozoidae', 'Aspergillus', 'Apolemia',\n",
       "       'Crustomastix stigmatica', 'Nereis heterocirrata', 'Dendronotus',\n",
       "       'Ostreopsidaceae', 'Caliciaceae', 'Sporobolomyces', 'Apostomatida',\n",
       "       'Loligo', 'Storeatula major', 'Gigartacon muelleri',\n",
       "       'Glycinde armigera', 'Oodinium pouchetii', 'Cystobasidiomycetes',\n",
       "       'Hibberdiaceae', 'Pirsonia formosa', 'Pavlova',\n",
       "       'Stephanoeca diplocostata', 'Ellobiopsis',\n",
       "       'Rhogostoma schuessleri', 'Leuckartiara octona',\n",
       "       'Coccidia sp. Lutjanus/HAW36/TMW-2003',\n",
       "       'Neogloboquadrina dutertrei', 'Hemiselmis andersenii',\n",
       "       'Obelia dichotoma', 'Polychaeta', 'Phyllospadix iwatensis',\n",
       "       'Hexaconus', 'Terebellidae', 'Scenedesmaceae', 'Bentheuphausiidae',\n",
       "       'Vexillifera sp. K9', 'Pseudotontonia sp. JG-2011a',\n",
       "       'Spiniferites', 'Taphrina', 'Halofilum ramosum', 'Chytridiomycota',\n",
       "       'Pinus', 'Abollifer', 'Cenangiaceae', 'Navicula sp. AT-145.08',\n",
       "       'Phaeophyceae', 'Oxytrichidae',\n",
       "       'Candida <clade Candida/Lodderomyces clade>',\n",
       "       'Uncultured marine alveolate Group II DH147-EKD6',\n",
       "       'Amphilithiidae', 'Protophysarum sp. PV64', 'Plagoniidae',\n",
       "       'Peziza echinospora', 'Dinophyceae sp. CCMP1878',\n",
       "       'Pedinomonadaceae', 'Paraonidae', 'Blastodinium',\n",
       "       'Trieres sp. TN-2014', 'Gotoius excentricus', 'Heterophrys',\n",
       "       'Heterolobosea', 'Actinochloris sp. BCP-LG3VF20', 'Ceratiaceae',\n",
       "       'Gadiformes', 'Fuscoporia ferrea', 'Chattonellaceae', 'Cuninidae',\n",
       "       'Planomonas micra', 'Diplonemida sp. 21sb RG-2016', 'Catiniidae',\n",
       "       'Rhynchomonas', 'Carinoma sp. SA-2011', 'Mallomonas',\n",
       "       'Phytoseiidae', 'Sarcoptiformes', 'Aplanochytrium',\n",
       "       'Rhizophlyctidaceae', 'Phacellophora camtschatica', 'Massisteria',\n",
       "       'Hemiptera', 'Tulasnella', 'Virgularia', 'Balamuthiidae',\n",
       "       'Oligohymenophorea', 'Aggregatidae', 'Entomobryidae',\n",
       "       'Corticiaceae', 'Chironomidae', 'Hartmannella', 'Vacuolariaceae',\n",
       "       'Poseidonemertes maslakovae', 'Phaeocystis cordata', 'Parafavella',\n",
       "       'Euterpina acutifrons', 'Thalassiosira oceanica', 'Biddulphia',\n",
       "       'Rhodomelaceae', 'Polycystididae', 'Chloropicon', 'Flabellulidae',\n",
       "       'Chlorellaceae', 'Campanulariidae', 'Kofoidinium',\n",
       "       'Protoperidinium excentricum', 'Eutonina indicans',\n",
       "       'Labyrinthulomycetes', 'Ceratiomyxella tahitiensis',\n",
       "       'Actinommidae', 'Cryptomonadaceae', 'Subulatomonas',\n",
       "       'Himatismenida', 'Gregarinidae', 'Catenophlyctis',\n",
       "       'Vermistella antarctica', 'Aequorea', 'Nephroselmidophyceae',\n",
       "       'Flamella arnhemensis', 'Oxyrrhis marina', 'Vibrionaceae',\n",
       "       'Aplanochytrium kerguelense', 'Chaetoceros sp. UNC1416',\n",
       "       'Leptothecata', 'Polykrikos sp. EP-2016a', 'Navicula',\n",
       "       'Lithomelissa setosa', 'Prymnesium palpebrale', 'Bougainvilliidae',\n",
       "       'Maculaura alaskensis', 'Polykrikos kofoidii',\n",
       "       'Chaetoceros didymus', 'Parauronematidae', 'Plagiocampidae',\n",
       "       'Amoebophrya sp. ex Ceratium tripos', 'Creseis sp. GG-2005',\n",
       "       'Dictyocha speculum', 'Lepidodinium viride', 'Idanthyrsus',\n",
       "       'Eutintinnus tenuis', 'Rhynchomolgidae',\n",
       "       'Protoperidinium denticulatum', 'Mesodiniidae',\n",
       "       'Amoebophrya sp. ex Prorocentrum micans', 'Pirsonia verrucosa',\n",
       "       'Lauderia borealis', 'Rhizosolenia setigera',\n",
       "       'Eutreptiella eupharyngea', 'Plagiolemma', 'Dinophysis fortii',\n",
       "       'Coscinodiscus radiatus', 'Membraniporidae',\n",
       "       'Chrysocampanula spinifera', 'Proteomonas', 'Fibrocapsa japonica',\n",
       "       'Corycaeus speciosus', 'Chrysochromulina scutellum',\n",
       "       'Goniodoma polyedricum', 'Picobiliphyte sp. MS584-11',\n",
       "       'Polykrikos', 'Symbiodiniaceae', 'Norrisiella sphaerica',\n",
       "       'Pseudochattonella', 'Mytilus', 'Adenoides eludens',\n",
       "       'Rimostrombidium veniliae', 'Chrysaora fuscescens',\n",
       "       'Nematodinium sp. BSL-2009a', 'Protoperidinium depressum',\n",
       "       'Trichocerca elongata', 'Cryothecomonas longipes',\n",
       "       'Achradina pulchra', 'Oithonidae', 'Chaetoceros constrictus',\n",
       "       'Triparma', 'Minidiscus', 'Pterocystis',\n",
       "       'Chrysochromulina sp. Hp34EHU', 'Minutocellus polymorphus',\n",
       "       'Biecheleriaceae', 'Pleurobrachia', 'Sagitta', 'Archaeperidinium',\n",
       "       'Diaphanoeca undulata', 'Skeletonemataceae', 'Doliolum nationalis',\n",
       "       'eukaryote clone OLI11005', 'Pleurosigma sp. LM-2002',\n",
       "       'Varistrombidium kielum', 'Dothideomycetes', 'Minorisa sp. SRT705',\n",
       "       'Stephanopyxis', 'Pseudokeronopsidae', 'Planktoniella sol',\n",
       "       'Scrippsiella lachrymosa', 'Chaetoceros costatus',\n",
       "       'Choanoflagellida sp. SL117#1', 'Telonema antarcticum',\n",
       "       'Pseudotontonia simplicidens', 'Metridia',\n",
       "       'Katablepharis japonica', 'Amylax triacantha', 'Echiuridae',\n",
       "       'Calciodinellum operosum', 'Geminigera cryophila',\n",
       "       'Gymnodinium catenatum', 'Dactylodinium', 'Skeletonema menzelii',\n",
       "       'Pentacheles laevis', 'Helicostomella subulata', 'Pelagophyceae',\n",
       "       'Protodrilidae', 'Attheya longicornis', 'Biddulphia sp. MACC 6',\n",
       "       'Clausidiidae', 'Cladosporium', 'Amphorides quadrilineata',\n",
       "       'Heterosigma akashiwo', 'Norrisiella', 'Ochromonadaceae',\n",
       "       'Ventrifissura sp. SRT122', 'Parasagitta megalophthalma',\n",
       "       'Nephroselmis pyriformis', 'Oithonidae sp. DZMB624',\n",
       "       'Ectinosomatidae'], dtype=object)"
      ]
     },
     "execution_count": 245,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ['scientificName'].unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**NEXT STEPS**\n",
    "- Are any of the weird names from BOLD or UNITE already?\n",
    "- If not, is there some way to search for these IDs using DNA sequences? \n",
    "- Anything  that can't ultimately be matched with a taxonomic rank or ID must ultimately be dropped."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**What does it mean when the number of reads is 0? Is that an absence record??**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "venv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
