{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# eDNA conversion of practice data - draft 2\n",
    "\n",
    "**Resources:**\n",
    "- https://docs.gbif-uat.org/publishing-dna-derived-data/1.0/en/#mapping-metabarcoding-edna-and-barcoding-data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Imports\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import random\n",
    "\n",
    "from datetime import datetime # for handling dates\n",
    "import pytz # for handling time zones"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Ensure my general functions for the MPA data integration project can be imported, and import them\n",
    "\n",
    "import sys\n",
    "sys.path.insert(0, \"C:\\\\Users\\\\dianalg\\\\PycharmProjects\\\\PythonScripts\\\\eDNA\")\n",
    "\n",
    "import WoRMS # functions for querying WoRMS REST API"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(280440, 11)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ASV</th>\n",
       "      <th>FilterID</th>\n",
       "      <th>Sequence_ID</th>\n",
       "      <th>Reads</th>\n",
       "      <th>Kingdom</th>\n",
       "      <th>Phylum</th>\n",
       "      <th>Class</th>\n",
       "      <th>Order</th>\n",
       "      <th>Family</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>05114c01_12_edna_1</td>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>14825</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>unassigned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>05114c01_12_edna_2</td>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>16094</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>unassigned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>05114c01_12_edna_3</td>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>22459</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>unassigned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>11216c01_12_edna_1</td>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>19312</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>unassigned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>11216c01_12_edna_2</td>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>16491</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>unassigned</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 ASV            FilterID  \\\n",
       "0  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  05114c01_12_edna_1   \n",
       "1  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  05114c01_12_edna_2   \n",
       "2  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  05114c01_12_edna_3   \n",
       "3  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  11216c01_12_edna_1   \n",
       "4  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  11216c01_12_edna_2   \n",
       "\n",
       "            Sequence_ID  Reads    Kingdom      Phylum        Class      Order  \\\n",
       "0  05114c01_12_edna_1_S  14825  Eukaryota  Arthropoda  Hexanauplia  Calanoida   \n",
       "1  05114c01_12_edna_2_S  16094  Eukaryota  Arthropoda  Hexanauplia  Calanoida   \n",
       "2  05114c01_12_edna_3_S  22459  Eukaryota  Arthropoda  Hexanauplia  Calanoida   \n",
       "3  11216c01_12_edna_1_S  19312  Eukaryota  Arthropoda  Hexanauplia  Calanoida   \n",
       "4  11216c01_12_edna_2_S  16491  Eukaryota  Arthropoda  Hexanauplia  Calanoida   \n",
       "\n",
       "          Family        Genus     Species  \n",
       "0  Paracalanidae  Paracalanus  unassigned  \n",
       "1  Paracalanidae  Paracalanus  unassigned  \n",
       "2  Paracalanidae  Paracalanus  unassigned  \n",
       "3  Paracalanidae  Paracalanus  unassigned  \n",
       "4  Paracalanidae  Paracalanus  unassigned  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Plate data\n",
    "\n",
    "plate = pd.read_csv('Plate_S_ASV_OBIS_data.csv')\n",
    "print(plate.shape)\n",
    "plate.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**FilterID** = SAMPLING_cruise (cruise where the sample was collected) + \n",
    "    SAMPLING_station_number (in this case c + a cast number because these data were collected via CTD cast) + _ +\n",
    "    SAMPLING_bottle (bottle number, if multiple samples were taken from a single bottle, there may be replicates indicated by a, b, c, etc.) + _ +\n",
    "    edna (filter indicator. edna means that the standard filter for the standard edna pipeline, i.e. a standard PBDF filter, was used. other options include hplc) + _ +\n",
    "    1, 2 or 3 (PCR replicate)\n",
    "\n",
    "To summarize, **FilterID** = cruise number + cast number + bottle number + filter indicator + replicate number.\n",
    "\n",
    "**Note** that not all IDs exactly follow this formula. Example:\n",
    "- JD10706C1_0m_1\n",
    "\n",
    "**Sequence_ID** adds the plate indicator to the end of FilterID. In this case, all samples are from plate S. Plates are labeled A-Z, AA-ZZ, etc."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(4711, 79)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ASV</th>\n",
       "      <th>Kingdom</th>\n",
       "      <th>Phylum</th>\n",
       "      <th>Class</th>\n",
       "      <th>Order</th>\n",
       "      <th>Family</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Species</th>\n",
       "      <th>14213c01_12_edna_1</th>\n",
       "      <th>14213c01_12_edna_2</th>\n",
       "      <th>...</th>\n",
       "      <th>JD33306C1_0m_3</th>\n",
       "      <th>pcrblank2</th>\n",
       "      <th>EB_20161121</th>\n",
       "      <th>EB_20161228</th>\n",
       "      <th>EB_20170117</th>\n",
       "      <th>pcrblank3</th>\n",
       "      <th>CB_CANON160925_1</th>\n",
       "      <th>CB_CANON160925_2</th>\n",
       "      <th>CB_CANON160925_3</th>\n",
       "      <th>ArtComm2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ASV_1</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>9</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>956</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1312</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>197</td>\n",
       "      <td>3</td>\n",
       "      <td>210</td>\n",
       "      <td>44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ASV_2</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Calanidae</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>4</td>\n",
       "      <td>7</td>\n",
       "      <td>...</td>\n",
       "      <td>358</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>535</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>118</td>\n",
       "      <td>162</td>\n",
       "      <td>139</td>\n",
       "      <td>5976</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ASV_3</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>Gymnodiniales</td>\n",
       "      <td>Gymnodiniaceae</td>\n",
       "      <td>Akashiwo</td>\n",
       "      <td>Akashiwo sanguinea</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>795</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>35</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ASV_4</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>Gymnodiniales</td>\n",
       "      <td>Gymnodiniaceae</td>\n",
       "      <td>Cochlodinium</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>82065</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>836</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ASV_5</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>148</td>\n",
       "      <td>147</td>\n",
       "      <td>...</td>\n",
       "      <td>14688</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>36</td>\n",
       "      <td>401</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>11</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 79 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     ASV    Kingdom      Phylum        Class          Order          Family  \\\n",
       "0  ASV_1  Eukaryota  Arthropoda  Hexanauplia      Calanoida   Paracalanidae   \n",
       "1  ASV_2  Eukaryota  Arthropoda  Hexanauplia      Calanoida       Calanidae   \n",
       "2  ASV_3  Eukaryota     unknown  Dinophyceae  Gymnodiniales  Gymnodiniaceae   \n",
       "3  ASV_4  Eukaryota     unknown  Dinophyceae  Gymnodiniales  Gymnodiniaceae   \n",
       "4  ASV_5  Eukaryota     unknown  Dinophyceae     unassigned      unassigned   \n",
       "\n",
       "          Genus             Species  14213c01_12_edna_1  14213c01_12_edna_2  \\\n",
       "0   Paracalanus          unassigned                   9                   4   \n",
       "1    unassigned          unassigned                   4                   7   \n",
       "2      Akashiwo  Akashiwo sanguinea                   2                   3   \n",
       "3  Cochlodinium          unassigned                   2                   3   \n",
       "4    unassigned          unassigned                 148                 147   \n",
       "\n",
       "   ...  JD33306C1_0m_3  pcrblank2  EB_20161121  EB_20161228  EB_20170117  \\\n",
       "0  ...             956          0            0         1312            1   \n",
       "1  ...             358          0            0          535            0   \n",
       "2  ...             795          0            0            0           35   \n",
       "3  ...           82065          0            0            0          836   \n",
       "4  ...           14688          0            0           36          401   \n",
       "\n",
       "   pcrblank3  CB_CANON160925_1  CB_CANON160925_2  CB_CANON160925_3  ArtComm2  \n",
       "0          0               197                 3               210        44  \n",
       "1          0               118               162               139      5976  \n",
       "2          0                 1                 0                 1         8  \n",
       "3          0                 1                 0                 0         7  \n",
       "4          0                 0                 4                11         7  \n",
       "\n",
       "[5 rows x 79 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Taxonomy table\n",
    "\n",
    "taxa = pd.read_csv('Filtered_ASV_taxa_table_all.csv')\n",
    "print(taxa.shape)\n",
    "taxa.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This shows the number of reads for each ASV detected for each FilterID (columns), plus the matched taxa information for each ASV, if available. \n",
    "\n",
    "Some columns are associated with controls:\n",
    "- **CB** = collection blank which was taken on the ship. Water that should be \"clean\" was passed through the filter, so there should be no reads in this sample, although it tends to be the \"dirtiest\" control. It checks for contamination during filtration, I think.\n",
    "- **EB** = extraction blank which was taken in the lab but contains no DNA (i.e. a negative control), so there should be no reads in this sample. It checks if contamination occurred during DNA extraction (lab work pre-PCR).\n",
    "- **pcrblank** = A sample that went through PCR but contained no input DNA (i.e. a negative control), so there should be no reads in this sample. It checks if contamination occurred during PCR.\n",
    "- **ArtComm** = artificial community that went through PCR and contained DNA from species that should not be in the study system (i.e. a positive control), so you know what results you expect.\n",
    "\n",
    "To see these columns, use:\n",
    "```python\n",
    "for col in taxa.columns:\n",
    "    if col not in plate['FilterID'].unique():\n",
    "        print(col)\n",
    "```\n",
    "\n",
    "Additionally, taxonomic columns include the following designations:\n",
    "- **unknown** = GenBank couldn't give a scientifically-agreed-upon name for a given taxonomic level. I.e., either the name doesn't exist, or there isn't enough scientific consensus to give a name.\n",
    "- **no_hit** = BLAST did not find any hits for the ASV.\n",
    "- **unassigned** = The ASV got BLAST hits, but the post-processing program Megan6 didn't assign the ASV to any taxonomic group.\n",
    "- **g_** or **s_** = After Megan there's an additional filtering step where Genus or Species designations are only left if they were assigned with a certain, high level of confidence or higher. If this threshold is not met, a g_ or s_ is given."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(60, 52)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SequenceID</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>order</th>\n",
       "      <th>tag_sequence</th>\n",
       "      <th>primer_sequence_F</th>\n",
       "      <th>primer_sequence_R</th>\n",
       "      <th>library_tag_combo</th>\n",
       "      <th>library</th>\n",
       "      <th>date_PCR</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>...</th>\n",
       "      <th>project_name</th>\n",
       "      <th>samp_vol_we_dna_ext</th>\n",
       "      <th>samp_filter_size_ext</th>\n",
       "      <th>samp_filter_ext_type</th>\n",
       "      <th>samp_store_temp</th>\n",
       "      <th>seq_meth</th>\n",
       "      <th>sequencing_facility</th>\n",
       "      <th>geo_loc_name</th>\n",
       "      <th>investigation_type</th>\n",
       "      <th>FilterID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>14213c01_12_edna_1_S</td>\n",
       "      <td>14213c01_12_edna_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ACGAGACTGATT</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>S1_ACGAGACTGATT</td>\n",
       "      <td>S1</td>\n",
       "      <td>20170124</td>\n",
       "      <td>environmental</td>\n",
       "      <td>...</td>\n",
       "      <td>MBON</td>\n",
       "      <td>100ml</td>\n",
       "      <td>0.22um</td>\n",
       "      <td>Poretics</td>\n",
       "      <td>-80C</td>\n",
       "      <td>NGS Illumina Miseq</td>\n",
       "      <td>Stanford</td>\n",
       "      <td>USA:California:Monterey Bay</td>\n",
       "      <td>eukaryote</td>\n",
       "      <td>14213c01_12_edna_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>14213c01_12_edna_2_S</td>\n",
       "      <td>14213c01_12_edna_2</td>\n",
       "      <td>2</td>\n",
       "      <td>GAATACCAAGTC</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>S2_GAATACCAAGTC</td>\n",
       "      <td>S2</td>\n",
       "      <td>20170124</td>\n",
       "      <td>environmental</td>\n",
       "      <td>...</td>\n",
       "      <td>MBON</td>\n",
       "      <td>100ml</td>\n",
       "      <td>0.22um</td>\n",
       "      <td>Poretics</td>\n",
       "      <td>-80C</td>\n",
       "      <td>NGS Illumina Miseq</td>\n",
       "      <td>Stanford</td>\n",
       "      <td>USA:California:Monterey Bay</td>\n",
       "      <td>eukaryote</td>\n",
       "      <td>14213c01_12_edna_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>14213c01_12_edna_3_S</td>\n",
       "      <td>14213c01_12_edna_3</td>\n",
       "      <td>3</td>\n",
       "      <td>CGAGGGAAAGTC</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>S3_CGAGGGAAAGTC</td>\n",
       "      <td>S3</td>\n",
       "      <td>20170124</td>\n",
       "      <td>environmental</td>\n",
       "      <td>...</td>\n",
       "      <td>MBON</td>\n",
       "      <td>100ml</td>\n",
       "      <td>0.22um</td>\n",
       "      <td>Poretics</td>\n",
       "      <td>-80C</td>\n",
       "      <td>NGS Illumina Miseq</td>\n",
       "      <td>Stanford</td>\n",
       "      <td>USA:California:Monterey Bay</td>\n",
       "      <td>eukaryote</td>\n",
       "      <td>14213c01_12_edna_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>22013c01_12_edna_1_S</td>\n",
       "      <td>22013c01_12_edna_1</td>\n",
       "      <td>4</td>\n",
       "      <td>GAACACTTTGGA</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>S4_GAACACTTTGGA</td>\n",
       "      <td>S4</td>\n",
       "      <td>20170124</td>\n",
       "      <td>environmental</td>\n",
       "      <td>...</td>\n",
       "      <td>MBON</td>\n",
       "      <td>100ml</td>\n",
       "      <td>0.22um</td>\n",
       "      <td>Poretics</td>\n",
       "      <td>-80C</td>\n",
       "      <td>NGS Illumina Miseq</td>\n",
       "      <td>Stanford</td>\n",
       "      <td>USA:California:Monterey Bay</td>\n",
       "      <td>eukaryote</td>\n",
       "      <td>22013c01_12_edna_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>22013c01_12_edna_2_S</td>\n",
       "      <td>22013c01_12_edna_2</td>\n",
       "      <td>5</td>\n",
       "      <td>ACTCACAGGAAT</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>S5_ACTCACAGGAAT</td>\n",
       "      <td>S5</td>\n",
       "      <td>20170124</td>\n",
       "      <td>environmental</td>\n",
       "      <td>...</td>\n",
       "      <td>MBON</td>\n",
       "      <td>100ml</td>\n",
       "      <td>0.22um</td>\n",
       "      <td>Poretics</td>\n",
       "      <td>-80C</td>\n",
       "      <td>NGS Illumina Miseq</td>\n",
       "      <td>Stanford</td>\n",
       "      <td>USA:California:Monterey Bay</td>\n",
       "      <td>eukaryote</td>\n",
       "      <td>22013c01_12_edna_2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 52 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             SequenceID         sample_name  order  tag_sequence  \\\n",
       "0  14213c01_12_edna_1_S  14213c01_12_edna_1      1  ACGAGACTGATT   \n",
       "1  14213c01_12_edna_2_S  14213c01_12_edna_2      2  GAATACCAAGTC   \n",
       "2  14213c01_12_edna_3_S  14213c01_12_edna_3      3  CGAGGGAAAGTC   \n",
       "3  22013c01_12_edna_1_S  22013c01_12_edna_1      4  GAACACTTTGGA   \n",
       "4  22013c01_12_edna_2_S  22013c01_12_edna_2      5  ACTCACAGGAAT   \n",
       "\n",
       "  primer_sequence_F         primer_sequence_R library_tag_combo library  \\\n",
       "0  GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC   S1_ACGAGACTGATT      S1   \n",
       "1  GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC   S2_GAATACCAAGTC      S2   \n",
       "2  GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC   S3_CGAGGGAAAGTC      S3   \n",
       "3  GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC   S4_GAACACTTTGGA      S4   \n",
       "4  GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC   S5_ACTCACAGGAAT      S5   \n",
       "\n",
       "   date_PCR    sample_type  ... project_name  samp_vol_we_dna_ext  \\\n",
       "0  20170124  environmental  ...         MBON                100ml   \n",
       "1  20170124  environmental  ...         MBON                100ml   \n",
       "2  20170124  environmental  ...         MBON                100ml   \n",
       "3  20170124  environmental  ...         MBON                100ml   \n",
       "4  20170124  environmental  ...         MBON                100ml   \n",
       "\n",
       "  samp_filter_size_ext samp_filter_ext_type samp_store_temp  \\\n",
       "0               0.22um             Poretics            -80C   \n",
       "1               0.22um             Poretics            -80C   \n",
       "2               0.22um             Poretics            -80C   \n",
       "3               0.22um             Poretics            -80C   \n",
       "4               0.22um             Poretics            -80C   \n",
       "\n",
       "             seq_meth  sequencing_facility                 geo_loc_name  \\\n",
       "0  NGS Illumina Miseq             Stanford  USA:California:Monterey Bay   \n",
       "1  NGS Illumina Miseq             Stanford  USA:California:Monterey Bay   \n",
       "2  NGS Illumina Miseq             Stanford  USA:California:Monterey Bay   \n",
       "3  NGS Illumina Miseq             Stanford  USA:California:Monterey Bay   \n",
       "4  NGS Illumina Miseq             Stanford  USA:California:Monterey Bay   \n",
       "\n",
       "  investigation_type            FilterID  \n",
       "0          eukaryote  14213c01_12_edna_1  \n",
       "1          eukaryote  14213c01_12_edna_2  \n",
       "2          eukaryote  14213c01_12_edna_3  \n",
       "3          eukaryote  22013c01_12_edna_1  \n",
       "4          eukaryote  22013c01_12_edna_2  \n",
       "\n",
       "[5 rows x 52 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Plate metadata\n",
    "\n",
    "plate_meta = pd.read_csv('Plate_S_meta_OBIS_data.csv')\n",
    "print(plate_meta.shape)\n",
    "plate_meta.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Conversion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(280440, 1)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID\n",
       "0  05114c01_12_edna_1_S\n",
       "1  05114c01_12_edna_2_S\n",
       "2  05114c01_12_edna_3_S\n",
       "3  11216c01_12_edna_1_S\n",
       "4  11216c01_12_edna_2_S"
      ]
     },
     "execution_count": 127,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## eventID\n",
    "\n",
    "occ = pd.DataFrame({'eventID':plate['Sequence_ID']})\n",
    "print(occ.shape)\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>pcr_primer_forward</th>\n",
       "      <th>pcr_primer_reverse</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20 15:33</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20 15:33</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20 15:33</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21 14:39</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21 14:39</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID         eventDate  decimalLatitude  decimalLongitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20 15:33          36.7958          -121.848   \n",
       "1  05114c01_12_edna_2_S  2014-02-20 15:33          36.7958          -121.848   \n",
       "2  05114c01_12_edna_3_S  2014-02-20 15:33          36.7958          -121.848   \n",
       "3  11216c01_12_edna_1_S  2016-04-21 14:39          36.7962          -121.846   \n",
       "4  11216c01_12_edna_2_S  2016-04-21 14:39          36.7962          -121.846   \n",
       "\n",
       "                 env_broad_scale                env_local_scale  \\\n",
       "0  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   \n",
       "1  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   \n",
       "2  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   \n",
       "3  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   \n",
       "4  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  pcr_primer_forward        pcr_primer_reverse  \n",
       "0   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  \n",
       "1   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  \n",
       "2   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  \n",
       "3   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  \n",
       "4   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  "
      ]
     },
     "execution_count": 128,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Merge with plate_meta to obtain eventDate, decimal Lat and Lon\n",
    "\n",
    "occ = occ.merge(plate_meta[['SequenceID', 'SAMPLING_date_time', 'SAMPLING_dec_lat', 'SAMPLING_dec_lon',\n",
    "                            'env_biome', 'env_feature', 'env_material', 'locus', 'primer_sequence_F',\n",
    "                            'primer_sequence_R']], how='left', left_on='eventID', right_on='SequenceID')\n",
    "occ.drop(columns='SequenceID', inplace=True)\n",
    "occ.columns = ['eventID', 'eventDate', 'decimalLatitude', 'decimalLongitude', 'env_broad_scale', 'env_local_scale', 'env_medium',\n",
    "               'target_gene', 'pcr_primer_forward', 'pcr_primer_reverse']\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>pcr_primer_name_forward</th>\n",
       "      <th>pcr_primer_forward</th>\n",
       "      <th>pcr_primer_name_reverse</th>\n",
       "      <th>pcr_primer_reverse</th>\n",
       "      <th>pcr_primer_reference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20 15:33</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>insert primer reference</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20 15:33</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>insert primer reference</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20 15:33</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>insert primer reference</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21 14:39</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>insert primer reference</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21 14:39</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>insert primer reference</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID         eventDate  decimalLatitude  decimalLongitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20 15:33          36.7958          -121.848   \n",
       "1  05114c01_12_edna_2_S  2014-02-20 15:33          36.7958          -121.848   \n",
       "2  05114c01_12_edna_3_S  2014-02-20 15:33          36.7958          -121.848   \n",
       "3  11216c01_12_edna_1_S  2016-04-21 14:39          36.7962          -121.846   \n",
       "4  11216c01_12_edna_2_S  2016-04-21 14:39          36.7962          -121.846   \n",
       "\n",
       "                 env_broad_scale                env_local_scale  \\\n",
       "0  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   \n",
       "1  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   \n",
       "2  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   \n",
       "3  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   \n",
       "4  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  pcr_primer_name_forward pcr_primer_forward pcr_primer_name_reverse  \\\n",
       "0      insert primer name   GTACACACCGCCCGTC      insert primer name   \n",
       "1      insert primer name   GTACACACCGCCCGTC      insert primer name   \n",
       "2      insert primer name   GTACACACCGCCCGTC      insert primer name   \n",
       "3      insert primer name   GTACACACCGCCCGTC      insert primer name   \n",
       "4      insert primer name   GTACACACCGCCCGTC      insert primer name   \n",
       "\n",
       "         pcr_primer_reverse     pcr_primer_reference  \n",
       "0  TGATCCTTCTGCAGGTTCACCTAC  insert primer reference  \n",
       "1  TGATCCTTCTGCAGGTTCACCTAC  insert primer reference  \n",
       "2  TGATCCTTCTGCAGGTTCACCTAC  insert primer reference  \n",
       "3  TGATCCTTCTGCAGGTTCACCTAC  insert primer reference  \n",
       "4  TGATCCTTCTGCAGGTTCACCTAC  insert primer reference  "
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Add primer names and references\n",
    "\n",
    "occ.insert(8, 'pcr_primer_name_forward', 'insert primer name')\n",
    "occ.insert(10, 'pcr_primer_name_reverse', 'insert primer name')\n",
    "occ.insert(12, 'pcr_primer_reference', 'insert primer reference')\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>pcr_primer_name_forward</th>\n",
       "      <th>pcr_primer_forward</th>\n",
       "      <th>pcr_primer_name_reverse</th>\n",
       "      <th>pcr_primer_reverse</th>\n",
       "      <th>pcr_primer_reference</th>\n",
       "      <th>sop</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20 15:33</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>insert primer reference</td>\n",
       "      <td>insert sop link/doi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20 15:33</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>insert primer reference</td>\n",
       "      <td>insert sop link/doi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20 15:33</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>insert primer reference</td>\n",
       "      <td>insert sop link/doi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21 14:39</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>insert primer reference</td>\n",
       "      <td>insert sop link/doi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21 14:39</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>insert primer reference</td>\n",
       "      <td>insert sop link/doi</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID         eventDate  decimalLatitude  decimalLongitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20 15:33          36.7958          -121.848   \n",
       "1  05114c01_12_edna_2_S  2014-02-20 15:33          36.7958          -121.848   \n",
       "2  05114c01_12_edna_3_S  2014-02-20 15:33          36.7958          -121.848   \n",
       "3  11216c01_12_edna_1_S  2016-04-21 14:39          36.7962          -121.846   \n",
       "4  11216c01_12_edna_2_S  2016-04-21 14:39          36.7962          -121.846   \n",
       "\n",
       "                 env_broad_scale                env_local_scale  \\\n",
       "0  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   \n",
       "1  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   \n",
       "2  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   \n",
       "3  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   \n",
       "4  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  pcr_primer_name_forward pcr_primer_forward pcr_primer_name_reverse  \\\n",
       "0      insert primer name   GTACACACCGCCCGTC      insert primer name   \n",
       "1      insert primer name   GTACACACCGCCCGTC      insert primer name   \n",
       "2      insert primer name   GTACACACCGCCCGTC      insert primer name   \n",
       "3      insert primer name   GTACACACCGCCCGTC      insert primer name   \n",
       "4      insert primer name   GTACACACCGCCCGTC      insert primer name   \n",
       "\n",
       "         pcr_primer_reverse     pcr_primer_reference                  sop  \n",
       "0  TGATCCTTCTGCAGGTTCACCTAC  insert primer reference  insert sop link/doi  \n",
       "1  TGATCCTTCTGCAGGTTCACCTAC  insert primer reference  insert sop link/doi  \n",
       "2  TGATCCTTCTGCAGGTTCACCTAC  insert primer reference  insert sop link/doi  \n",
       "3  TGATCCTTCTGCAGGTTCACCTAC  insert primer reference  insert sop link/doi  \n",
       "4  TGATCCTTCTGCAGGTTCACCTAC  insert primer reference  insert sop link/doi  "
      ]
     },
     "execution_count": 130,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Add sop\n",
    "\n",
    "occ.insert(13, 'sop', 'insert sop link/doi')\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Note** that pcr_primer_name_forward, pcr_primer_name_reverse, pcr_primer_reference, and sop will be in Katie's revised metadata file, so ultimately I should be able to incorporate all of them in the original merge step."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>pcr_primer_name_forward</th>\n",
       "      <th>pcr_primer_forward</th>\n",
       "      <th>pcr_primer_name_reverse</th>\n",
       "      <th>pcr_primer_reverse</th>\n",
       "      <th>pcr_primer_reference</th>\n",
       "      <th>sop</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>insert primer reference</td>\n",
       "      <td>insert sop link/doi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>insert primer reference</td>\n",
       "      <td>insert sop link/doi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>insert primer reference</td>\n",
       "      <td>insert sop link/doi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>insert primer reference</td>\n",
       "      <td>insert sop link/doi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>insert primer reference</td>\n",
       "      <td>insert sop link/doi</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude                env_broad_scale  \\\n",
       "0          -121.848  marine biome  (ENVO:00000447)   \n",
       "1          -121.848  marine biome  (ENVO:00000447)   \n",
       "2          -121.848  marine biome  (ENVO:00000447)   \n",
       "3          -121.846  marine biome  (ENVO:00000447)   \n",
       "4          -121.846  marine biome  (ENVO:00000447)   \n",
       "\n",
       "                 env_local_scale  \\\n",
       "0  coastal water (ENVO:00001250)   \n",
       "1  coastal water (ENVO:00001250)   \n",
       "2  coastal water (ENVO:00001250)   \n",
       "3  coastal water (ENVO:00001250)   \n",
       "4  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  pcr_primer_name_forward pcr_primer_forward pcr_primer_name_reverse  \\\n",
       "0      insert primer name   GTACACACCGCCCGTC      insert primer name   \n",
       "1      insert primer name   GTACACACCGCCCGTC      insert primer name   \n",
       "2      insert primer name   GTACACACCGCCCGTC      insert primer name   \n",
       "3      insert primer name   GTACACACCGCCCGTC      insert primer name   \n",
       "4      insert primer name   GTACACACCGCCCGTC      insert primer name   \n",
       "\n",
       "         pcr_primer_reverse     pcr_primer_reference                  sop  \n",
       "0  TGATCCTTCTGCAGGTTCACCTAC  insert primer reference  insert sop link/doi  \n",
       "1  TGATCCTTCTGCAGGTTCACCTAC  insert primer reference  insert sop link/doi  \n",
       "2  TGATCCTTCTGCAGGTTCACCTAC  insert primer reference  insert sop link/doi  \n",
       "3  TGATCCTTCTGCAGGTTCACCTAC  insert primer reference  insert sop link/doi  \n",
       "4  TGATCCTTCTGCAGGTTCACCTAC  insert primer reference  insert sop link/doi  "
      ]
     },
     "execution_count": 131,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Format eventDate\n",
    "\n",
    "pst = pytz.timezone('America/Los_Angeles')\n",
    "eventDate = [pst.localize(datetime.strptime(dt, '%Y-%m-%d %H:%M')).isoformat() for dt in occ['eventDate']]\n",
    "occ['eventDate'] = eventDate\n",
    "\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>pcr_primer_name_forward</th>\n",
       "      <th>pcr_primer_forward</th>\n",
       "      <th>pcr_primer_name_reverse</th>\n",
       "      <th>pcr_primer_reverse</th>\n",
       "      <th>pcr_primer_reference</th>\n",
       "      <th>sop</th>\n",
       "      <th>DNA_sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>insert primer reference</td>\n",
       "      <td>insert sop link/doi</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>insert primer reference</td>\n",
       "      <td>insert sop link/doi</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>insert primer reference</td>\n",
       "      <td>insert sop link/doi</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>insert primer reference</td>\n",
       "      <td>insert sop link/doi</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>insert primer reference</td>\n",
       "      <td>insert sop link/doi</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude                env_broad_scale  \\\n",
       "0          -121.848  marine biome  (ENVO:00000447)   \n",
       "1          -121.848  marine biome  (ENVO:00000447)   \n",
       "2          -121.848  marine biome  (ENVO:00000447)   \n",
       "3          -121.846  marine biome  (ENVO:00000447)   \n",
       "4          -121.846  marine biome  (ENVO:00000447)   \n",
       "\n",
       "                 env_local_scale  \\\n",
       "0  coastal water (ENVO:00001250)   \n",
       "1  coastal water (ENVO:00001250)   \n",
       "2  coastal water (ENVO:00001250)   \n",
       "3  coastal water (ENVO:00001250)   \n",
       "4  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  pcr_primer_name_forward pcr_primer_forward pcr_primer_name_reverse  \\\n",
       "0      insert primer name   GTACACACCGCCCGTC      insert primer name   \n",
       "1      insert primer name   GTACACACCGCCCGTC      insert primer name   \n",
       "2      insert primer name   GTACACACCGCCCGTC      insert primer name   \n",
       "3      insert primer name   GTACACACCGCCCGTC      insert primer name   \n",
       "4      insert primer name   GTACACACCGCCCGTC      insert primer name   \n",
       "\n",
       "         pcr_primer_reverse     pcr_primer_reference                  sop  \\\n",
       "0  TGATCCTTCTGCAGGTTCACCTAC  insert primer reference  insert sop link/doi   \n",
       "1  TGATCCTTCTGCAGGTTCACCTAC  insert primer reference  insert sop link/doi   \n",
       "2  TGATCCTTCTGCAGGTTCACCTAC  insert primer reference  insert sop link/doi   \n",
       "3  TGATCCTTCTGCAGGTTCACCTAC  insert primer reference  insert sop link/doi   \n",
       "4  TGATCCTTCTGCAGGTTCACCTAC  insert primer reference  insert sop link/doi   \n",
       "\n",
       "                                        DNA_sequence  \n",
       "0  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  \n",
       "1  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  \n",
       "2  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  \n",
       "3  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  \n",
       "4  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  "
      ]
     },
     "execution_count": 132,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Add DNA_sequence\n",
    "\n",
    "occ['DNA_sequence'] = plate['ASV']\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>pcr_primer_name_forward</th>\n",
       "      <th>pcr_primer_forward</th>\n",
       "      <th>...</th>\n",
       "      <th>pcr_primer_reference</th>\n",
       "      <th>sop</th>\n",
       "      <th>DNA_sequence</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>...</td>\n",
       "      <td>insert primer reference</td>\n",
       "      <td>insert sop link/doi</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>...</td>\n",
       "      <td>insert primer reference</td>\n",
       "      <td>insert sop link/doi</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>...</td>\n",
       "      <td>insert primer reference</td>\n",
       "      <td>insert sop link/doi</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>...</td>\n",
       "      <td>insert primer reference</td>\n",
       "      <td>insert sop link/doi</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>...</td>\n",
       "      <td>insert primer reference</td>\n",
       "      <td>insert sop link/doi</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude                env_broad_scale  \\\n",
       "0          -121.848  marine biome  (ENVO:00000447)   \n",
       "1          -121.848  marine biome  (ENVO:00000447)   \n",
       "2          -121.848  marine biome  (ENVO:00000447)   \n",
       "3          -121.846  marine biome  (ENVO:00000447)   \n",
       "4          -121.846  marine biome  (ENVO:00000447)   \n",
       "\n",
       "                 env_local_scale  \\\n",
       "0  coastal water (ENVO:00001250)   \n",
       "1  coastal water (ENVO:00001250)   \n",
       "2  coastal water (ENVO:00001250)   \n",
       "3  coastal water (ENVO:00001250)   \n",
       "4  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  pcr_primer_name_forward pcr_primer_forward  ...     pcr_primer_reference  \\\n",
       "0      insert primer name   GTACACACCGCCCGTC  ...  insert primer reference   \n",
       "1      insert primer name   GTACACACCGCCCGTC  ...  insert primer reference   \n",
       "2      insert primer name   GTACACACCGCCCGTC  ...  insert primer reference   \n",
       "3      insert primer name   GTACACACCGCCCGTC  ...  insert primer reference   \n",
       "4      insert primer name   GTACACACCGCCCGTC  ...  insert primer reference   \n",
       "\n",
       "                   sop                                       DNA_sequence  \\\n",
       "0  insert sop link/doi  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...   \n",
       "1  insert sop link/doi  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...   \n",
       "2  insert sop link/doi  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...   \n",
       "3  insert sop link/doi  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...   \n",
       "4  insert sop link/doi  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...   \n",
       "\n",
       "  scientificName    kingdom      phylum        class      order  \\\n",
       "0     unassigned  Eukaryota  Arthropoda  Hexanauplia  Calanoida   \n",
       "1     unassigned  Eukaryota  Arthropoda  Hexanauplia  Calanoida   \n",
       "2     unassigned  Eukaryota  Arthropoda  Hexanauplia  Calanoida   \n",
       "3     unassigned  Eukaryota  Arthropoda  Hexanauplia  Calanoida   \n",
       "4     unassigned  Eukaryota  Arthropoda  Hexanauplia  Calanoida   \n",
       "\n",
       "          family        genus  \n",
       "0  Paracalanidae  Paracalanus  \n",
       "1  Paracalanidae  Paracalanus  \n",
       "2  Paracalanidae  Paracalanus  \n",
       "3  Paracalanidae  Paracalanus  \n",
       "4  Paracalanidae  Paracalanus  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "execution_count": 133,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## scientificName, taxonomic info\n",
    "\n",
    "occ['scientificName'] = plate['Species']\n",
    "occ['kingdom'] = plate['Kingdom']\n",
    "occ['phylum'] = plate['Phylum']\n",
    "occ['class'] = plate['Class']\n",
    "occ['order'] = plate['Order']\n",
    "occ['family'] = plate['Family']\n",
    "occ['genus'] = plate['Genus']\n",
    "\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>pcr_primer_name_forward</th>\n",
       "      <th>pcr_primer_forward</th>\n",
       "      <th>...</th>\n",
       "      <th>pcr_primer_reference</th>\n",
       "      <th>sop</th>\n",
       "      <th>DNA_sequence</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>...</td>\n",
       "      <td>insert primer reference</td>\n",
       "      <td>insert sop link/doi</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>...</td>\n",
       "      <td>insert primer reference</td>\n",
       "      <td>insert sop link/doi</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>...</td>\n",
       "      <td>insert primer reference</td>\n",
       "      <td>insert sop link/doi</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>...</td>\n",
       "      <td>insert primer reference</td>\n",
       "      <td>insert sop link/doi</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>...</td>\n",
       "      <td>insert primer reference</td>\n",
       "      <td>insert sop link/doi</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude                env_broad_scale  \\\n",
       "0          -121.848  marine biome  (ENVO:00000447)   \n",
       "1          -121.848  marine biome  (ENVO:00000447)   \n",
       "2          -121.848  marine biome  (ENVO:00000447)   \n",
       "3          -121.846  marine biome  (ENVO:00000447)   \n",
       "4          -121.846  marine biome  (ENVO:00000447)   \n",
       "\n",
       "                 env_local_scale  \\\n",
       "0  coastal water (ENVO:00001250)   \n",
       "1  coastal water (ENVO:00001250)   \n",
       "2  coastal water (ENVO:00001250)   \n",
       "3  coastal water (ENVO:00001250)   \n",
       "4  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  pcr_primer_name_forward pcr_primer_forward  ...     pcr_primer_reference  \\\n",
       "0      insert primer name   GTACACACCGCCCGTC  ...  insert primer reference   \n",
       "1      insert primer name   GTACACACCGCCCGTC  ...  insert primer reference   \n",
       "2      insert primer name   GTACACACCGCCCGTC  ...  insert primer reference   \n",
       "3      insert primer name   GTACACACCGCCCGTC  ...  insert primer reference   \n",
       "4      insert primer name   GTACACACCGCCCGTC  ...  insert primer reference   \n",
       "\n",
       "                   sop                                       DNA_sequence  \\\n",
       "0  insert sop link/doi  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...   \n",
       "1  insert sop link/doi  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...   \n",
       "2  insert sop link/doi  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...   \n",
       "3  insert sop link/doi  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...   \n",
       "4  insert sop link/doi  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...   \n",
       "\n",
       "  scientificName    kingdom      phylum        class      order  \\\n",
       "0            NaN  Eukaryota  Arthropoda  Hexanauplia  Calanoida   \n",
       "1            NaN  Eukaryota  Arthropoda  Hexanauplia  Calanoida   \n",
       "2            NaN  Eukaryota  Arthropoda  Hexanauplia  Calanoida   \n",
       "3            NaN  Eukaryota  Arthropoda  Hexanauplia  Calanoida   \n",
       "4            NaN  Eukaryota  Arthropoda  Hexanauplia  Calanoida   \n",
       "\n",
       "          family        genus  \n",
       "0  Paracalanidae  Paracalanus  \n",
       "1  Paracalanidae  Paracalanus  \n",
       "2  Paracalanidae  Paracalanus  \n",
       "3  Paracalanidae  Paracalanus  \n",
       "4  Paracalanidae  Paracalanus  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "execution_count": 134,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Replace 'unknown', 'unassigned', etc. in scientificName and taxonomy columns with NaN\n",
    "\n",
    "cols = ['scientificName', 'kingdom', 'phylum', 'class', 'order', 'family', 'genus']\n",
    "occ[cols] = occ[cols].replace({'unassigned':np.nan,\n",
    "                              's_':np.nan,\n",
    "                              'g_':np.nan,\n",
    "                              'unknown':np.nan,\n",
    "                              'no_hit':np.nan})\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Note** that there are a number of entries in the scientificName column, like \"uncultured marine eukaryote\", \"eukaryote clone OLI11007\", and \"Acantharian sp. 6201\", that are not proper Latin species names. It seems reasonable to replace these with NaN as well. **I've used a simple rule to filter these out, and it would be worth checking it as the data set expands, just to be extra careful.**\n",
    "\n",
    "To see the entries:\n",
    "```python\n",
    "names = occ['scientificName'].unique()\n",
    "names = names[~pd.isnull(names)] # remove NaN\n",
    "for name in names:\n",
    "    words_in_name = name.split(' ')\n",
    "    if len(words_in_name) > 2:\n",
    "        print(name)\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Replace non-Latin species names with NaN\n",
    "\n",
    "# Get names\n",
    "names = occ['scientificName'].unique()\n",
    "names = names[~pd.isnull(names)] # remove NaN\n",
    "\n",
    "# Get non-Latin names\n",
    "non_latin_names = []\n",
    "for name in names:\n",
    "    words_in_name = name.split(' ')\n",
    "    if len(words_in_name) > 2:\n",
    "        non_latin_names.append(name)\n",
    "non_latin_names_dict = {i:np.nan for i in non_latin_names}\n",
    "\n",
    "# Add any names that didn't get caught in the simple filter\n",
    "non_latin_names_dict['phototrophic eukaryote'] = np.nan\n",
    "\n",
    "# Replace\n",
    "occ['scientificName'].replace(non_latin_names_dict, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Fill missing values in the species column with the lowest available taxonomic rank\n",
    "\n",
    "occ['scientificName'] = occ['scientificName'].combine_first(occ['genus'])\n",
    "occ['scientificName'] = occ['scientificName'].combine_first(occ['family'])\n",
    "occ['scientificName'] = occ['scientificName'].combine_first(occ['order'])\n",
    "occ['scientificName'] = occ['scientificName'].combine_first(occ['class'])\n",
    "occ['scientificName'] = occ['scientificName'].combine_first(occ['phylum'])\n",
    "occ['scientificName'] = occ['scientificName'].combine_first(occ['kingdom'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Note** that there are 33360 records (~12% of the data) where no taxonomic information was obtained at all (i.e. scientificName is still NaN). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "756"
      ]
     },
     "execution_count": 137,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Get unique species names\n",
    "\n",
    "names = occ['scientificName'].unique()\n",
    "names = names[~pd.isnull(names)] # remove NaN\n",
    "len(names)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Note:** These names do not have to be WoRMS-approved, as far as I can tell. As long as there's some taxonomic information, the record can be submitted to OBIS. But just to be aware, there are a number of names in scientificName that are not approved by WoRMS. To see them, use:\n",
    "```python\n",
    "name_id_dict, name_name_dict, name_taxid_dict, name_class_dict = WoRMS.run_get_worms_from_scientific_name(names, verbose_flag=True)\n",
    "```\n",
    "\n",
    "**Also**, many entries have no other information other than the kingdom is \"Eukaryota\". This troubles me a little bit, since that's a domain, not a kingdom. **Will this be a problem? Is it possible to get the actual kingdom information in the kingdom field?**\n",
    "\n",
    "**Finally** the 33360 records with no taxonomic information will need to be dropped eventually, but I'll leave it until the end in case we want to add any more columns from the original data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>pcr_primer_name_forward</th>\n",
       "      <th>pcr_primer_forward</th>\n",
       "      <th>...</th>\n",
       "      <th>sop</th>\n",
       "      <th>DNA_sequence</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>basisOfRecord</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>...</td>\n",
       "      <td>insert sop link/doi</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>MaterialSample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>...</td>\n",
       "      <td>insert sop link/doi</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>MaterialSample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>...</td>\n",
       "      <td>insert sop link/doi</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>MaterialSample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>...</td>\n",
       "      <td>insert sop link/doi</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>MaterialSample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>...</td>\n",
       "      <td>insert sop link/doi</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>MaterialSample</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude                env_broad_scale  \\\n",
       "0          -121.848  marine biome  (ENVO:00000447)   \n",
       "1          -121.848  marine biome  (ENVO:00000447)   \n",
       "2          -121.848  marine biome  (ENVO:00000447)   \n",
       "3          -121.846  marine biome  (ENVO:00000447)   \n",
       "4          -121.846  marine biome  (ENVO:00000447)   \n",
       "\n",
       "                 env_local_scale  \\\n",
       "0  coastal water (ENVO:00001250)   \n",
       "1  coastal water (ENVO:00001250)   \n",
       "2  coastal water (ENVO:00001250)   \n",
       "3  coastal water (ENVO:00001250)   \n",
       "4  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  pcr_primer_name_forward pcr_primer_forward  ...                  sop  \\\n",
       "0      insert primer name   GTACACACCGCCCGTC  ...  insert sop link/doi   \n",
       "1      insert primer name   GTACACACCGCCCGTC  ...  insert sop link/doi   \n",
       "2      insert primer name   GTACACACCGCCCGTC  ...  insert sop link/doi   \n",
       "3      insert primer name   GTACACACCGCCCGTC  ...  insert sop link/doi   \n",
       "4      insert primer name   GTACACACCGCCCGTC  ...  insert sop link/doi   \n",
       "\n",
       "                                        DNA_sequence scientificName  \\\n",
       "0  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...    Paracalanus   \n",
       "1  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...    Paracalanus   \n",
       "2  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...    Paracalanus   \n",
       "3  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...    Paracalanus   \n",
       "4  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...    Paracalanus   \n",
       "\n",
       "     kingdom      phylum        class      order         family        genus  \\\n",
       "0  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   \n",
       "1  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   \n",
       "2  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   \n",
       "3  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   \n",
       "4  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   \n",
       "\n",
       "    basisOfRecord  \n",
       "0  MaterialSample  \n",
       "1  MaterialSample  \n",
       "2  MaterialSample  \n",
       "3  MaterialSample  \n",
       "4  MaterialSample  \n",
       "\n",
       "[5 rows x 23 columns]"
      ]
     },
     "execution_count": 138,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## basisOfRecord\n",
    "\n",
    "occ['basisOfRecord'] = 'MaterialSample'\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>pcr_primer_name_forward</th>\n",
       "      <th>pcr_primer_forward</th>\n",
       "      <th>...</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>basisOfRecord</th>\n",
       "      <th>materialSampleID</th>\n",
       "      <th>identificationRemarks</th>\n",
       "      <th>identificationReferences</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>...</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>insert doi/SRA number (?) if sequence has been...</td>\n",
       "      <td>insert the GenBank version used, the fact that...</td>\n",
       "      <td>insert links/dois for published papers, and li...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>...</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>insert doi/SRA number (?) if sequence has been...</td>\n",
       "      <td>insert the GenBank version used, the fact that...</td>\n",
       "      <td>insert links/dois for published papers, and li...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>...</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>insert doi/SRA number (?) if sequence has been...</td>\n",
       "      <td>insert the GenBank version used, the fact that...</td>\n",
       "      <td>insert links/dois for published papers, and li...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>...</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>insert doi/SRA number (?) if sequence has been...</td>\n",
       "      <td>insert the GenBank version used, the fact that...</td>\n",
       "      <td>insert links/dois for published papers, and li...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>...</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>insert doi/SRA number (?) if sequence has been...</td>\n",
       "      <td>insert the GenBank version used, the fact that...</td>\n",
       "      <td>insert links/dois for published papers, and li...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude                env_broad_scale  \\\n",
       "0          -121.848  marine biome  (ENVO:00000447)   \n",
       "1          -121.848  marine biome  (ENVO:00000447)   \n",
       "2          -121.848  marine biome  (ENVO:00000447)   \n",
       "3          -121.846  marine biome  (ENVO:00000447)   \n",
       "4          -121.846  marine biome  (ENVO:00000447)   \n",
       "\n",
       "                 env_local_scale  \\\n",
       "0  coastal water (ENVO:00001250)   \n",
       "1  coastal water (ENVO:00001250)   \n",
       "2  coastal water (ENVO:00001250)   \n",
       "3  coastal water (ENVO:00001250)   \n",
       "4  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  pcr_primer_name_forward pcr_primer_forward  ...    kingdom      phylum  \\\n",
       "0      insert primer name   GTACACACCGCCCGTC  ...  Eukaryota  Arthropoda   \n",
       "1      insert primer name   GTACACACCGCCCGTC  ...  Eukaryota  Arthropoda   \n",
       "2      insert primer name   GTACACACCGCCCGTC  ...  Eukaryota  Arthropoda   \n",
       "3      insert primer name   GTACACACCGCCCGTC  ...  Eukaryota  Arthropoda   \n",
       "4      insert primer name   GTACACACCGCCCGTC  ...  Eukaryota  Arthropoda   \n",
       "\n",
       "         class      order         family        genus   basisOfRecord  \\\n",
       "0  Hexanauplia  Calanoida  Paracalanidae  Paracalanus  MaterialSample   \n",
       "1  Hexanauplia  Calanoida  Paracalanidae  Paracalanus  MaterialSample   \n",
       "2  Hexanauplia  Calanoida  Paracalanidae  Paracalanus  MaterialSample   \n",
       "3  Hexanauplia  Calanoida  Paracalanidae  Paracalanus  MaterialSample   \n",
       "4  Hexanauplia  Calanoida  Paracalanidae  Paracalanus  MaterialSample   \n",
       "\n",
       "                                    materialSampleID  \\\n",
       "0  insert doi/SRA number (?) if sequence has been...   \n",
       "1  insert doi/SRA number (?) if sequence has been...   \n",
       "2  insert doi/SRA number (?) if sequence has been...   \n",
       "3  insert doi/SRA number (?) if sequence has been...   \n",
       "4  insert doi/SRA number (?) if sequence has been...   \n",
       "\n",
       "                               identificationRemarks  \\\n",
       "0  insert the GenBank version used, the fact that...   \n",
       "1  insert the GenBank version used, the fact that...   \n",
       "2  insert the GenBank version used, the fact that...   \n",
       "3  insert the GenBank version used, the fact that...   \n",
       "4  insert the GenBank version used, the fact that...   \n",
       "\n",
       "                            identificationReferences  \n",
       "0  insert links/dois for published papers, and li...  \n",
       "1  insert links/dois for published papers, and li...  \n",
       "2  insert links/dois for published papers, and li...  \n",
       "3  insert links/dois for published papers, and li...  \n",
       "4  insert links/dois for published papers, and li...  \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 139,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Add materialSampleID, identificationRemarks, identificationReferences -- can probably do this through merging metadata when Katie adds this info\n",
    "\n",
    "occ['materialSampleID'] = 'insert doi/SRA number (?) if sequence has been archived'\n",
    "occ['identificationRemarks'] = 'insert the GenBank version used, the fact that BLAST and Megan6 were used, and the BLAST confidence level'\n",
    "occ['identificationReferences'] = 'insert links/dois for published papers, and link to github code'\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>pcr_primer_name_forward</th>\n",
       "      <th>pcr_primer_forward</th>\n",
       "      <th>...</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>basisOfRecord</th>\n",
       "      <th>materialSampleID</th>\n",
       "      <th>identificationRemarks</th>\n",
       "      <th>identificationReferences</th>\n",
       "      <th>organismQuantity</th>\n",
       "      <th>organismQuantityType</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>...</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>insert doi/SRA number (?) if sequence has been...</td>\n",
       "      <td>insert the GenBank version used, the fact that...</td>\n",
       "      <td>insert links/dois for published papers, and li...</td>\n",
       "      <td>14825</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>...</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>insert doi/SRA number (?) if sequence has been...</td>\n",
       "      <td>insert the GenBank version used, the fact that...</td>\n",
       "      <td>insert links/dois for published papers, and li...</td>\n",
       "      <td>16094</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>...</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>insert doi/SRA number (?) if sequence has been...</td>\n",
       "      <td>insert the GenBank version used, the fact that...</td>\n",
       "      <td>insert links/dois for published papers, and li...</td>\n",
       "      <td>22459</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>...</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>insert doi/SRA number (?) if sequence has been...</td>\n",
       "      <td>insert the GenBank version used, the fact that...</td>\n",
       "      <td>insert links/dois for published papers, and li...</td>\n",
       "      <td>19312</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>...</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>insert doi/SRA number (?) if sequence has been...</td>\n",
       "      <td>insert the GenBank version used, the fact that...</td>\n",
       "      <td>insert links/dois for published papers, and li...</td>\n",
       "      <td>16491</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude                env_broad_scale  \\\n",
       "0          -121.848  marine biome  (ENVO:00000447)   \n",
       "1          -121.848  marine biome  (ENVO:00000447)   \n",
       "2          -121.848  marine biome  (ENVO:00000447)   \n",
       "3          -121.846  marine biome  (ENVO:00000447)   \n",
       "4          -121.846  marine biome  (ENVO:00000447)   \n",
       "\n",
       "                 env_local_scale  \\\n",
       "0  coastal water (ENVO:00001250)   \n",
       "1  coastal water (ENVO:00001250)   \n",
       "2  coastal water (ENVO:00001250)   \n",
       "3  coastal water (ENVO:00001250)   \n",
       "4  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  pcr_primer_name_forward pcr_primer_forward  ...        class      order  \\\n",
       "0      insert primer name   GTACACACCGCCCGTC  ...  Hexanauplia  Calanoida   \n",
       "1      insert primer name   GTACACACCGCCCGTC  ...  Hexanauplia  Calanoida   \n",
       "2      insert primer name   GTACACACCGCCCGTC  ...  Hexanauplia  Calanoida   \n",
       "3      insert primer name   GTACACACCGCCCGTC  ...  Hexanauplia  Calanoida   \n",
       "4      insert primer name   GTACACACCGCCCGTC  ...  Hexanauplia  Calanoida   \n",
       "\n",
       "          family        genus   basisOfRecord  \\\n",
       "0  Paracalanidae  Paracalanus  MaterialSample   \n",
       "1  Paracalanidae  Paracalanus  MaterialSample   \n",
       "2  Paracalanidae  Paracalanus  MaterialSample   \n",
       "3  Paracalanidae  Paracalanus  MaterialSample   \n",
       "4  Paracalanidae  Paracalanus  MaterialSample   \n",
       "\n",
       "                                    materialSampleID  \\\n",
       "0  insert doi/SRA number (?) if sequence has been...   \n",
       "1  insert doi/SRA number (?) if sequence has been...   \n",
       "2  insert doi/SRA number (?) if sequence has been...   \n",
       "3  insert doi/SRA number (?) if sequence has been...   \n",
       "4  insert doi/SRA number (?) if sequence has been...   \n",
       "\n",
       "                               identificationRemarks  \\\n",
       "0  insert the GenBank version used, the fact that...   \n",
       "1  insert the GenBank version used, the fact that...   \n",
       "2  insert the GenBank version used, the fact that...   \n",
       "3  insert the GenBank version used, the fact that...   \n",
       "4  insert the GenBank version used, the fact that...   \n",
       "\n",
       "                            identificationReferences organismQuantity  \\\n",
       "0  insert links/dois for published papers, and li...            14825   \n",
       "1  insert links/dois for published papers, and li...            16094   \n",
       "2  insert links/dois for published papers, and li...            22459   \n",
       "3  insert links/dois for published papers, and li...            19312   \n",
       "4  insert links/dois for published papers, and li...            16491   \n",
       "\n",
       "  organismQuantityType  \n",
       "0   DNA sequence reads  \n",
       "1   DNA sequence reads  \n",
       "2   DNA sequence reads  \n",
       "3   DNA sequence reads  \n",
       "4   DNA sequence reads  \n",
       "\n",
       "[5 rows x 28 columns]"
      ]
     },
     "execution_count": 140,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## organismQuantity (number of reads)\n",
    "\n",
    "occ['organismQuantity'] = plate['Reads']\n",
    "occ['organismQuantityType'] = 'DNA sequence reads'\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Note** that there are 215537 rows where the number of reads is 0. These will need to be dropped at the end of the script."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(280440, 29)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>pcr_primer_name_forward</th>\n",
       "      <th>pcr_primer_forward</th>\n",
       "      <th>...</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>basisOfRecord</th>\n",
       "      <th>materialSampleID</th>\n",
       "      <th>identificationRemarks</th>\n",
       "      <th>identificationReferences</th>\n",
       "      <th>organismQuantity</th>\n",
       "      <th>organismQuantityType</th>\n",
       "      <th>sampleSizeValue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>...</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>insert doi/SRA number (?) if sequence has been...</td>\n",
       "      <td>insert the GenBank version used, the fact that...</td>\n",
       "      <td>insert links/dois for published papers, and li...</td>\n",
       "      <td>14825</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>85600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>...</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>insert doi/SRA number (?) if sequence has been...</td>\n",
       "      <td>insert the GenBank version used, the fact that...</td>\n",
       "      <td>insert links/dois for published papers, and li...</td>\n",
       "      <td>16094</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>90702</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>...</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>insert doi/SRA number (?) if sequence has been...</td>\n",
       "      <td>insert the GenBank version used, the fact that...</td>\n",
       "      <td>insert links/dois for published papers, and li...</td>\n",
       "      <td>22459</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>130275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>...</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>insert doi/SRA number (?) if sequence has been...</td>\n",
       "      <td>insert the GenBank version used, the fact that...</td>\n",
       "      <td>insert links/dois for published papers, and li...</td>\n",
       "      <td>19312</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>147220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>...</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>insert doi/SRA number (?) if sequence has been...</td>\n",
       "      <td>insert the GenBank version used, the fact that...</td>\n",
       "      <td>insert links/dois for published papers, and li...</td>\n",
       "      <td>16491</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>121419</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude                env_broad_scale  \\\n",
       "0          -121.848  marine biome  (ENVO:00000447)   \n",
       "1          -121.848  marine biome  (ENVO:00000447)   \n",
       "2          -121.848  marine biome  (ENVO:00000447)   \n",
       "3          -121.846  marine biome  (ENVO:00000447)   \n",
       "4          -121.846  marine biome  (ENVO:00000447)   \n",
       "\n",
       "                 env_local_scale  \\\n",
       "0  coastal water (ENVO:00001250)   \n",
       "1  coastal water (ENVO:00001250)   \n",
       "2  coastal water (ENVO:00001250)   \n",
       "3  coastal water (ENVO:00001250)   \n",
       "4  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  pcr_primer_name_forward pcr_primer_forward  ...      order         family  \\\n",
       "0      insert primer name   GTACACACCGCCCGTC  ...  Calanoida  Paracalanidae   \n",
       "1      insert primer name   GTACACACCGCCCGTC  ...  Calanoida  Paracalanidae   \n",
       "2      insert primer name   GTACACACCGCCCGTC  ...  Calanoida  Paracalanidae   \n",
       "3      insert primer name   GTACACACCGCCCGTC  ...  Calanoida  Paracalanidae   \n",
       "4      insert primer name   GTACACACCGCCCGTC  ...  Calanoida  Paracalanidae   \n",
       "\n",
       "         genus   basisOfRecord  \\\n",
       "0  Paracalanus  MaterialSample   \n",
       "1  Paracalanus  MaterialSample   \n",
       "2  Paracalanus  MaterialSample   \n",
       "3  Paracalanus  MaterialSample   \n",
       "4  Paracalanus  MaterialSample   \n",
       "\n",
       "                                    materialSampleID  \\\n",
       "0  insert doi/SRA number (?) if sequence has been...   \n",
       "1  insert doi/SRA number (?) if sequence has been...   \n",
       "2  insert doi/SRA number (?) if sequence has been...   \n",
       "3  insert doi/SRA number (?) if sequence has been...   \n",
       "4  insert doi/SRA number (?) if sequence has been...   \n",
       "\n",
       "                               identificationRemarks  \\\n",
       "0  insert the GenBank version used, the fact that...   \n",
       "1  insert the GenBank version used, the fact that...   \n",
       "2  insert the GenBank version used, the fact that...   \n",
       "3  insert the GenBank version used, the fact that...   \n",
       "4  insert the GenBank version used, the fact that...   \n",
       "\n",
       "                            identificationReferences organismQuantity  \\\n",
       "0  insert links/dois for published papers, and li...            14825   \n",
       "1  insert links/dois for published papers, and li...            16094   \n",
       "2  insert links/dois for published papers, and li...            22459   \n",
       "3  insert links/dois for published papers, and li...            19312   \n",
       "4  insert links/dois for published papers, and li...            16491   \n",
       "\n",
       "  organismQuantityType sampleSizeValue  \n",
       "0   DNA sequence reads           85600  \n",
       "1   DNA sequence reads           90702  \n",
       "2   DNA sequence reads          130275  \n",
       "3   DNA sequence reads          147220  \n",
       "4   DNA sequence reads          121419  \n",
       "\n",
       "[5 rows x 29 columns]"
      ]
     },
     "execution_count": 141,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## sampleSizeValue\n",
    "\n",
    "count_by_seq = plate.groupby('Sequence_ID', as_index=False)['Reads'].sum()\n",
    "occ = occ.merge(count_by_seq, how='left', left_on='eventID', right_on='Sequence_ID')\n",
    "occ.drop(columns='Sequence_ID', inplace=True)\n",
    "occ.rename(columns={'Reads':'sampleSizeValue'}, inplace=True)\n",
    "print(occ.shape)\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>pcr_primer_name_forward</th>\n",
       "      <th>pcr_primer_forward</th>\n",
       "      <th>...</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>basisOfRecord</th>\n",
       "      <th>materialSampleID</th>\n",
       "      <th>identificationRemarks</th>\n",
       "      <th>identificationReferences</th>\n",
       "      <th>organismQuantity</th>\n",
       "      <th>organismQuantityType</th>\n",
       "      <th>sampleSizeValue</th>\n",
       "      <th>sampleSizeUnit</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>...</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>insert doi/SRA number (?) if sequence has been...</td>\n",
       "      <td>insert the GenBank version used, the fact that...</td>\n",
       "      <td>insert links/dois for published papers, and li...</td>\n",
       "      <td>14825</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>85600</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>...</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>insert doi/SRA number (?) if sequence has been...</td>\n",
       "      <td>insert the GenBank version used, the fact that...</td>\n",
       "      <td>insert links/dois for published papers, and li...</td>\n",
       "      <td>16094</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>90702</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>...</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>insert doi/SRA number (?) if sequence has been...</td>\n",
       "      <td>insert the GenBank version used, the fact that...</td>\n",
       "      <td>insert links/dois for published papers, and li...</td>\n",
       "      <td>22459</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>130275</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>...</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>insert doi/SRA number (?) if sequence has been...</td>\n",
       "      <td>insert the GenBank version used, the fact that...</td>\n",
       "      <td>insert links/dois for published papers, and li...</td>\n",
       "      <td>19312</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>147220</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>insert primer name</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>...</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>insert doi/SRA number (?) if sequence has been...</td>\n",
       "      <td>insert the GenBank version used, the fact that...</td>\n",
       "      <td>insert links/dois for published papers, and li...</td>\n",
       "      <td>16491</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>121419</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude                env_broad_scale  \\\n",
       "0          -121.848  marine biome  (ENVO:00000447)   \n",
       "1          -121.848  marine biome  (ENVO:00000447)   \n",
       "2          -121.848  marine biome  (ENVO:00000447)   \n",
       "3          -121.846  marine biome  (ENVO:00000447)   \n",
       "4          -121.846  marine biome  (ENVO:00000447)   \n",
       "\n",
       "                 env_local_scale  \\\n",
       "0  coastal water (ENVO:00001250)   \n",
       "1  coastal water (ENVO:00001250)   \n",
       "2  coastal water (ENVO:00001250)   \n",
       "3  coastal water (ENVO:00001250)   \n",
       "4  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  pcr_primer_name_forward pcr_primer_forward  ...         family        genus  \\\n",
       "0      insert primer name   GTACACACCGCCCGTC  ...  Paracalanidae  Paracalanus   \n",
       "1      insert primer name   GTACACACCGCCCGTC  ...  Paracalanidae  Paracalanus   \n",
       "2      insert primer name   GTACACACCGCCCGTC  ...  Paracalanidae  Paracalanus   \n",
       "3      insert primer name   GTACACACCGCCCGTC  ...  Paracalanidae  Paracalanus   \n",
       "4      insert primer name   GTACACACCGCCCGTC  ...  Paracalanidae  Paracalanus   \n",
       "\n",
       "    basisOfRecord                                   materialSampleID  \\\n",
       "0  MaterialSample  insert doi/SRA number (?) if sequence has been...   \n",
       "1  MaterialSample  insert doi/SRA number (?) if sequence has been...   \n",
       "2  MaterialSample  insert doi/SRA number (?) if sequence has been...   \n",
       "3  MaterialSample  insert doi/SRA number (?) if sequence has been...   \n",
       "4  MaterialSample  insert doi/SRA number (?) if sequence has been...   \n",
       "\n",
       "                               identificationRemarks  \\\n",
       "0  insert the GenBank version used, the fact that...   \n",
       "1  insert the GenBank version used, the fact that...   \n",
       "2  insert the GenBank version used, the fact that...   \n",
       "3  insert the GenBank version used, the fact that...   \n",
       "4  insert the GenBank version used, the fact that...   \n",
       "\n",
       "                            identificationReferences organismQuantity  \\\n",
       "0  insert links/dois for published papers, and li...            14825   \n",
       "1  insert links/dois for published papers, and li...            16094   \n",
       "2  insert links/dois for published papers, and li...            22459   \n",
       "3  insert links/dois for published papers, and li...            19312   \n",
       "4  insert links/dois for published papers, and li...            16491   \n",
       "\n",
       "  organismQuantityType sampleSizeValue      sampleSizeUnit  \n",
       "0   DNA sequence reads           85600  DNA sequence reads  \n",
       "1   DNA sequence reads           90702  DNA sequence reads  \n",
       "2   DNA sequence reads          130275  DNA sequence reads  \n",
       "3   DNA sequence reads          147220  DNA sequence reads  \n",
       "4   DNA sequence reads          121419  DNA sequence reads  \n",
       "\n",
       "[5 rows x 30 columns]"
      ]
     },
     "execution_count": 142,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## sampleSizeUnit\n",
    "\n",
    "occ['sampleSizeUnit'] = 'DNA sequence reads'\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**I'm still a little unclear about the difference between materialSampleID and associatedSequences. Maybe I can ask Abby about this.**\n",
    "\n",
    "## Tidy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(59625, 30)\n"
     ]
    }
   ],
   "source": [
    "## Drop records where scientificName = NaN (no taxonomic information was available) and where organismQuantity = 0\n",
    "\n",
    "occ = occ[(occ['scientificName'].isna() == False) & (occ['organismQuantity'] > 0)]\n",
    "print(occ.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Note** that the vast majority of this data has 0 reads (215537 records out of 280440, or about 75%). I guess that makes sense - it's sparse data - but it might be good to check?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Replace NaN values in text fields with ''\n",
    "\n",
    "text_fields = ['eventID', 'env_broad_scale', 'env_local_scale', 'env_medium', 'target_gene', 'pcr_primer_name_forward', 'pcr_primer_forward', \n",
    "               'pcr_primer_name_reverse', 'pcr_primer_reverse', 'pcr_primer_reference', 'sop', 'DNA_sequence', 'scientificName', 'kingdom', \n",
    "               'phylum', 'class', 'order', 'family', 'genus', 'basisOfRecord', 'materialSampleID', 'identificationRemarks', 'identificationReferences', \n",
    "               'organismQuantityType', 'sampleSizeUnit']\n",
    "occ[text_fields] = occ[text_fields].replace(np.nan, '')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Save"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Save\n",
    "\n",
    "occ.to_csv('eDNA_practice_plate_occ_20210108.csv', index=False, na_rep='NaN')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Questions\n",
    "\n",
    "1. What's up with the set of FilterID's that don't follow the formula? (E.g. JD10706C1_0m_1)\n",
    "2. \"Eukaryota\" is a very common entry in the Kingdom field, but it's not a Kingdom, it's a Domain. Is this a problem? If so, is there some way to get the actual Kingdom information, or retain the Eukaryota classification elsewhere in the data set? **Yes, it's a problem. All occurrences have to have some entry in the scientific name column that matches on WoRMS, even though this is not clear from the GBIF guide. As such, things in the Kingdom column should be true Kingdoms, if possible. Maybe Katie can make sure this happens when she gets information off of GenBank? Additionally, if the only taxonomic information for an organism is \"Eukaryota\", just put \"Biota\" in the scientificName column (which will match on WoRMS - http://marinespecies.org/aphia.php?p=taxdetails&id=1) and NaN everywhere else. This is also possibly an option for records where no taxonomic data at all is available. Note that Pieter is willing to run the data through OBIS's automated system to see how it manages matching on WoRMS. Also, he says there's a way to query the WoRMS API using GenBank ID (AphiaRecordByExternalID - https://www.marinespecies.org/rest/AphiaRecordByExternalID/399303?type=ncbi). Finally, GenBank IDs can likely be included in a taxonID field. Or the taxonConceptID field. To follow a related issue on GBIF's GitHub: https://github.com/gbif/doc-publishing-dna-derived-data/issues/35**\n",
    "3. Difference between materialSampleID and associatedSequences term definitions? **Abby is emailing contacts at GBIF to get more information on this (1/13). But also, maybe Katie would understand the definitions better than I do? After talking with Pieter and Saara, it sounds like materialSampleID is probably the ID of the actual physical water sample, if it is frozen somewhere. associatedSequences would be the biosampleID from SRA. If I want, I could get in touch with the developers of the GBIF guide and let them know that this is confusing.**\n",
    "4. How might one include the control data in DwC format? Abby suggested including it in the MoF file somehow? **No one knows. Abby will connect me to her contact at OBIS who is working through an eDNA data set; they might be at least thinking about this issue (1/13). Pieter and Saara contributed a number of questions/suggestions:**\n",
    "    **1. Has the control information been incorporated into this data set already? Are the control data still useful if you're dealing with the fully processed data set?**\n",
    "    **2. Are the control data archived and therefore accessible via biosampleID?**\n",
    "    **3. At worst, the records could be included in the data set with no scientificName and no coordinates to make it clear they're not regular occurrences.**\n",
    "    \n",
    "    **These questions would be best addressed to Katie.**\n",
    "\n",
    "Other than that, just waiting for updated metadata file from Katie.\n",
    "\n",
    "**Note** a new DwC extension is being developed for sequence derived data: https://rs.gbif.org/sandbox/extension/dna_derived_data.xml\n",
    "\n",
    "Also **note**: We can choose for these data to become a use case (i.e. an example for others trying to prepare their sequence-derived data for submission to OBIS). This could be done with part or all of the data set, and would mean that your conversion process would be available as an example on GitHub. Would Francisco like to do this? For examples, look here: https://github.com/tdwg/dwc-for-biologging/wiki"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "venv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
