{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# eDNA conversion of practice data - draft 3\n",
    "\n",
    "**Resources:**\n",
    "- https://docs.gbif-uat.org/publishing-dna-derived-data/1.0/en/#mapping-metabarcoding-edna-and-barcoding-data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Imports\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import random\n",
    "\n",
    "from datetime import datetime # for handling dates\n",
    "import pytz # for handling time zones"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Ensure my general functions for the MPA data integration project can be imported, and import them\n",
    "\n",
    "import sys\n",
    "sys.path.insert(0, \"C:\\\\Users\\\\dianalg\\\\PycharmProjects\\\\PythonScripts\\\\eDNA\")\n",
    "\n",
    "import WoRMS # functions for querying WoRMS REST API"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(280440, 11)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ASV</th>\n",
       "      <th>FilterID</th>\n",
       "      <th>Sequence_ID</th>\n",
       "      <th>Reads</th>\n",
       "      <th>Kingdom</th>\n",
       "      <th>Phylum</th>\n",
       "      <th>Class</th>\n",
       "      <th>Order</th>\n",
       "      <th>Family</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>05114c01_12_edna_1</td>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>14825</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>unassigned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>05114c01_12_edna_2</td>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>16094</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>unassigned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>05114c01_12_edna_3</td>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>22459</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>unassigned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>11216c01_12_edna_1</td>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>19312</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>unassigned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>11216c01_12_edna_2</td>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>16491</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>unassigned</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 ASV            FilterID  \\\n",
       "0  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  05114c01_12_edna_1   \n",
       "1  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  05114c01_12_edna_2   \n",
       "2  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  05114c01_12_edna_3   \n",
       "3  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  11216c01_12_edna_1   \n",
       "4  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  11216c01_12_edna_2   \n",
       "\n",
       "            Sequence_ID  Reads    Kingdom      Phylum        Class      Order  \\\n",
       "0  05114c01_12_edna_1_S  14825  Eukaryota  Arthropoda  Hexanauplia  Calanoida   \n",
       "1  05114c01_12_edna_2_S  16094  Eukaryota  Arthropoda  Hexanauplia  Calanoida   \n",
       "2  05114c01_12_edna_3_S  22459  Eukaryota  Arthropoda  Hexanauplia  Calanoida   \n",
       "3  11216c01_12_edna_1_S  19312  Eukaryota  Arthropoda  Hexanauplia  Calanoida   \n",
       "4  11216c01_12_edna_2_S  16491  Eukaryota  Arthropoda  Hexanauplia  Calanoida   \n",
       "\n",
       "          Family        Genus     Species  \n",
       "0  Paracalanidae  Paracalanus  unassigned  \n",
       "1  Paracalanidae  Paracalanus  unassigned  \n",
       "2  Paracalanidae  Paracalanus  unassigned  \n",
       "3  Paracalanidae  Paracalanus  unassigned  \n",
       "4  Paracalanidae  Paracalanus  unassigned  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Plate data\n",
    "\n",
    "plate = pd.read_csv('Plate_S_ASV_OBIS_data.csv')\n",
    "print(plate.shape)\n",
    "plate.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**FilterID** = SAMPLING_cruise (cruise where the sample was collected) + \n",
    "    SAMPLING_station_number (in this case c + a cast number because these data were collected via CTD cast) + _ +\n",
    "    SAMPLING_bottle (bottle number, if multiple samples were taken from a single bottle, there may be replicates indicated by a, b, c, etc.) + _ +\n",
    "    edna (filter indicator. edna means that the standard filter for the standard edna pipeline, i.e. a standard PBDF filter, was used. other options include hplc) + _ +\n",
    "    1, 2 or 3 (PCR replicate)\n",
    "\n",
    "To summarize, **FilterID** = cruise number + cast number + bottle number + filter indicator + replicate number.\n",
    "\n",
    "**Note** that not all IDs exactly follow this formula. Example:\n",
    "- JD10706C1_0m_1\n",
    "\n",
    "**Sequence_ID** adds the plate indicator to the end of FilterID. In this case, all samples are from plate S. Plates are labeled A-Z, AA-ZZ, etc."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(4711, 79)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ASV</th>\n",
       "      <th>Kingdom</th>\n",
       "      <th>Phylum</th>\n",
       "      <th>Class</th>\n",
       "      <th>Order</th>\n",
       "      <th>Family</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Species</th>\n",
       "      <th>14213c01_12_edna_1</th>\n",
       "      <th>14213c01_12_edna_2</th>\n",
       "      <th>...</th>\n",
       "      <th>JD33306C1_0m_3</th>\n",
       "      <th>pcrblank2</th>\n",
       "      <th>EB_20161121</th>\n",
       "      <th>EB_20161228</th>\n",
       "      <th>EB_20170117</th>\n",
       "      <th>pcrblank3</th>\n",
       "      <th>CB_CANON160925_1</th>\n",
       "      <th>CB_CANON160925_2</th>\n",
       "      <th>CB_CANON160925_3</th>\n",
       "      <th>ArtComm2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ASV_1</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>9</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>956</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1312</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>197</td>\n",
       "      <td>3</td>\n",
       "      <td>210</td>\n",
       "      <td>44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ASV_2</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Calanidae</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>4</td>\n",
       "      <td>7</td>\n",
       "      <td>...</td>\n",
       "      <td>358</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>535</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>118</td>\n",
       "      <td>162</td>\n",
       "      <td>139</td>\n",
       "      <td>5976</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ASV_3</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>Gymnodiniales</td>\n",
       "      <td>Gymnodiniaceae</td>\n",
       "      <td>Akashiwo</td>\n",
       "      <td>Akashiwo sanguinea</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>795</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>35</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ASV_4</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>Gymnodiniales</td>\n",
       "      <td>Gymnodiniaceae</td>\n",
       "      <td>Cochlodinium</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>82065</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>836</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ASV_5</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>148</td>\n",
       "      <td>147</td>\n",
       "      <td>...</td>\n",
       "      <td>14688</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>36</td>\n",
       "      <td>401</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>11</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 79 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     ASV    Kingdom      Phylum        Class          Order          Family  \\\n",
       "0  ASV_1  Eukaryota  Arthropoda  Hexanauplia      Calanoida   Paracalanidae   \n",
       "1  ASV_2  Eukaryota  Arthropoda  Hexanauplia      Calanoida       Calanidae   \n",
       "2  ASV_3  Eukaryota     unknown  Dinophyceae  Gymnodiniales  Gymnodiniaceae   \n",
       "3  ASV_4  Eukaryota     unknown  Dinophyceae  Gymnodiniales  Gymnodiniaceae   \n",
       "4  ASV_5  Eukaryota     unknown  Dinophyceae     unassigned      unassigned   \n",
       "\n",
       "          Genus             Species  14213c01_12_edna_1  14213c01_12_edna_2  \\\n",
       "0   Paracalanus          unassigned                   9                   4   \n",
       "1    unassigned          unassigned                   4                   7   \n",
       "2      Akashiwo  Akashiwo sanguinea                   2                   3   \n",
       "3  Cochlodinium          unassigned                   2                   3   \n",
       "4    unassigned          unassigned                 148                 147   \n",
       "\n",
       "   ...  JD33306C1_0m_3  pcrblank2  EB_20161121  EB_20161228  EB_20170117  \\\n",
       "0  ...             956          0            0         1312            1   \n",
       "1  ...             358          0            0          535            0   \n",
       "2  ...             795          0            0            0           35   \n",
       "3  ...           82065          0            0            0          836   \n",
       "4  ...           14688          0            0           36          401   \n",
       "\n",
       "   pcrblank3  CB_CANON160925_1  CB_CANON160925_2  CB_CANON160925_3  ArtComm2  \n",
       "0          0               197                 3               210        44  \n",
       "1          0               118               162               139      5976  \n",
       "2          0                 1                 0                 1         8  \n",
       "3          0                 1                 0                 0         7  \n",
       "4          0                 0                 4                11         7  \n",
       "\n",
       "[5 rows x 79 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Taxonomy table\n",
    "\n",
    "taxa = pd.read_csv('Filtered_ASV_taxa_table_all.csv')\n",
    "print(taxa.shape)\n",
    "taxa.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This shows the number of reads for each ASV detected for each FilterID (columns), plus the matched taxa information for each ASV, if available. \n",
    "\n",
    "Some columns are associated with controls:\n",
    "- **CB** = collection blank which was taken on the ship. Water that should be \"clean\" was passed through the filter, so there should be no reads in this sample, although it tends to be the \"dirtiest\" control. It checks for contamination during filtration, I think.\n",
    "- **EB** = extraction blank which was taken in the lab but contains no DNA (i.e. a negative control), so there should be no reads in this sample. It checks if contamination occurred during DNA extraction (lab work pre-PCR).\n",
    "- **pcrblank** = A sample that went through PCR but contained no input DNA (i.e. a negative control), so there should be no reads in this sample. It checks if contamination occurred during PCR.\n",
    "- **ArtComm** = artificial community that went through PCR and contained DNA from species that should not be in the study system (i.e. a positive control), so you know what results you expect.\n",
    "\n",
    "To see these columns, use:\n",
    "```python\n",
    "for col in taxa.columns:\n",
    "    if col not in plate['FilterID'].unique():\n",
    "        print(col)\n",
    "```\n",
    "\n",
    "Additionally, taxonomic columns include the following designations:\n",
    "- **unknown** = GenBank couldn't give a scientifically-agreed-upon name for a given taxonomic level. I.e., either the name doesn't exist, or there isn't enough scientific consensus to give a name.\n",
    "- **no_hit** = BLAST did not find any hits for the ASV.\n",
    "- **unassigned** = The ASV got BLAST hits, but the post-processing program Megan6 didn't assign the ASV to any taxonomic group.\n",
    "- **g_** or **s_** = After Megan there's an additional filtering step where Genus or Species designations are only left if they were assigned with a certain, high level of confidence or higher. If this threshold is not met, a g_ or s_ is given."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(60, 72)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>library</th>\n",
       "      <th>tag_sequence</th>\n",
       "      <th>primer_sequence_F</th>\n",
       "      <th>primer_sequence_R</th>\n",
       "      <th>R1</th>\n",
       "      <th>R2</th>\n",
       "      <th>PlateID</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>locus</th>\n",
       "      <th>...</th>\n",
       "      <th>sop</th>\n",
       "      <th>pcr_primer_name_forward</th>\n",
       "      <th>pcr_primer_name_reverse</th>\n",
       "      <th>pcr_primer_reference</th>\n",
       "      <th>seq_meth</th>\n",
       "      <th>sequencing_facility</th>\n",
       "      <th>seqID</th>\n",
       "      <th>identificationRemarks</th>\n",
       "      <th>identificationReferences</th>\n",
       "      <th>FilterID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>14213c01_12_eDNA_1</td>\n",
       "      <td>S1</td>\n",
       "      <td>ACGAGACTGATT</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>14213c01_12_edna_1_S1_L001_R1_001.fastq.gz</td>\n",
       "      <td>14213c01_12_edna_1_S1_L001_R2_001.fastq.gz</td>\n",
       "      <td>S</td>\n",
       "      <td>environmental</td>\n",
       "      <td>18S</td>\n",
       "      <td>...</td>\n",
       "      <td>ext:dx.doi.org/10.17504/protocols.io.xjufknw;a...</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>Amaral-Zettler et al. (2009)</td>\n",
       "      <td>NGS Illumina Miseq</td>\n",
       "      <td>Stanford</td>\n",
       "      <td>14213c01_12_edna_1_S</td>\n",
       "      <td>Genbank nr September 20 2017;&gt;80% identity</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>14213c01_12_edna</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>14213c01_12_eDNA_2</td>\n",
       "      <td>S2</td>\n",
       "      <td>GAATACCAAGTC</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>14213c01_12_edna_2_S2_L001_R1_001.fastq.gz</td>\n",
       "      <td>14213c01_12_edna_2_S2_L001_R2_001.fastq.gz</td>\n",
       "      <td>S</td>\n",
       "      <td>environmental</td>\n",
       "      <td>18S</td>\n",
       "      <td>...</td>\n",
       "      <td>ext:dx.doi.org/10.17504/protocols.io.xjufknw;a...</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>Amaral-Zettler et al. (2009)</td>\n",
       "      <td>NGS Illumina Miseq</td>\n",
       "      <td>Stanford</td>\n",
       "      <td>14213c01_12_edna_2_S</td>\n",
       "      <td>Genbank nr September 20 2017;&gt;80% identity</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>14213c01_12_edna</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>14213c01_12_eDNA_3</td>\n",
       "      <td>S3</td>\n",
       "      <td>CGAGGGAAAGTC</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>14213c01_12_edna_3_S3_L001_R1_001.fastq.gz</td>\n",
       "      <td>14213c01_12_edna_3_S3_L001_R2_001.fastq.gz</td>\n",
       "      <td>S</td>\n",
       "      <td>environmental</td>\n",
       "      <td>18S</td>\n",
       "      <td>...</td>\n",
       "      <td>ext:dx.doi.org/10.17504/protocols.io.xjufknw;a...</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>Amaral-Zettler et al. (2009)</td>\n",
       "      <td>NGS Illumina Miseq</td>\n",
       "      <td>Stanford</td>\n",
       "      <td>14213c01_12_edna_3_S</td>\n",
       "      <td>Genbank nr September 20 2017;&gt;80% identity</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>14213c01_12_edna</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>22013c01_12_eDNA_1</td>\n",
       "      <td>S4</td>\n",
       "      <td>GAACACTTTGGA</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>22013c01_12_edna_1_S4_L001_R1_001.fastq.gz</td>\n",
       "      <td>22013c01_12_edna_1_S4_L001_R2_001.fastq.gz</td>\n",
       "      <td>S</td>\n",
       "      <td>environmental</td>\n",
       "      <td>18S</td>\n",
       "      <td>...</td>\n",
       "      <td>ext:dx.doi.org/10.17504/protocols.io.xjufknw;a...</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>Amaral-Zettler et al. (2009)</td>\n",
       "      <td>NGS Illumina Miseq</td>\n",
       "      <td>Stanford</td>\n",
       "      <td>22013c01_12_edna_1_S</td>\n",
       "      <td>Genbank nr September 20 2017;&gt;80% identity</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>22013c01_12_edna</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>22013c01_12_eDNA_2</td>\n",
       "      <td>S5</td>\n",
       "      <td>ACTCACAGGAAT</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>22013c01_12_edna_2_S5_L001_R1_001.fastq.gz</td>\n",
       "      <td>22013c01_12_edna_2_S5_L001_R2_001.fastq.gz</td>\n",
       "      <td>S</td>\n",
       "      <td>environmental</td>\n",
       "      <td>18S</td>\n",
       "      <td>...</td>\n",
       "      <td>ext:dx.doi.org/10.17504/protocols.io.xjufknw;a...</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>Amaral-Zettler et al. (2009)</td>\n",
       "      <td>NGS Illumina Miseq</td>\n",
       "      <td>Stanford</td>\n",
       "      <td>22013c01_12_edna_2_S</td>\n",
       "      <td>Genbank nr September 20 2017;&gt;80% identity</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>22013c01_12_edna</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 72 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          sample_name library  tag_sequence primer_sequence_F  \\\n",
       "0  14213c01_12_eDNA_1      S1  ACGAGACTGATT  GTACACACCGCCCGTC   \n",
       "1  14213c01_12_eDNA_2      S2  GAATACCAAGTC  GTACACACCGCCCGTC   \n",
       "2  14213c01_12_eDNA_3      S3  CGAGGGAAAGTC  GTACACACCGCCCGTC   \n",
       "3  22013c01_12_eDNA_1      S4  GAACACTTTGGA  GTACACACCGCCCGTC   \n",
       "4  22013c01_12_eDNA_2      S5  ACTCACAGGAAT  GTACACACCGCCCGTC   \n",
       "\n",
       "          primer_sequence_R                                          R1  \\\n",
       "0  TGATCCTTCTGCAGGTTCACCTAC  14213c01_12_edna_1_S1_L001_R1_001.fastq.gz   \n",
       "1  TGATCCTTCTGCAGGTTCACCTAC  14213c01_12_edna_2_S2_L001_R1_001.fastq.gz   \n",
       "2  TGATCCTTCTGCAGGTTCACCTAC  14213c01_12_edna_3_S3_L001_R1_001.fastq.gz   \n",
       "3  TGATCCTTCTGCAGGTTCACCTAC  22013c01_12_edna_1_S4_L001_R1_001.fastq.gz   \n",
       "4  TGATCCTTCTGCAGGTTCACCTAC  22013c01_12_edna_2_S5_L001_R1_001.fastq.gz   \n",
       "\n",
       "                                           R2 PlateID    sample_type locus  \\\n",
       "0  14213c01_12_edna_1_S1_L001_R2_001.fastq.gz       S  environmental   18S   \n",
       "1  14213c01_12_edna_2_S2_L001_R2_001.fastq.gz       S  environmental   18S   \n",
       "2  14213c01_12_edna_3_S3_L001_R2_001.fastq.gz       S  environmental   18S   \n",
       "3  22013c01_12_edna_1_S4_L001_R2_001.fastq.gz       S  environmental   18S   \n",
       "4  22013c01_12_edna_2_S5_L001_R2_001.fastq.gz       S  environmental   18S   \n",
       "\n",
       "   ...                                                sop  \\\n",
       "0  ...  ext:dx.doi.org/10.17504/protocols.io.xjufknw;a...   \n",
       "1  ...  ext:dx.doi.org/10.17504/protocols.io.xjufknw;a...   \n",
       "2  ...  ext:dx.doi.org/10.17504/protocols.io.xjufknw;a...   \n",
       "3  ...  ext:dx.doi.org/10.17504/protocols.io.xjufknw;a...   \n",
       "4  ...  ext:dx.doi.org/10.17504/protocols.io.xjufknw;a...   \n",
       "\n",
       "  pcr_primer_name_forward pcr_primer_name_reverse  \\\n",
       "0                   1391f                   EukBr   \n",
       "1                   1391f                   EukBr   \n",
       "2                   1391f                   EukBr   \n",
       "3                   1391f                   EukBr   \n",
       "4                   1391f                   EukBr   \n",
       "\n",
       "           pcr_primer_reference            seq_meth sequencing_facility  \\\n",
       "0  Amaral-Zettler et al. (2009)  NGS Illumina Miseq            Stanford   \n",
       "1  Amaral-Zettler et al. (2009)  NGS Illumina Miseq            Stanford   \n",
       "2  Amaral-Zettler et al. (2009)  NGS Illumina Miseq            Stanford   \n",
       "3  Amaral-Zettler et al. (2009)  NGS Illumina Miseq            Stanford   \n",
       "4  Amaral-Zettler et al. (2009)  NGS Illumina Miseq            Stanford   \n",
       "\n",
       "                  seqID                       identificationRemarks  \\\n",
       "0  14213c01_12_edna_1_S  Genbank nr September 20 2017;>80% identity   \n",
       "1  14213c01_12_edna_2_S  Genbank nr September 20 2017;>80% identity   \n",
       "2  14213c01_12_edna_3_S  Genbank nr September 20 2017;>80% identity   \n",
       "3  22013c01_12_edna_1_S  Genbank nr September 20 2017;>80% identity   \n",
       "4  22013c01_12_edna_2_S  Genbank nr September 20 2017;>80% identity   \n",
       "\n",
       "                            identificationReferences          FilterID  \n",
       "0  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...  14213c01_12_edna  \n",
       "1  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...  14213c01_12_edna  \n",
       "2  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...  14213c01_12_edna  \n",
       "3  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...  22013c01_12_edna  \n",
       "4  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...  22013c01_12_edna  \n",
       "\n",
       "[5 rows x 72 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Plate metadata\n",
    "\n",
    "plate_meta = pd.read_csv('Plate_S_meta_OBIS_data_022321.csv')\n",
    "print(plate_meta.shape)\n",
    "plate_meta.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Conversion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(280440, 1)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID\n",
       "0  05114c01_12_edna_1_S\n",
       "1  05114c01_12_edna_2_S\n",
       "2  05114c01_12_edna_3_S\n",
       "3  11216c01_12_edna_1_S\n",
       "4  11216c01_12_edna_2_S"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## eventID\n",
    "\n",
    "occ = pd.DataFrame({'eventID':plate['Sequence_ID']})\n",
    "print(occ.shape)\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>primer_sequence_forward</th>\n",
       "      <th>primer_sequence_reverse</th>\n",
       "      <th>pcr_primer_name_forward</th>\n",
       "      <th>pcr_primer_name_reverse</th>\n",
       "      <th>pcr_primer_reference</th>\n",
       "      <th>sop</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2/20/2014 15:33</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>Amaral-Zettler et al. (2009)</td>\n",
       "      <td>ext:dx.doi.org/10.17504/protocols.io.xjufknw;a...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2/20/2014 15:33</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>Amaral-Zettler et al. (2009)</td>\n",
       "      <td>ext:dx.doi.org/10.17504/protocols.io.xjufknw;a...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2/20/2014 15:33</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>Amaral-Zettler et al. (2009)</td>\n",
       "      <td>ext:dx.doi.org/10.17504/protocols.io.xjufknw;a...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>4/21/2016 14:39</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>Amaral-Zettler et al. (2009)</td>\n",
       "      <td>ext:dx.doi.org/10.17504/protocols.io.xjufknw;a...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>4/21/2016 14:39</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>Amaral-Zettler et al. (2009)</td>\n",
       "      <td>ext:dx.doi.org/10.17504/protocols.io.xjufknw;a...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID        eventDate  decimalLatitude  decimalLongitude  \\\n",
       "0  05114c01_12_edna_1_S  2/20/2014 15:33          36.7958          -121.848   \n",
       "1  05114c01_12_edna_2_S  2/20/2014 15:33          36.7958          -121.848   \n",
       "2  05114c01_12_edna_3_S  2/20/2014 15:33          36.7958          -121.848   \n",
       "3  11216c01_12_edna_1_S  4/21/2016 14:39          36.7962          -121.846   \n",
       "4  11216c01_12_edna_2_S  4/21/2016 14:39          36.7962          -121.846   \n",
       "\n",
       "                 env_broad_scale                env_local_scale  \\\n",
       "0  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   \n",
       "1  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   \n",
       "2  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   \n",
       "3  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   \n",
       "4  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  primer_sequence_forward   primer_sequence_reverse pcr_primer_name_forward  \\\n",
       "0        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC                   1391f   \n",
       "1        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC                   1391f   \n",
       "2        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC                   1391f   \n",
       "3        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC                   1391f   \n",
       "4        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC                   1391f   \n",
       "\n",
       "  pcr_primer_name_reverse          pcr_primer_reference  \\\n",
       "0                   EukBr  Amaral-Zettler et al. (2009)   \n",
       "1                   EukBr  Amaral-Zettler et al. (2009)   \n",
       "2                   EukBr  Amaral-Zettler et al. (2009)   \n",
       "3                   EukBr  Amaral-Zettler et al. (2009)   \n",
       "4                   EukBr  Amaral-Zettler et al. (2009)   \n",
       "\n",
       "                                                 sop  \n",
       "0  ext:dx.doi.org/10.17504/protocols.io.xjufknw;a...  \n",
       "1  ext:dx.doi.org/10.17504/protocols.io.xjufknw;a...  \n",
       "2  ext:dx.doi.org/10.17504/protocols.io.xjufknw;a...  \n",
       "3  ext:dx.doi.org/10.17504/protocols.io.xjufknw;a...  \n",
       "4  ext:dx.doi.org/10.17504/protocols.io.xjufknw;a...  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Merge with plate_meta to obtain eventDate, decimalLatitude, decimalLongitude, and other columns that can be added directly from metadata\n",
    "\n",
    "metadata_cols = [\n",
    "    'seqID',\n",
    "    'SAMPLING_date_time', \n",
    "    'SAMPLING_dec_lat', \n",
    "    'SAMPLING_dec_lon',\n",
    "    'env_biome',\n",
    "    'env_feature',\n",
    "    'env_material',\n",
    "    'locus',\n",
    "    'primer_sequence_F',\n",
    "    'primer_sequence_R',\n",
    "    'pcr_primer_name_forward',\n",
    "    'pcr_primer_name_reverse',\n",
    "    'pcr_primer_reference',\n",
    "    'sop',\n",
    "]\n",
    "\n",
    "dwc_cols = [\n",
    "    'eventID',\n",
    "    'eventDate', \n",
    "    'decimalLatitude', \n",
    "    'decimalLongitude',\n",
    "    'env_broad_scale',\n",
    "    'env_local_scale',\n",
    "    'env_medium',\n",
    "    'target_gene',\n",
    "    'primer_sequence_forward',\n",
    "    'primer_sequence_reverse',\n",
    "    'pcr_primer_name_forward',\n",
    "    'pcr_primer_name_reverse',\n",
    "    'pcr_primer_reference',\n",
    "    'sop',\n",
    "]\n",
    "\n",
    "occ = occ.merge(plate_meta[metadata_cols], how='left', left_on='eventID', right_on='seqID')\n",
    "occ.drop(columns='seqID', inplace=True)\n",
    "occ.columns = dwc_cols\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Note** that for pcr_primer_reference, I think it would be best to either put a DOI that links to the article, or the entire article citation (i.e. title, journal name included). This isn't a requirement, it just seems like the safest way to ensure the correct article would be identified."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>primer_sequence_forward</th>\n",
       "      <th>primer_sequence_reverse</th>\n",
       "      <th>pcr_primer_name_forward</th>\n",
       "      <th>pcr_primer_name_reverse</th>\n",
       "      <th>pcr_primer_reference</th>\n",
       "      <th>sop</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>Amaral-Zettler et al. (2009)</td>\n",
       "      <td>ext:dx.doi.org/10.17504/protocols.io.xjufknw;a...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>Amaral-Zettler et al. (2009)</td>\n",
       "      <td>ext:dx.doi.org/10.17504/protocols.io.xjufknw;a...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>Amaral-Zettler et al. (2009)</td>\n",
       "      <td>ext:dx.doi.org/10.17504/protocols.io.xjufknw;a...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>Amaral-Zettler et al. (2009)</td>\n",
       "      <td>ext:dx.doi.org/10.17504/protocols.io.xjufknw;a...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>Amaral-Zettler et al. (2009)</td>\n",
       "      <td>ext:dx.doi.org/10.17504/protocols.io.xjufknw;a...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude                env_broad_scale  \\\n",
       "0          -121.848  marine biome  (ENVO:00000447)   \n",
       "1          -121.848  marine biome  (ENVO:00000447)   \n",
       "2          -121.848  marine biome  (ENVO:00000447)   \n",
       "3          -121.846  marine biome  (ENVO:00000447)   \n",
       "4          -121.846  marine biome  (ENVO:00000447)   \n",
       "\n",
       "                 env_local_scale  \\\n",
       "0  coastal water (ENVO:00001250)   \n",
       "1  coastal water (ENVO:00001250)   \n",
       "2  coastal water (ENVO:00001250)   \n",
       "3  coastal water (ENVO:00001250)   \n",
       "4  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  primer_sequence_forward   primer_sequence_reverse pcr_primer_name_forward  \\\n",
       "0        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC                   1391f   \n",
       "1        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC                   1391f   \n",
       "2        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC                   1391f   \n",
       "3        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC                   1391f   \n",
       "4        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC                   1391f   \n",
       "\n",
       "  pcr_primer_name_reverse          pcr_primer_reference  \\\n",
       "0                   EukBr  Amaral-Zettler et al. (2009)   \n",
       "1                   EukBr  Amaral-Zettler et al. (2009)   \n",
       "2                   EukBr  Amaral-Zettler et al. (2009)   \n",
       "3                   EukBr  Amaral-Zettler et al. (2009)   \n",
       "4                   EukBr  Amaral-Zettler et al. (2009)   \n",
       "\n",
       "                                                 sop  \n",
       "0  ext:dx.doi.org/10.17504/protocols.io.xjufknw;a...  \n",
       "1  ext:dx.doi.org/10.17504/protocols.io.xjufknw;a...  \n",
       "2  ext:dx.doi.org/10.17504/protocols.io.xjufknw;a...  \n",
       "3  ext:dx.doi.org/10.17504/protocols.io.xjufknw;a...  \n",
       "4  ext:dx.doi.org/10.17504/protocols.io.xjufknw;a...  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Format eventDate\n",
    "\n",
    "pst = pytz.timezone('America/Los_Angeles')\n",
    "eventDate = [pst.localize(datetime.strptime(dt, '%m/%d/%Y %H:%M')).isoformat() for dt in occ['eventDate']]\n",
    "occ['eventDate'] = eventDate\n",
    "\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>primer_sequence_forward</th>\n",
       "      <th>primer_sequence_reverse</th>\n",
       "      <th>pcr_primer_name_forward</th>\n",
       "      <th>pcr_primer_name_reverse</th>\n",
       "      <th>pcr_primer_reference</th>\n",
       "      <th>sop</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>Amaral-Zettler et al. (2009)</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw | dx....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>Amaral-Zettler et al. (2009)</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw | dx....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>Amaral-Zettler et al. (2009)</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw | dx....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>Amaral-Zettler et al. (2009)</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw | dx....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>Amaral-Zettler et al. (2009)</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw | dx....</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude                env_broad_scale  \\\n",
       "0          -121.848  marine biome  (ENVO:00000447)   \n",
       "1          -121.848  marine biome  (ENVO:00000447)   \n",
       "2          -121.848  marine biome  (ENVO:00000447)   \n",
       "3          -121.846  marine biome  (ENVO:00000447)   \n",
       "4          -121.846  marine biome  (ENVO:00000447)   \n",
       "\n",
       "                 env_local_scale  \\\n",
       "0  coastal water (ENVO:00001250)   \n",
       "1  coastal water (ENVO:00001250)   \n",
       "2  coastal water (ENVO:00001250)   \n",
       "3  coastal water (ENVO:00001250)   \n",
       "4  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  primer_sequence_forward   primer_sequence_reverse pcr_primer_name_forward  \\\n",
       "0        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC                   1391f   \n",
       "1        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC                   1391f   \n",
       "2        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC                   1391f   \n",
       "3        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC                   1391f   \n",
       "4        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC                   1391f   \n",
       "\n",
       "  pcr_primer_name_reverse          pcr_primer_reference  \\\n",
       "0                   EukBr  Amaral-Zettler et al. (2009)   \n",
       "1                   EukBr  Amaral-Zettler et al. (2009)   \n",
       "2                   EukBr  Amaral-Zettler et al. (2009)   \n",
       "3                   EukBr  Amaral-Zettler et al. (2009)   \n",
       "4                   EukBr  Amaral-Zettler et al. (2009)   \n",
       "\n",
       "                                                 sop  \n",
       "0  dx.doi.org/10.17504/protocols.io.xjufknw | dx....  \n",
       "1  dx.doi.org/10.17504/protocols.io.xjufknw | dx....  \n",
       "2  dx.doi.org/10.17504/protocols.io.xjufknw | dx....  \n",
       "3  dx.doi.org/10.17504/protocols.io.xjufknw | dx....  \n",
       "4  dx.doi.org/10.17504/protocols.io.xjufknw | dx....  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Clean sop\n",
    "\n",
    "# Replace semicolons with pipes\n",
    "occ['sop'] = occ['sop'].str.replace(';', ' | ')\n",
    "\n",
    "# Remove clarifying text that Katie added\n",
    "occ['sop'] = occ['sop'].replace('ext:|amp:|bioinformatics:', '', regex=True).iloc[0]\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>primer_sequence_forward</th>\n",
       "      <th>primer_sequence_reverse</th>\n",
       "      <th>pcr_primer_name_forward</th>\n",
       "      <th>pcr_primer_name_reverse</th>\n",
       "      <th>pcr_primer_reference</th>\n",
       "      <th>sop</th>\n",
       "      <th>DNA_sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>Amaral-Zettler et al. (2009)</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw | dx....</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>Amaral-Zettler et al. (2009)</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw | dx....</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>Amaral-Zettler et al. (2009)</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw | dx....</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>Amaral-Zettler et al. (2009)</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw | dx....</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>Amaral-Zettler et al. (2009)</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw | dx....</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude                env_broad_scale  \\\n",
       "0          -121.848  marine biome  (ENVO:00000447)   \n",
       "1          -121.848  marine biome  (ENVO:00000447)   \n",
       "2          -121.848  marine biome  (ENVO:00000447)   \n",
       "3          -121.846  marine biome  (ENVO:00000447)   \n",
       "4          -121.846  marine biome  (ENVO:00000447)   \n",
       "\n",
       "                 env_local_scale  \\\n",
       "0  coastal water (ENVO:00001250)   \n",
       "1  coastal water (ENVO:00001250)   \n",
       "2  coastal water (ENVO:00001250)   \n",
       "3  coastal water (ENVO:00001250)   \n",
       "4  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  primer_sequence_forward   primer_sequence_reverse pcr_primer_name_forward  \\\n",
       "0        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC                   1391f   \n",
       "1        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC                   1391f   \n",
       "2        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC                   1391f   \n",
       "3        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC                   1391f   \n",
       "4        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC                   1391f   \n",
       "\n",
       "  pcr_primer_name_reverse          pcr_primer_reference  \\\n",
       "0                   EukBr  Amaral-Zettler et al. (2009)   \n",
       "1                   EukBr  Amaral-Zettler et al. (2009)   \n",
       "2                   EukBr  Amaral-Zettler et al. (2009)   \n",
       "3                   EukBr  Amaral-Zettler et al. (2009)   \n",
       "4                   EukBr  Amaral-Zettler et al. (2009)   \n",
       "\n",
       "                                                 sop  \\\n",
       "0  dx.doi.org/10.17504/protocols.io.xjufknw | dx....   \n",
       "1  dx.doi.org/10.17504/protocols.io.xjufknw | dx....   \n",
       "2  dx.doi.org/10.17504/protocols.io.xjufknw | dx....   \n",
       "3  dx.doi.org/10.17504/protocols.io.xjufknw | dx....   \n",
       "4  dx.doi.org/10.17504/protocols.io.xjufknw | dx....   \n",
       "\n",
       "                                        DNA_sequence  \n",
       "0  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  \n",
       "1  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  \n",
       "2  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  \n",
       "3  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  \n",
       "4  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Add DNA_sequence\n",
    "\n",
    "occ['DNA_sequence'] = plate['ASV']\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>primer_sequence_forward</th>\n",
       "      <th>primer_sequence_reverse</th>\n",
       "      <th>...</th>\n",
       "      <th>pcr_primer_reference</th>\n",
       "      <th>sop</th>\n",
       "      <th>DNA_sequence</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Amaral-Zettler et al. (2009)</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw | dx....</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Amaral-Zettler et al. (2009)</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw | dx....</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Amaral-Zettler et al. (2009)</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw | dx....</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Amaral-Zettler et al. (2009)</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw | dx....</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Amaral-Zettler et al. (2009)</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw | dx....</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude                env_broad_scale  \\\n",
       "0          -121.848  marine biome  (ENVO:00000447)   \n",
       "1          -121.848  marine biome  (ENVO:00000447)   \n",
       "2          -121.848  marine biome  (ENVO:00000447)   \n",
       "3          -121.846  marine biome  (ENVO:00000447)   \n",
       "4          -121.846  marine biome  (ENVO:00000447)   \n",
       "\n",
       "                 env_local_scale  \\\n",
       "0  coastal water (ENVO:00001250)   \n",
       "1  coastal water (ENVO:00001250)   \n",
       "2  coastal water (ENVO:00001250)   \n",
       "3  coastal water (ENVO:00001250)   \n",
       "4  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  primer_sequence_forward   primer_sequence_reverse  ...  \\\n",
       "0        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   \n",
       "1        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   \n",
       "2        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   \n",
       "3        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   \n",
       "4        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   \n",
       "\n",
       "           pcr_primer_reference  \\\n",
       "0  Amaral-Zettler et al. (2009)   \n",
       "1  Amaral-Zettler et al. (2009)   \n",
       "2  Amaral-Zettler et al. (2009)   \n",
       "3  Amaral-Zettler et al. (2009)   \n",
       "4  Amaral-Zettler et al. (2009)   \n",
       "\n",
       "                                                 sop  \\\n",
       "0  dx.doi.org/10.17504/protocols.io.xjufknw | dx....   \n",
       "1  dx.doi.org/10.17504/protocols.io.xjufknw | dx....   \n",
       "2  dx.doi.org/10.17504/protocols.io.xjufknw | dx....   \n",
       "3  dx.doi.org/10.17504/protocols.io.xjufknw | dx....   \n",
       "4  dx.doi.org/10.17504/protocols.io.xjufknw | dx....   \n",
       "\n",
       "                                        DNA_sequence scientificName  \\\n",
       "0  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...     unassigned   \n",
       "1  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...     unassigned   \n",
       "2  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...     unassigned   \n",
       "3  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...     unassigned   \n",
       "4  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...     unassigned   \n",
       "\n",
       "     kingdom      phylum        class      order         family        genus  \n",
       "0  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus  \n",
       "1  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus  \n",
       "2  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus  \n",
       "3  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus  \n",
       "4  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## scientificName, taxonomic info\n",
    "\n",
    "occ['scientificName'] = plate['Species']\n",
    "occ['kingdom'] = plate['Kingdom']\n",
    "occ['phylum'] = plate['Phylum']\n",
    "occ['class'] = plate['Class']\n",
    "occ['order'] = plate['Order']\n",
    "occ['family'] = plate['Family']\n",
    "occ['genus'] = plate['Genus']\n",
    "\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>primer_sequence_forward</th>\n",
       "      <th>primer_sequence_reverse</th>\n",
       "      <th>...</th>\n",
       "      <th>pcr_primer_reference</th>\n",
       "      <th>sop</th>\n",
       "      <th>DNA_sequence</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Amaral-Zettler et al. (2009)</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw | dx....</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Amaral-Zettler et al. (2009)</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw | dx....</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Amaral-Zettler et al. (2009)</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw | dx....</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Amaral-Zettler et al. (2009)</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw | dx....</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Amaral-Zettler et al. (2009)</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw | dx....</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude                env_broad_scale  \\\n",
       "0          -121.848  marine biome  (ENVO:00000447)   \n",
       "1          -121.848  marine biome  (ENVO:00000447)   \n",
       "2          -121.848  marine biome  (ENVO:00000447)   \n",
       "3          -121.846  marine biome  (ENVO:00000447)   \n",
       "4          -121.846  marine biome  (ENVO:00000447)   \n",
       "\n",
       "                 env_local_scale  \\\n",
       "0  coastal water (ENVO:00001250)   \n",
       "1  coastal water (ENVO:00001250)   \n",
       "2  coastal water (ENVO:00001250)   \n",
       "3  coastal water (ENVO:00001250)   \n",
       "4  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  primer_sequence_forward   primer_sequence_reverse  ...  \\\n",
       "0        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   \n",
       "1        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   \n",
       "2        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   \n",
       "3        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   \n",
       "4        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   \n",
       "\n",
       "           pcr_primer_reference  \\\n",
       "0  Amaral-Zettler et al. (2009)   \n",
       "1  Amaral-Zettler et al. (2009)   \n",
       "2  Amaral-Zettler et al. (2009)   \n",
       "3  Amaral-Zettler et al. (2009)   \n",
       "4  Amaral-Zettler et al. (2009)   \n",
       "\n",
       "                                                 sop  \\\n",
       "0  dx.doi.org/10.17504/protocols.io.xjufknw | dx....   \n",
       "1  dx.doi.org/10.17504/protocols.io.xjufknw | dx....   \n",
       "2  dx.doi.org/10.17504/protocols.io.xjufknw | dx....   \n",
       "3  dx.doi.org/10.17504/protocols.io.xjufknw | dx....   \n",
       "4  dx.doi.org/10.17504/protocols.io.xjufknw | dx....   \n",
       "\n",
       "                                        DNA_sequence scientificName  \\\n",
       "0  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...            NaN   \n",
       "1  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...            NaN   \n",
       "2  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...            NaN   \n",
       "3  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...            NaN   \n",
       "4  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...            NaN   \n",
       "\n",
       "     kingdom      phylum        class      order         family        genus  \n",
       "0  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus  \n",
       "1  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus  \n",
       "2  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus  \n",
       "3  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus  \n",
       "4  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Replace 'unknown', 'unassigned', etc. in scientificName and taxonomy columns with NaN\n",
    "\n",
    "cols = ['scientificName', 'kingdom', 'phylum', 'class', 'order', 'family', 'genus']\n",
    "occ[cols] = occ[cols].replace({'unassigned':np.nan,\n",
    "                              's_':np.nan,\n",
    "                              'g_':np.nan,\n",
    "                              'unknown':np.nan,\n",
    "                              'no_hit':np.nan})\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "323\n"
     ]
    }
   ],
   "source": [
    "## Get unique species names\n",
    "\n",
    "names = occ['scientificName'].unique()\n",
    "names = names[~pd.isnull(names)] # remove NaN\n",
    "print(len(names))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "These names **do** have to be WoRMS-approved in order to be submitted to OBIS. \n",
    "\n",
    "There are a number of entries in the scientificName column, like \"uncultured marine eukaryote\", \"eukaryote clone OLI11007\", and \"Acantharian sp. 6201\", that are **not proper Latin species names**. It seems reasonable to replace these with NaN as well. **I've used a simple rule to filter these out, and it would be worth checking it as the data set expands, just to be extra careful.**\n",
    "\n",
    "To see the entries:\n",
    "```python\n",
    "names = occ['scientificName'].unique()\n",
    "names = names[~pd.isnull(names)] # remove NaN\n",
    "for name in names:\n",
    "    words_in_name = name.split(' ')\n",
    "    if len(words_in_name) > 2:\n",
    "        print(name)\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Replace non-Latin species names with NaN\n",
    "\n",
    "# Get non-Latin names\n",
    "non_latin_names = []\n",
    "for name in names:\n",
    "    words_in_name = name.split(' ')\n",
    "    if len(words_in_name) > 2:\n",
    "        non_latin_names.append(name)\n",
    "non_latin_names_dict = {i:np.nan for i in non_latin_names}\n",
    "\n",
    "# Add any names that didn't get caught in the simple filter\n",
    "non_latin_names_dict['phototrophic eukaryote'] = np.nan\n",
    "non_latin_names_dict['Candida <clade Candida/Lodderomyces clade>'] = np.nan\n",
    "\n",
    "# Replace\n",
    "occ['scientificName'].replace(non_latin_names_dict, inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In addition, 52380 records **only give \"Eukaryota\" as the scientific name** (i.e. Eukaryota is in the kingdom field, and there is no more taxonomic information). These should be replaced with Biota (http://marinespecies.org/aphia.php?p=taxdetails&id=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Replace entries where kingdom = 'Eukaryota' with the WoRMS-approved 'Biota'\n",
    "\n",
    "occ.loc[occ['kingdom'] == 'Eukaryota', 'kingdom'] = 'Biota'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally **GenBank and WoRMS seem to disagree on their taxonomy**. Two examples:\n",
    "\n",
    "1. 120 records have \"Pelargonium\" as the genus. [On GenBank](https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?mode=Info&id=4030&lvl=3&lin=f&keep=1&srchmode=1&unlock) the full taxonomy is given as:\n",
    "    - superkingdom Eukaryota\n",
    "    - kingdom Viridiplantae (subkingdom on WoRMS)\n",
    "    - phylum Streptophyta (infrakingdom on WoRMS, phylum listed as Tracheophyta)\n",
    "    - subphylum Streptophytina (Spermatophytina on WoRMS)\n",
    "    - class Magnoliopsida (matches on WoRMS)\n",
    "        - this was indicated as 'unknown' in original data, i.e. GenBank couldn't give a scientifically-agreed-upon name for a given taxonomic level. **This doesn't make sense to me.**\n",
    "    - order Geraniales (matches on WoRMS)\n",
    "    - family Geraniaceae ([matches on WoRMS](http://marinespecies.org/aphia.php?p=taxdetails&id=382461))\n",
    "    - genus Pelargonium (does NOT match on WoRMS)\n",
    "\n",
    "\n",
    "2. 600 records have family Hemistasiidae (and no further taxonomic information). [On GenBank](https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?mode=Info&id=2015512&lvl=3&lin=f&keep=1&srchmode=1&unlock), the full taxonomy is given as:\n",
    "    - superkingdom Eukaryota\n",
    "    - phylum Euglenozoa (matches on WoRMS)\n",
    "    - order Diplonemea ([matches on WoRMS](http://marinespecies.org/aphia.php?p=taxdetails&id=582176))\n",
    "    - family Hemistasiidae (does NOT match on WoRMS)\n",
    "\n",
    "Interestingly, there are 60 records where species is Hemistasia phaeocysticola. These records also have family Hemistasiidae. The genus Hemistasia [matches on WoRMS](http://marinespecies.org/aphia.php?p=taxdetails&id=146165), but with a totally different taxonomy than [on GenBank](https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?mode=Undef&id=1503927&lvl=3&srchmode=1&keep=1&unlock):\n",
    "- kingdom Protozoa\n",
    "- phylum Euglenozoa\n",
    "- class Kinetoplastea\n",
    "- order not listed\n",
    "- family not listed\n",
    "\n",
    "**Is this messed up, or is this fine?**\n",
    "\n",
    "**In both examples above**, the lowest taxonomic rank name given by the dataset does not match on WoRMS. I therefore need a system for searching through the higher taxonomic ranks given, finding the lowest one that will match on WoRMS, and putting that name in the scientificName column. The following few code blocks do this."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Functions for finding the lowest available taxonomic rank that will match on WoRMS\n",
    "\n",
    "def fill_lowest_taxon(df, cols):\n",
    "    \"\"\" Takes the occurrence pandas data frame and fills missing values in scientificName with values from the first non-missing taxonomic rank column, listed in cols. \"\"\"\n",
    "    \n",
    "    cols.reverse()\n",
    "    \n",
    "    for col in cols[:-1]:\n",
    "        df['scientificName'] = df['scientificName'].combine_first(df[col])\n",
    "    \n",
    "    cols.reverse()\n",
    "    \n",
    "    return(df)\n",
    "\n",
    "def find_not_matched(df, name_dict):\n",
    "    \"\"\" Takes the occurrence pandas data frame and name_dict matching scientificName values with names on WoRMS and returns a list of names that did not match on WoRMS. \"\"\"\n",
    "    \n",
    "    not_matched = []\n",
    "    \n",
    "    for name in df['scientificName'].unique():\n",
    "        if name not in name_dict.keys():\n",
    "            not_matched.append(name)\n",
    "    \n",
    "    try:\n",
    "        not_matched.remove(np.nan)\n",
    "    except ValueError:\n",
    "        pass\n",
    "            \n",
    "    return(not_matched)\n",
    "\n",
    "def replace_not_matched(df, not_matched, cols):\n",
    "    \"\"\" Takes the occurrence pandas data frame and a list of scientificName values that did not match on WoRMS and replaces those values with NaN in the columns specified by cols. \"\"\"\n",
    "    \n",
    "    df[cols] = df[cols].replace(not_matched, np.nan)\n",
    "    \n",
    "    return(df)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Matching 756 names on WoRMS.\n",
      "Length of name_name_dict: 694\n",
      "Number of names not matched: 62\n",
      "Matching 36 names on WoRMS.\n",
      "Length of name_name_dict: 714\n",
      "Number of names not matched: 16\n",
      "Matching 12 names on WoRMS.\n",
      "Length of name_name_dict: 720\n",
      "Number of names not matched: 6\n",
      "Matching 3 names on WoRMS.\n",
      "Length of name_name_dict: 721\n",
      "Number of names not matched: 2\n",
      "Matching 1 names on WoRMS.\n",
      "Length of name_name_dict: 721\n",
      "Number of names not matched: 1\n",
      "Matching 0 names on WoRMS.\n",
      "Length of name_name_dict: 721\n",
      "Number of names not matched: 0\n"
     ]
    }
   ],
   "source": [
    "## Iterate to match lowest possible taxonomic rank on WoRMS\n",
    "\n",
    "# Note that cols (list of taxonomic column names) was defined when cleanin them above\n",
    "\n",
    "# Initialize dictionaries\n",
    "name_name_dict = {}\n",
    "name_id_dict = {}\n",
    "name_taxid_dict = {}\n",
    "name_class_dict = {}\n",
    "\n",
    "# Initialize not_matched\n",
    "not_matched = [1]\n",
    "\n",
    "# Iterate\n",
    "while len(not_matched) > 0:\n",
    "    \n",
    "    # Step 1 - fill\n",
    "    occ = fill_lowest_taxon(occ, cols)\n",
    "\n",
    "    # Step 2 - get names to match\n",
    "    to_match = find_not_matched(occ, name_name_dict)\n",
    "\n",
    "    # Step 3 - match\n",
    "    print('Matching {num} names on WoRMS.'.format(num = len(to_match)))\n",
    "    name_id, name_name, name_taxid, name_class = WoRMS.run_get_worms_from_scientific_name(to_match, verbose_flag=False)\n",
    "    name_id_dict = {**name_id_dict, **name_id}\n",
    "    name_name_dict = {**name_name_dict, **name_name}\n",
    "    name_taxid_dict = {**name_taxid_dict, **name_taxid}\n",
    "    name_class_dict = {**name_class_dict, **name_class}\n",
    "    print('Length of name_name_dict: {length}'.format(length = len(name_name_dict)))\n",
    "    \n",
    "    ## ---- Add in a progress bar? Also, a better/additional stopping criterion? What if not all names can be matched?\n",
    "\n",
    "    # Step 4 - get names that didn't match\n",
    "    not_matched = find_not_matched(occ, name_name_dict)\n",
    "    print('Number of names not matched: {num}'.format(num = len(not_matched)))\n",
    "\n",
    "    # Step 5 - replace these values with NaN\n",
    "    occ = replace_not_matched(occ, not_matched, cols)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Note** that there are 33360 records (~12% of the data) where no taxonomic information was obtained at all (i.e. scientificName is still NaN). **Pieter mentioned that I could possibly use 'Biota' for these records as well. Would that work?**\n",
    "\n",
    "```python\n",
    "occ[occ['scientificName'].isna() == True].shape\n",
    "```\n",
    "\n",
    "**These tentatively should be dropped, since they won't appear on OBIS/GBIF, but perhaps there's some argument to be made for leaving them in anyway?** I'm leaving them in for now.\n",
    "\n",
    "Finally, during the above process, **I totally messed up the taxonomy columns in order to obtain the best possible scientificName column**. I'll fix that below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Fix taxonomy columns\n",
    "\n",
    "# Replace with original data\n",
    "occ[cols[1:]] = plate[['Kingdom', 'Phylum', 'Class', 'Order', 'Family', 'Genus']].copy()\n",
    "\n",
    "# Replace missing data indicators in original data with empty strings ('')\n",
    "occ[cols[1:]] = occ[cols[1:]].replace({\n",
    "    'unassigned':'',\n",
    "    's_':'',\n",
    "    'g_':'',\n",
    "    'unknown':'',\n",
    "    'no_hit':''})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check using:\n",
    "\n",
    "```python\n",
    "occ[occ['scientificName'] == 'Diplonemea'].head()\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>primer_sequence_forward</th>\n",
       "      <th>primer_sequence_reverse</th>\n",
       "      <th>...</th>\n",
       "      <th>DNA_sequence</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>taxonID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude                env_broad_scale  \\\n",
       "0          -121.848  marine biome  (ENVO:00000447)   \n",
       "1          -121.848  marine biome  (ENVO:00000447)   \n",
       "2          -121.848  marine biome  (ENVO:00000447)   \n",
       "3          -121.846  marine biome  (ENVO:00000447)   \n",
       "4          -121.846  marine biome  (ENVO:00000447)   \n",
       "\n",
       "                 env_local_scale  \\\n",
       "0  coastal water (ENVO:00001250)   \n",
       "1  coastal water (ENVO:00001250)   \n",
       "2  coastal water (ENVO:00001250)   \n",
       "3  coastal water (ENVO:00001250)   \n",
       "4  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  primer_sequence_forward   primer_sequence_reverse  ...  \\\n",
       "0        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   \n",
       "1        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   \n",
       "2        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   \n",
       "3        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   \n",
       "4        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   \n",
       "\n",
       "                                        DNA_sequence scientificName  \\\n",
       "0  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...    Paracalanus   \n",
       "1  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...    Paracalanus   \n",
       "2  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...    Paracalanus   \n",
       "3  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...    Paracalanus   \n",
       "4  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...    Paracalanus   \n",
       "\n",
       "     kingdom      phylum        class      order         family        genus  \\\n",
       "0  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   \n",
       "1  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   \n",
       "2  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   \n",
       "3  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   \n",
       "4  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   \n",
       "\n",
       "                            scientificNameID   taxonID  \n",
       "0  urn:lsid:marinespecies.org:taxname:104196  104196.0  \n",
       "1  urn:lsid:marinespecies.org:taxname:104196  104196.0  \n",
       "2  urn:lsid:marinespecies.org:taxname:104196  104196.0  \n",
       "3  urn:lsid:marinespecies.org:taxname:104196  104196.0  \n",
       "4  urn:lsid:marinespecies.org:taxname:104196  104196.0  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Add scientific name-related columns\n",
    "\n",
    "occ['scientificNameID'] = occ['scientificName']\n",
    "occ['scientificNameID'].replace(name_id_dict, inplace=True)\n",
    "\n",
    "occ['taxonID'] = occ['scientificName']\n",
    "occ['taxonID'].replace(name_taxid_dict, inplace=True)\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Note** that Pieter said that taxonID or taxonConceptID could potentially be used to include a GenBank ID here (the latter seems like a better fit to me). **Can I request these from Katie? Also, where would I specify that this number is a GenBankID? In the taxonRemarks?**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>primer_sequence_forward</th>\n",
       "      <th>primer_sequence_reverse</th>\n",
       "      <th>...</th>\n",
       "      <th>DNA_sequence</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>taxonID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude                env_broad_scale  \\\n",
       "0          -121.848  marine biome  (ENVO:00000447)   \n",
       "1          -121.848  marine biome  (ENVO:00000447)   \n",
       "2          -121.848  marine biome  (ENVO:00000447)   \n",
       "3          -121.846  marine biome  (ENVO:00000447)   \n",
       "4          -121.846  marine biome  (ENVO:00000447)   \n",
       "\n",
       "                 env_local_scale  \\\n",
       "0  coastal water (ENVO:00001250)   \n",
       "1  coastal water (ENVO:00001250)   \n",
       "2  coastal water (ENVO:00001250)   \n",
       "3  coastal water (ENVO:00001250)   \n",
       "4  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  primer_sequence_forward   primer_sequence_reverse  ...  \\\n",
       "0        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   \n",
       "1        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   \n",
       "2        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   \n",
       "3        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   \n",
       "4        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   \n",
       "\n",
       "                                        DNA_sequence scientificName  \\\n",
       "0  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...    Paracalanus   \n",
       "1  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...    Paracalanus   \n",
       "2  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...    Paracalanus   \n",
       "3  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...    Paracalanus   \n",
       "4  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...    Paracalanus   \n",
       "\n",
       "     kingdom      phylum        class      order         family        genus  \\\n",
       "0  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   \n",
       "1  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   \n",
       "2  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   \n",
       "3  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   \n",
       "4  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   \n",
       "\n",
       "                            scientificNameID   taxonID  \n",
       "0  urn:lsid:marinespecies.org:taxname:104196  104196.0  \n",
       "1  urn:lsid:marinespecies.org:taxname:104196  104196.0  \n",
       "2  urn:lsid:marinespecies.org:taxname:104196  104196.0  \n",
       "3  urn:lsid:marinespecies.org:taxname:104196  104196.0  \n",
       "4  urn:lsid:marinespecies.org:taxname:104196  104196.0  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Replace scientificName with the name that matched on WoRMS\n",
    "\n",
    "occ['scientificName'].replace(name_name_dict, inplace=True)\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>primer_sequence_forward</th>\n",
       "      <th>primer_sequence_reverse</th>\n",
       "      <th>...</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>taxonID</th>\n",
       "      <th>nameAccordingTo</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196.0</td>\n",
       "      <td>WoRMS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196.0</td>\n",
       "      <td>WoRMS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196.0</td>\n",
       "      <td>WoRMS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196.0</td>\n",
       "      <td>WoRMS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196.0</td>\n",
       "      <td>WoRMS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude                env_broad_scale  \\\n",
       "0          -121.848  marine biome  (ENVO:00000447)   \n",
       "1          -121.848  marine biome  (ENVO:00000447)   \n",
       "2          -121.848  marine biome  (ENVO:00000447)   \n",
       "3          -121.846  marine biome  (ENVO:00000447)   \n",
       "4          -121.846  marine biome  (ENVO:00000447)   \n",
       "\n",
       "                 env_local_scale  \\\n",
       "0  coastal water (ENVO:00001250)   \n",
       "1  coastal water (ENVO:00001250)   \n",
       "2  coastal water (ENVO:00001250)   \n",
       "3  coastal water (ENVO:00001250)   \n",
       "4  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  primer_sequence_forward   primer_sequence_reverse  ... scientificName  \\\n",
       "0        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...    Paracalanus   \n",
       "1        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...    Paracalanus   \n",
       "2        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...    Paracalanus   \n",
       "3        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...    Paracalanus   \n",
       "4        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...    Paracalanus   \n",
       "\n",
       "     kingdom      phylum        class      order         family        genus  \\\n",
       "0  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   \n",
       "1  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   \n",
       "2  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   \n",
       "3  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   \n",
       "4  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   \n",
       "\n",
       "                            scientificNameID   taxonID nameAccordingTo  \n",
       "0  urn:lsid:marinespecies.org:taxname:104196  104196.0           WoRMS  \n",
       "1  urn:lsid:marinespecies.org:taxname:104196  104196.0           WoRMS  \n",
       "2  urn:lsid:marinespecies.org:taxname:104196  104196.0           WoRMS  \n",
       "3  urn:lsid:marinespecies.org:taxname:104196  104196.0           WoRMS  \n",
       "4  urn:lsid:marinespecies.org:taxname:104196  104196.0           WoRMS  \n",
       "\n",
       "[5 rows x 25 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Add nameAccordingTo\n",
    "\n",
    "occ['nameAccordingTo'] = 'WoRMS'\n",
    "\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>primer_sequence_forward</th>\n",
       "      <th>primer_sequence_reverse</th>\n",
       "      <th>...</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>taxonID</th>\n",
       "      <th>nameAccordingTo</th>\n",
       "      <th>basisOfRecord</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196.0</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>MaterialSample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196.0</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>MaterialSample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196.0</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>MaterialSample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196.0</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>MaterialSample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196.0</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>MaterialSample</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude                env_broad_scale  \\\n",
       "0          -121.848  marine biome  (ENVO:00000447)   \n",
       "1          -121.848  marine biome  (ENVO:00000447)   \n",
       "2          -121.848  marine biome  (ENVO:00000447)   \n",
       "3          -121.846  marine biome  (ENVO:00000447)   \n",
       "4          -121.846  marine biome  (ENVO:00000447)   \n",
       "\n",
       "                 env_local_scale  \\\n",
       "0  coastal water (ENVO:00001250)   \n",
       "1  coastal water (ENVO:00001250)   \n",
       "2  coastal water (ENVO:00001250)   \n",
       "3  coastal water (ENVO:00001250)   \n",
       "4  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  primer_sequence_forward   primer_sequence_reverse  ...    kingdom  \\\n",
       "0        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Eukaryota   \n",
       "1        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Eukaryota   \n",
       "2        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Eukaryota   \n",
       "3        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Eukaryota   \n",
       "4        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Eukaryota   \n",
       "\n",
       "       phylum        class      order         family        genus  \\\n",
       "0  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   \n",
       "1  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   \n",
       "2  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   \n",
       "3  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   \n",
       "4  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   \n",
       "\n",
       "                            scientificNameID   taxonID nameAccordingTo  \\\n",
       "0  urn:lsid:marinespecies.org:taxname:104196  104196.0           WoRMS   \n",
       "1  urn:lsid:marinespecies.org:taxname:104196  104196.0           WoRMS   \n",
       "2  urn:lsid:marinespecies.org:taxname:104196  104196.0           WoRMS   \n",
       "3  urn:lsid:marinespecies.org:taxname:104196  104196.0           WoRMS   \n",
       "4  urn:lsid:marinespecies.org:taxname:104196  104196.0           WoRMS   \n",
       "\n",
       "    basisOfRecord  \n",
       "0  MaterialSample  \n",
       "1  MaterialSample  \n",
       "2  MaterialSample  \n",
       "3  MaterialSample  \n",
       "4  MaterialSample  \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## basisOfRecord\n",
    "\n",
    "occ['basisOfRecord'] = 'MaterialSample'\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>primer_sequence_forward</th>\n",
       "      <th>primer_sequence_reverse</th>\n",
       "      <th>...</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>taxonID</th>\n",
       "      <th>nameAccordingTo</th>\n",
       "      <th>basisOfRecord</th>\n",
       "      <th>materialSampleID</th>\n",
       "      <th>identificationRemarks</th>\n",
       "      <th>identificationReferences</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196.0</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>insert ID for actual frozen sample if exists</td>\n",
       "      <td>Genbank nr September 20 2017;&gt;80% identity</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196.0</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>insert ID for actual frozen sample if exists</td>\n",
       "      <td>Genbank nr September 20 2017;&gt;80% identity</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196.0</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>insert ID for actual frozen sample if exists</td>\n",
       "      <td>Genbank nr September 20 2017;&gt;80% identity</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196.0</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>insert ID for actual frozen sample if exists</td>\n",
       "      <td>Genbank nr September 20 2017;&gt;80% identity</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196.0</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>insert ID for actual frozen sample if exists</td>\n",
       "      <td>Genbank nr September 20 2017;&gt;80% identity</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude                env_broad_scale  \\\n",
       "0          -121.848  marine biome  (ENVO:00000447)   \n",
       "1          -121.848  marine biome  (ENVO:00000447)   \n",
       "2          -121.848  marine biome  (ENVO:00000447)   \n",
       "3          -121.846  marine biome  (ENVO:00000447)   \n",
       "4          -121.846  marine biome  (ENVO:00000447)   \n",
       "\n",
       "                 env_local_scale  \\\n",
       "0  coastal water (ENVO:00001250)   \n",
       "1  coastal water (ENVO:00001250)   \n",
       "2  coastal water (ENVO:00001250)   \n",
       "3  coastal water (ENVO:00001250)   \n",
       "4  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  primer_sequence_forward   primer_sequence_reverse  ...      order  \\\n",
       "0        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Calanoida   \n",
       "1        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Calanoida   \n",
       "2        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Calanoida   \n",
       "3        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Calanoida   \n",
       "4        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Calanoida   \n",
       "\n",
       "          family        genus                           scientificNameID  \\\n",
       "0  Paracalanidae  Paracalanus  urn:lsid:marinespecies.org:taxname:104196   \n",
       "1  Paracalanidae  Paracalanus  urn:lsid:marinespecies.org:taxname:104196   \n",
       "2  Paracalanidae  Paracalanus  urn:lsid:marinespecies.org:taxname:104196   \n",
       "3  Paracalanidae  Paracalanus  urn:lsid:marinespecies.org:taxname:104196   \n",
       "4  Paracalanidae  Paracalanus  urn:lsid:marinespecies.org:taxname:104196   \n",
       "\n",
       "    taxonID nameAccordingTo   basisOfRecord  \\\n",
       "0  104196.0           WoRMS  MaterialSample   \n",
       "1  104196.0           WoRMS  MaterialSample   \n",
       "2  104196.0           WoRMS  MaterialSample   \n",
       "3  104196.0           WoRMS  MaterialSample   \n",
       "4  104196.0           WoRMS  MaterialSample   \n",
       "\n",
       "                               materialSampleID  \\\n",
       "0  insert ID for actual frozen sample if exists   \n",
       "1  insert ID for actual frozen sample if exists   \n",
       "2  insert ID for actual frozen sample if exists   \n",
       "3  insert ID for actual frozen sample if exists   \n",
       "4  insert ID for actual frozen sample if exists   \n",
       "\n",
       "                        identificationRemarks  \\\n",
       "0  Genbank nr September 20 2017;>80% identity   \n",
       "1  Genbank nr September 20 2017;>80% identity   \n",
       "2  Genbank nr September 20 2017;>80% identity   \n",
       "3  Genbank nr September 20 2017;>80% identity   \n",
       "4  Genbank nr September 20 2017;>80% identity   \n",
       "\n",
       "                            identificationReferences  \n",
       "0  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...  \n",
       "1  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...  \n",
       "2  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...  \n",
       "3  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...  \n",
       "4  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...  \n",
       "\n",
       "[5 rows x 29 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Add materialSampleID, identificationRemarks, identificationReferences -- can probably do this through merging metadata when Katie adds this info\n",
    "\n",
    "occ['materialSampleID'] = 'insert ID for actual frozen sample if exists'\n",
    "\n",
    "occ = occ.merge(plate_meta[['seqID', 'identificationRemarks', 'identificationReferences']], how='left', left_on='eventID', right_on='seqID')\n",
    "occ.drop(columns='seqID', inplace=True)\n",
    "\n",
    "# Clean identificationRemarks ---- make this more human-readable after clarifying what exactly it means w/ Katie\n",
    "\n",
    "# Clean identificationReferences\n",
    "occ['identificationReferences'] = occ['identificationReferences'].str.replace('; ', ' | ')\n",
    "occ['identificationReferences'] = occ['identificationReferences'].str.replace(' D.*\\)', '', regex=True)\n",
    "\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Note** that I should ask Katie to explain the content of identificationRemarks, and then make the field contents more descriptive. **I also need to ask if these actual water samples are frozen somewhere. If they are, that information can be included under materialSampleID.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>primer_sequence_forward</th>\n",
       "      <th>primer_sequence_reverse</th>\n",
       "      <th>...</th>\n",
       "      <th>genus</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>taxonID</th>\n",
       "      <th>nameAccordingTo</th>\n",
       "      <th>basisOfRecord</th>\n",
       "      <th>materialSampleID</th>\n",
       "      <th>identificationRemarks</th>\n",
       "      <th>identificationReferences</th>\n",
       "      <th>organismQuantity</th>\n",
       "      <th>organismQuantityType</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196.0</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>insert ID for actual frozen sample if exists</td>\n",
       "      <td>Genbank nr September 20 2017;&gt;80% identity</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>14825</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196.0</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>insert ID for actual frozen sample if exists</td>\n",
       "      <td>Genbank nr September 20 2017;&gt;80% identity</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>16094</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196.0</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>insert ID for actual frozen sample if exists</td>\n",
       "      <td>Genbank nr September 20 2017;&gt;80% identity</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>22459</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196.0</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>insert ID for actual frozen sample if exists</td>\n",
       "      <td>Genbank nr September 20 2017;&gt;80% identity</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>19312</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196.0</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>insert ID for actual frozen sample if exists</td>\n",
       "      <td>Genbank nr September 20 2017;&gt;80% identity</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>16491</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude                env_broad_scale  \\\n",
       "0          -121.848  marine biome  (ENVO:00000447)   \n",
       "1          -121.848  marine biome  (ENVO:00000447)   \n",
       "2          -121.848  marine biome  (ENVO:00000447)   \n",
       "3          -121.846  marine biome  (ENVO:00000447)   \n",
       "4          -121.846  marine biome  (ENVO:00000447)   \n",
       "\n",
       "                 env_local_scale  \\\n",
       "0  coastal water (ENVO:00001250)   \n",
       "1  coastal water (ENVO:00001250)   \n",
       "2  coastal water (ENVO:00001250)   \n",
       "3  coastal water (ENVO:00001250)   \n",
       "4  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  primer_sequence_forward   primer_sequence_reverse  ...        genus  \\\n",
       "0        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Paracalanus   \n",
       "1        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Paracalanus   \n",
       "2        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Paracalanus   \n",
       "3        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Paracalanus   \n",
       "4        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Paracalanus   \n",
       "\n",
       "                            scientificNameID   taxonID nameAccordingTo  \\\n",
       "0  urn:lsid:marinespecies.org:taxname:104196  104196.0           WoRMS   \n",
       "1  urn:lsid:marinespecies.org:taxname:104196  104196.0           WoRMS   \n",
       "2  urn:lsid:marinespecies.org:taxname:104196  104196.0           WoRMS   \n",
       "3  urn:lsid:marinespecies.org:taxname:104196  104196.0           WoRMS   \n",
       "4  urn:lsid:marinespecies.org:taxname:104196  104196.0           WoRMS   \n",
       "\n",
       "    basisOfRecord                              materialSampleID  \\\n",
       "0  MaterialSample  insert ID for actual frozen sample if exists   \n",
       "1  MaterialSample  insert ID for actual frozen sample if exists   \n",
       "2  MaterialSample  insert ID for actual frozen sample if exists   \n",
       "3  MaterialSample  insert ID for actual frozen sample if exists   \n",
       "4  MaterialSample  insert ID for actual frozen sample if exists   \n",
       "\n",
       "                        identificationRemarks  \\\n",
       "0  Genbank nr September 20 2017;>80% identity   \n",
       "1  Genbank nr September 20 2017;>80% identity   \n",
       "2  Genbank nr September 20 2017;>80% identity   \n",
       "3  Genbank nr September 20 2017;>80% identity   \n",
       "4  Genbank nr September 20 2017;>80% identity   \n",
       "\n",
       "                            identificationReferences organismQuantity  \\\n",
       "0  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...            14825   \n",
       "1  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...            16094   \n",
       "2  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...            22459   \n",
       "3  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...            19312   \n",
       "4  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...            16491   \n",
       "\n",
       "  organismQuantityType  \n",
       "0   DNA sequence reads  \n",
       "1   DNA sequence reads  \n",
       "2   DNA sequence reads  \n",
       "3   DNA sequence reads  \n",
       "4   DNA sequence reads  \n",
       "\n",
       "[5 rows x 31 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## organismQuantity (number of reads)\n",
    "\n",
    "occ['organismQuantity'] = plate['Reads']\n",
    "occ['organismQuantityType'] = 'DNA sequence reads'\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Note** that there are 215537 rows where the number of reads is 0. These will need to be dropped at the end of the script."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(280440, 32)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>primer_sequence_forward</th>\n",
       "      <th>primer_sequence_reverse</th>\n",
       "      <th>...</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>taxonID</th>\n",
       "      <th>nameAccordingTo</th>\n",
       "      <th>basisOfRecord</th>\n",
       "      <th>materialSampleID</th>\n",
       "      <th>identificationRemarks</th>\n",
       "      <th>identificationReferences</th>\n",
       "      <th>organismQuantity</th>\n",
       "      <th>organismQuantityType</th>\n",
       "      <th>sampleSizeValue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196.0</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>insert ID for actual frozen sample if exists</td>\n",
       "      <td>Genbank nr September 20 2017;&gt;80% identity</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>14825</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>85600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196.0</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>insert ID for actual frozen sample if exists</td>\n",
       "      <td>Genbank nr September 20 2017;&gt;80% identity</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>16094</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>90702</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196.0</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>insert ID for actual frozen sample if exists</td>\n",
       "      <td>Genbank nr September 20 2017;&gt;80% identity</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>22459</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>130275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196.0</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>insert ID for actual frozen sample if exists</td>\n",
       "      <td>Genbank nr September 20 2017;&gt;80% identity</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>19312</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>147220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196.0</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>insert ID for actual frozen sample if exists</td>\n",
       "      <td>Genbank nr September 20 2017;&gt;80% identity</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>16491</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>121419</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude                env_broad_scale  \\\n",
       "0          -121.848  marine biome  (ENVO:00000447)   \n",
       "1          -121.848  marine biome  (ENVO:00000447)   \n",
       "2          -121.848  marine biome  (ENVO:00000447)   \n",
       "3          -121.846  marine biome  (ENVO:00000447)   \n",
       "4          -121.846  marine biome  (ENVO:00000447)   \n",
       "\n",
       "                 env_local_scale  \\\n",
       "0  coastal water (ENVO:00001250)   \n",
       "1  coastal water (ENVO:00001250)   \n",
       "2  coastal water (ENVO:00001250)   \n",
       "3  coastal water (ENVO:00001250)   \n",
       "4  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  primer_sequence_forward   primer_sequence_reverse  ...  \\\n",
       "0        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   \n",
       "1        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   \n",
       "2        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   \n",
       "3        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   \n",
       "4        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   \n",
       "\n",
       "                            scientificNameID   taxonID nameAccordingTo  \\\n",
       "0  urn:lsid:marinespecies.org:taxname:104196  104196.0           WoRMS   \n",
       "1  urn:lsid:marinespecies.org:taxname:104196  104196.0           WoRMS   \n",
       "2  urn:lsid:marinespecies.org:taxname:104196  104196.0           WoRMS   \n",
       "3  urn:lsid:marinespecies.org:taxname:104196  104196.0           WoRMS   \n",
       "4  urn:lsid:marinespecies.org:taxname:104196  104196.0           WoRMS   \n",
       "\n",
       "    basisOfRecord                              materialSampleID  \\\n",
       "0  MaterialSample  insert ID for actual frozen sample if exists   \n",
       "1  MaterialSample  insert ID for actual frozen sample if exists   \n",
       "2  MaterialSample  insert ID for actual frozen sample if exists   \n",
       "3  MaterialSample  insert ID for actual frozen sample if exists   \n",
       "4  MaterialSample  insert ID for actual frozen sample if exists   \n",
       "\n",
       "                        identificationRemarks  \\\n",
       "0  Genbank nr September 20 2017;>80% identity   \n",
       "1  Genbank nr September 20 2017;>80% identity   \n",
       "2  Genbank nr September 20 2017;>80% identity   \n",
       "3  Genbank nr September 20 2017;>80% identity   \n",
       "4  Genbank nr September 20 2017;>80% identity   \n",
       "\n",
       "                            identificationReferences organismQuantity  \\\n",
       "0  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...            14825   \n",
       "1  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...            16094   \n",
       "2  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...            22459   \n",
       "3  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...            19312   \n",
       "4  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...            16491   \n",
       "\n",
       "  organismQuantityType sampleSizeValue  \n",
       "0   DNA sequence reads           85600  \n",
       "1   DNA sequence reads           90702  \n",
       "2   DNA sequence reads          130275  \n",
       "3   DNA sequence reads          147220  \n",
       "4   DNA sequence reads          121419  \n",
       "\n",
       "[5 rows x 32 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## sampleSizeValue\n",
    "\n",
    "count_by_seq = plate.groupby('Sequence_ID', as_index=False)['Reads'].sum()\n",
    "occ = occ.merge(count_by_seq, how='left', left_on='eventID', right_on='Sequence_ID')\n",
    "occ.drop(columns='Sequence_ID', inplace=True)\n",
    "occ.rename(columns={'Reads':'sampleSizeValue'}, inplace=True)\n",
    "print(occ.shape)\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>primer_sequence_forward</th>\n",
       "      <th>primer_sequence_reverse</th>\n",
       "      <th>...</th>\n",
       "      <th>taxonID</th>\n",
       "      <th>nameAccordingTo</th>\n",
       "      <th>basisOfRecord</th>\n",
       "      <th>materialSampleID</th>\n",
       "      <th>identificationRemarks</th>\n",
       "      <th>identificationReferences</th>\n",
       "      <th>organismQuantity</th>\n",
       "      <th>organismQuantityType</th>\n",
       "      <th>sampleSizeValue</th>\n",
       "      <th>sampleSizeUnit</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>104196.0</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>insert ID for actual frozen sample if exists</td>\n",
       "      <td>Genbank nr September 20 2017;&gt;80% identity</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>14825</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>85600</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>104196.0</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>insert ID for actual frozen sample if exists</td>\n",
       "      <td>Genbank nr September 20 2017;&gt;80% identity</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>16094</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>90702</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>104196.0</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>insert ID for actual frozen sample if exists</td>\n",
       "      <td>Genbank nr September 20 2017;&gt;80% identity</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>22459</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>130275</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>104196.0</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>insert ID for actual frozen sample if exists</td>\n",
       "      <td>Genbank nr September 20 2017;&gt;80% identity</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>19312</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>147220</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>104196.0</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>insert ID for actual frozen sample if exists</td>\n",
       "      <td>Genbank nr September 20 2017;&gt;80% identity</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>16491</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>121419</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude                env_broad_scale  \\\n",
       "0          -121.848  marine biome  (ENVO:00000447)   \n",
       "1          -121.848  marine biome  (ENVO:00000447)   \n",
       "2          -121.848  marine biome  (ENVO:00000447)   \n",
       "3          -121.846  marine biome  (ENVO:00000447)   \n",
       "4          -121.846  marine biome  (ENVO:00000447)   \n",
       "\n",
       "                 env_local_scale  \\\n",
       "0  coastal water (ENVO:00001250)   \n",
       "1  coastal water (ENVO:00001250)   \n",
       "2  coastal water (ENVO:00001250)   \n",
       "3  coastal water (ENVO:00001250)   \n",
       "4  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  primer_sequence_forward   primer_sequence_reverse  ...   taxonID  \\\n",
       "0        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  104196.0   \n",
       "1        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  104196.0   \n",
       "2        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  104196.0   \n",
       "3        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  104196.0   \n",
       "4        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  104196.0   \n",
       "\n",
       "  nameAccordingTo   basisOfRecord  \\\n",
       "0           WoRMS  MaterialSample   \n",
       "1           WoRMS  MaterialSample   \n",
       "2           WoRMS  MaterialSample   \n",
       "3           WoRMS  MaterialSample   \n",
       "4           WoRMS  MaterialSample   \n",
       "\n",
       "                               materialSampleID  \\\n",
       "0  insert ID for actual frozen sample if exists   \n",
       "1  insert ID for actual frozen sample if exists   \n",
       "2  insert ID for actual frozen sample if exists   \n",
       "3  insert ID for actual frozen sample if exists   \n",
       "4  insert ID for actual frozen sample if exists   \n",
       "\n",
       "                        identificationRemarks  \\\n",
       "0  Genbank nr September 20 2017;>80% identity   \n",
       "1  Genbank nr September 20 2017;>80% identity   \n",
       "2  Genbank nr September 20 2017;>80% identity   \n",
       "3  Genbank nr September 20 2017;>80% identity   \n",
       "4  Genbank nr September 20 2017;>80% identity   \n",
       "\n",
       "                            identificationReferences organismQuantity  \\\n",
       "0  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...            14825   \n",
       "1  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...            16094   \n",
       "2  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...            22459   \n",
       "3  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...            19312   \n",
       "4  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...            16491   \n",
       "\n",
       "  organismQuantityType sampleSizeValue      sampleSizeUnit  \n",
       "0   DNA sequence reads           85600  DNA sequence reads  \n",
       "1   DNA sequence reads           90702  DNA sequence reads  \n",
       "2   DNA sequence reads          130275  DNA sequence reads  \n",
       "3   DNA sequence reads          147220  DNA sequence reads  \n",
       "4   DNA sequence reads          121419  DNA sequence reads  \n",
       "\n",
       "[5 rows x 33 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## sampleSizeUnit\n",
    "\n",
    "occ['sampleSizeUnit'] = 'DNA sequence reads'\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>primer_sequence_forward</th>\n",
       "      <th>primer_sequence_reverse</th>\n",
       "      <th>...</th>\n",
       "      <th>nameAccordingTo</th>\n",
       "      <th>basisOfRecord</th>\n",
       "      <th>materialSampleID</th>\n",
       "      <th>identificationRemarks</th>\n",
       "      <th>identificationReferences</th>\n",
       "      <th>organismQuantity</th>\n",
       "      <th>organismQuantityType</th>\n",
       "      <th>sampleSizeValue</th>\n",
       "      <th>sampleSizeUnit</th>\n",
       "      <th>associatedSequences</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>insert ID for actual frozen sample if exists</td>\n",
       "      <td>Genbank nr September 20 2017;&gt;80% identity</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>14825</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>85600</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>biosampleID from SRA if available</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>insert ID for actual frozen sample if exists</td>\n",
       "      <td>Genbank nr September 20 2017;&gt;80% identity</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>16094</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>90702</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>biosampleID from SRA if available</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>insert ID for actual frozen sample if exists</td>\n",
       "      <td>Genbank nr September 20 2017;&gt;80% identity</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>22459</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>130275</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>biosampleID from SRA if available</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>insert ID for actual frozen sample if exists</td>\n",
       "      <td>Genbank nr September 20 2017;&gt;80% identity</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>19312</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>147220</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>biosampleID from SRA if available</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>insert ID for actual frozen sample if exists</td>\n",
       "      <td>Genbank nr September 20 2017;&gt;80% identity</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>16491</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>121419</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>biosampleID from SRA if available</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude                env_broad_scale  \\\n",
       "0          -121.848  marine biome  (ENVO:00000447)   \n",
       "1          -121.848  marine biome  (ENVO:00000447)   \n",
       "2          -121.848  marine biome  (ENVO:00000447)   \n",
       "3          -121.846  marine biome  (ENVO:00000447)   \n",
       "4          -121.846  marine biome  (ENVO:00000447)   \n",
       "\n",
       "                 env_local_scale  \\\n",
       "0  coastal water (ENVO:00001250)   \n",
       "1  coastal water (ENVO:00001250)   \n",
       "2  coastal water (ENVO:00001250)   \n",
       "3  coastal water (ENVO:00001250)   \n",
       "4  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  primer_sequence_forward   primer_sequence_reverse  ... nameAccordingTo  \\\n",
       "0        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...           WoRMS   \n",
       "1        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...           WoRMS   \n",
       "2        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...           WoRMS   \n",
       "3        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...           WoRMS   \n",
       "4        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...           WoRMS   \n",
       "\n",
       "    basisOfRecord                              materialSampleID  \\\n",
       "0  MaterialSample  insert ID for actual frozen sample if exists   \n",
       "1  MaterialSample  insert ID for actual frozen sample if exists   \n",
       "2  MaterialSample  insert ID for actual frozen sample if exists   \n",
       "3  MaterialSample  insert ID for actual frozen sample if exists   \n",
       "4  MaterialSample  insert ID for actual frozen sample if exists   \n",
       "\n",
       "                        identificationRemarks  \\\n",
       "0  Genbank nr September 20 2017;>80% identity   \n",
       "1  Genbank nr September 20 2017;>80% identity   \n",
       "2  Genbank nr September 20 2017;>80% identity   \n",
       "3  Genbank nr September 20 2017;>80% identity   \n",
       "4  Genbank nr September 20 2017;>80% identity   \n",
       "\n",
       "                            identificationReferences organismQuantity  \\\n",
       "0  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...            14825   \n",
       "1  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...            16094   \n",
       "2  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...            22459   \n",
       "3  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...            19312   \n",
       "4  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...            16491   \n",
       "\n",
       "  organismQuantityType sampleSizeValue      sampleSizeUnit  \\\n",
       "0   DNA sequence reads           85600  DNA sequence reads   \n",
       "1   DNA sequence reads           90702  DNA sequence reads   \n",
       "2   DNA sequence reads          130275  DNA sequence reads   \n",
       "3   DNA sequence reads          147220  DNA sequence reads   \n",
       "4   DNA sequence reads          121419  DNA sequence reads   \n",
       "\n",
       "                 associatedSequences  \n",
       "0  biosampleID from SRA if available  \n",
       "1  biosampleID from SRA if available  \n",
       "2  biosampleID from SRA if available  \n",
       "3  biosampleID from SRA if available  \n",
       "4  biosampleID from SRA if available  \n",
       "\n",
       "[5 rows x 34 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## associatedSequences\n",
    "\n",
    "occ['associatedSequences'] = 'biosampleID from SRA if available'\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Ask Katie about associatedSequences value.**\n",
    "\n",
    "## Tidy\n",
    "\n",
    "**Recall:** I'm leaving records where scientificName is missing in for now."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(64903, 34)\n"
     ]
    }
   ],
   "source": [
    "## Drop records where organismQuantity = 0 (absences are not meaningful for this data set)\n",
    "\n",
    "occ = occ[occ['organismQuantity'] > 0]\n",
    "print(occ.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Note** that the vast majority of this data has 0 reads (215537 records out of 280440, or about 75%). I guess that makes sense - it's sparse data - but it might be good to check?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "eventID                        0\n",
       "eventDate                      0\n",
       "decimalLatitude                0\n",
       "decimalLongitude               0\n",
       "env_broad_scale                0\n",
       "env_local_scale                0\n",
       "env_medium                     0\n",
       "target_gene                    0\n",
       "primer_sequence_forward        0\n",
       "primer_sequence_reverse        0\n",
       "pcr_primer_name_forward        0\n",
       "pcr_primer_name_reverse        0\n",
       "pcr_primer_reference           0\n",
       "sop                            0\n",
       "DNA_sequence                   0\n",
       "scientificName                 0\n",
       "kingdom                        0\n",
       "phylum                         0\n",
       "class                          0\n",
       "order                          0\n",
       "family                         0\n",
       "genus                          0\n",
       "scientificNameID               0\n",
       "taxonID                     5278\n",
       "nameAccordingTo                0\n",
       "basisOfRecord                  0\n",
       "materialSampleID               0\n",
       "identificationRemarks          0\n",
       "identificationReferences       0\n",
       "organismQuantity               0\n",
       "organismQuantityType           0\n",
       "sampleSizeValue                0\n",
       "sampleSizeUnit                 0\n",
       "associatedSequences            0\n",
       "dtype: int64"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Replace NaN values in text fields with ''\n",
    "\n",
    "text_fields = ['scientificName', 'scientificNameID']\n",
    "occ[text_fields] = occ[text_fields].replace(np.nan, '')\n",
    "occ.isna().sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Save"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Save\n",
    "\n",
    "occ.to_csv('eDNA_practice_plate_occ_20210308.csv', index=False, na_rep='NaN')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Questions\n",
    "\n",
    "1. For pcr_primer_reference, I think it would be best to either put a DOI that links to the article, or the entire article citation (i.e. title, journal name included). This isn't a requirement, it just seems like the safest way to ensure the correct article would be identified.\n",
    "2. Cleaning sop field. Why is it formatted like this? Do you just type it in? If so, maybe can adhere to DwC formatting requirements (pipes, no text)? If not, no worries, I can handle it.\n",
    "3. GenBank and WoRMS often disagree on taxonomy, especially when you look at the whole tree. Is this a problem? Also, talk through general scheme of getting a WoRMS match for scientificName. **Note that this process and assigning scientificName, taxonID, and taxonConceptID to organisms with no Linnaean name remains under debate, especially in the context of how it will be interpreted by OBIS search. Katie had some interesting thoughts about searching eDNA data, if I recall correctly?**\n",
    "4. Would putting 'Biota' be appropriate for taxa where nothing matched on GenBank at any taxonomic level? (http://marinespecies.org/aphia.php?p=taxdetails&id=1)\n",
    "5. Is it possible to obtain GenBankID for each taxon? I would like to include this information. \n",
    "6. **Question for Saara or Pieter or Abby:** Where can I include the fact that the taxonConceptID is a GenBankID?\n",
    "7. materialSampleID - are these water samples frozen somewhere? Do they have an ID number?\n",
    "8. What is in the identificationRemarks column? Can we make the wording more clear?\n",
    "9. associatedSequences - are there biosampleIDs from SRA that I can include?\n",
    "10. 75% of the rows in this dataset have 0 reads. Does that seem right?\n",
    "11. Follow up questions on control data: Has the control information been incorporated into this data set already? Are the control data still useful if you're dealing with the fully processed data set? Are the control data archived and therefore accessible via biosampleID?\n",
    "\n",
    "\n",
    "**Note to self**: Would it work just as well to drop the records with 0 reads up top? I don't know that it would change how long the WoRMS lookup takes (since I think the 0's appear because the ASV was present in another sample on the same plate), but it might speed up some simple processes like formatting dates and performing replacements. Maybe."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "venv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
