{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# eDNA conversion of practice data\n",
    "\n",
    "**Resources:**\n",
    "- https://docs.gbif-uat.org/publishing-sequence-derived-data/1.0/en/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Imports\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import random\n",
    "\n",
    "from datetime import datetime # for handling dates\n",
    "import pytz # for handling time zones"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Ensure my general functions for the MPA data integration project can be imported, and import them\n",
    "\n",
    "import sys\n",
    "sys.path.insert(0, \"C:\\\\Users\\\\dianalg\\\\PycharmProjects\\\\PythonScripts\\\\eDNA\")\n",
    "\n",
    "import WoRMS # functions for querying WoRMS REST API"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(280440, 11)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ASV</th>\n",
       "      <th>FilterID</th>\n",
       "      <th>Sequence_ID</th>\n",
       "      <th>Reads</th>\n",
       "      <th>Kingdom</th>\n",
       "      <th>Phylum</th>\n",
       "      <th>Class</th>\n",
       "      <th>Order</th>\n",
       "      <th>Family</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>05114c01_12_edna_1</td>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>14825</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>unassigned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>05114c01_12_edna_2</td>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>16094</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>unassigned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>05114c01_12_edna_3</td>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>22459</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>unassigned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>11216c01_12_edna_1</td>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>19312</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>unassigned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>11216c01_12_edna_2</td>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>16491</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>unassigned</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 ASV            FilterID  \\\n",
       "0  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  05114c01_12_edna_1   \n",
       "1  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  05114c01_12_edna_2   \n",
       "2  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  05114c01_12_edna_3   \n",
       "3  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  11216c01_12_edna_1   \n",
       "4  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  11216c01_12_edna_2   \n",
       "\n",
       "            Sequence_ID  Reads    Kingdom      Phylum        Class      Order  \\\n",
       "0  05114c01_12_edna_1_S  14825  Eukaryota  Arthropoda  Hexanauplia  Calanoida   \n",
       "1  05114c01_12_edna_2_S  16094  Eukaryota  Arthropoda  Hexanauplia  Calanoida   \n",
       "2  05114c01_12_edna_3_S  22459  Eukaryota  Arthropoda  Hexanauplia  Calanoida   \n",
       "3  11216c01_12_edna_1_S  19312  Eukaryota  Arthropoda  Hexanauplia  Calanoida   \n",
       "4  11216c01_12_edna_2_S  16491  Eukaryota  Arthropoda  Hexanauplia  Calanoida   \n",
       "\n",
       "          Family        Genus     Species  \n",
       "0  Paracalanidae  Paracalanus  unassigned  \n",
       "1  Paracalanidae  Paracalanus  unassigned  \n",
       "2  Paracalanidae  Paracalanus  unassigned  \n",
       "3  Paracalanidae  Paracalanus  unassigned  \n",
       "4  Paracalanidae  Paracalanus  unassigned  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Plate data\n",
    "\n",
    "plate = pd.read_csv('Plate_S_ASV_OBIS_data.csv')\n",
    "print(plate.shape)\n",
    "plate.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Note** that there are 60 unique FilterIDs and 60 unique Sequence_IDs in this table. Is that usual, to have # FilterID = # Sequence_ID? **IDs are a combination of what information?** Sequence_ID appears to be just the FilterID with underscore and S added."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(60, 2)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>FilterID</th>\n",
       "      <th>Sequence_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1</td>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2</td>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3</td>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1</td>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2</td>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             FilterID           Sequence_ID\n",
       "0  05114c01_12_edna_1  05114c01_12_edna_1_S\n",
       "1  05114c01_12_edna_2  05114c01_12_edna_2_S\n",
       "2  05114c01_12_edna_3  05114c01_12_edna_3_S\n",
       "3  11216c01_12_edna_1  11216c01_12_edna_1_S\n",
       "4  11216c01_12_edna_2  11216c01_12_edna_2_S"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Sequence_ID is just FilterID + '_S'\n",
    "\n",
    "temp = plate[['FilterID', 'Sequence_ID']].copy()\n",
    "temp.drop_duplicates(inplace=True)\n",
    "print(temp.shape)\n",
    "temp.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**FilterID seems to be composed of:**\n",
    "- SAMPLING_cruise from plate_meta\n",
    "- c + SAMPLING_station_number (or just SAMPLING_station, lower case and zero-padded?)\n",
    "- _ + SAMPLING_bottle\n",
    "- _ + edna\n",
    "- _ + 1, 2 or 3 (replicate?)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(4711, 79)\n",
      "Index(['ASV', 'Kingdom', 'Phylum', 'Class', 'Order', 'Family', 'Genus',\n",
      "       'Species', '14213c01_12_edna_1', '14213c01_12_edna_2',\n",
      "       '14213c01_12_edna_3', '22013c01_12_edna_1', '22013c01_12_edna_2',\n",
      "       '22013c01_12_edna_3', 'CN13Dc01_12_edna_1', 'CN13Dc01_12_edna_2',\n",
      "       'CN13Dc01_12_edna_3', '05114c01_12_edna_1', '05114c01_12_edna_2',\n",
      "       '05114c01_12_edna_3', '14714c01_12_edna_1', '14714c01_12_edna_2',\n",
      "       '14714c01_12_edna_3', '19114c01_12_edna_1', '19114c01_12_edna_2',\n",
      "       '19114c01_12_edna_3', '30214c01_12_edna_1', '30214c01_12_edna_2',\n",
      "       '30214c01_12_edna_3', '32414c01_12_edna_1', '32414c01_12_edna_2',\n",
      "       '32414c01_12_edna_3', '12015c01_12_edna_1', '12015c01_12_edna_2',\n",
      "       '12015c01_12_edna_3', '18815c01_12_edna_1', '18815c01_12_edna_2',\n",
      "       '18815c01_12_edna_3', 'EB_20161116', 'pcrblank_1', '28215c01_12_edna_1',\n",
      "       '28215c01_12_edna_2', '28215c01_12_edna_3', '34915c01_12_edna_1',\n",
      "       '34915c01_12_edna_2', '34915c01_12_edna_3', '11216c01_12_edna_1',\n",
      "       '11216c01_12_edna_2', '11216c01_12_edna_3', '24416c01_12_edna_1',\n",
      "       '24416c01_12_2', '24416c01_12_edna_3', 'CANON16c14_12_edna_1',\n",
      "       'CANON16c14_12_edna_2', 'CANON16c14_12_edna_3', '34916c01_12_edna_1',\n",
      "       '34916c01_12_edna_2', '34916c01_12_edna_3', 'JD10706C1_0m_1',\n",
      "       'JD10706C1_0m_2', 'JD10706C1_0m_3', 'JD15806C1_0m_1', 'JD15806C1_0m_2',\n",
      "       'JD15806C1_0m_3', 'JD27006C1_0m_1', 'JD27006C1_0m_2', 'JD27006C1_0m_3',\n",
      "       'JD33306C1_0m_1', 'JD33306C1_0m_2', 'JD33306C1_0m_3', 'pcrblank2',\n",
      "       'EB_20161121', 'EB_20161228', 'EB_20170117', 'pcrblank3',\n",
      "       'CB_CANON160925_1', 'CB_CANON160925_2', 'CB_CANON160925_3', 'ArtComm2'],\n",
      "      dtype='object')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ASV</th>\n",
       "      <th>Kingdom</th>\n",
       "      <th>Phylum</th>\n",
       "      <th>Class</th>\n",
       "      <th>Order</th>\n",
       "      <th>Family</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Species</th>\n",
       "      <th>14213c01_12_edna_1</th>\n",
       "      <th>14213c01_12_edna_2</th>\n",
       "      <th>...</th>\n",
       "      <th>JD33306C1_0m_3</th>\n",
       "      <th>pcrblank2</th>\n",
       "      <th>EB_20161121</th>\n",
       "      <th>EB_20161228</th>\n",
       "      <th>EB_20170117</th>\n",
       "      <th>pcrblank3</th>\n",
       "      <th>CB_CANON160925_1</th>\n",
       "      <th>CB_CANON160925_2</th>\n",
       "      <th>CB_CANON160925_3</th>\n",
       "      <th>ArtComm2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ASV_1</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>9</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>956</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1312</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>197</td>\n",
       "      <td>3</td>\n",
       "      <td>210</td>\n",
       "      <td>44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ASV_2</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Calanidae</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>4</td>\n",
       "      <td>7</td>\n",
       "      <td>...</td>\n",
       "      <td>358</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>535</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>118</td>\n",
       "      <td>162</td>\n",
       "      <td>139</td>\n",
       "      <td>5976</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ASV_3</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>Gymnodiniales</td>\n",
       "      <td>Gymnodiniaceae</td>\n",
       "      <td>Akashiwo</td>\n",
       "      <td>Akashiwo sanguinea</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>795</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>35</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ASV_4</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>Gymnodiniales</td>\n",
       "      <td>Gymnodiniaceae</td>\n",
       "      <td>Cochlodinium</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>82065</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>836</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ASV_5</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>148</td>\n",
       "      <td>147</td>\n",
       "      <td>...</td>\n",
       "      <td>14688</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>36</td>\n",
       "      <td>401</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>11</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 79 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     ASV    Kingdom      Phylum        Class          Order          Family  \\\n",
       "0  ASV_1  Eukaryota  Arthropoda  Hexanauplia      Calanoida   Paracalanidae   \n",
       "1  ASV_2  Eukaryota  Arthropoda  Hexanauplia      Calanoida       Calanidae   \n",
       "2  ASV_3  Eukaryota     unknown  Dinophyceae  Gymnodiniales  Gymnodiniaceae   \n",
       "3  ASV_4  Eukaryota     unknown  Dinophyceae  Gymnodiniales  Gymnodiniaceae   \n",
       "4  ASV_5  Eukaryota     unknown  Dinophyceae     unassigned      unassigned   \n",
       "\n",
       "          Genus             Species  14213c01_12_edna_1  14213c01_12_edna_2  \\\n",
       "0   Paracalanus          unassigned                   9                   4   \n",
       "1    unassigned          unassigned                   4                   7   \n",
       "2      Akashiwo  Akashiwo sanguinea                   2                   3   \n",
       "3  Cochlodinium          unassigned                   2                   3   \n",
       "4    unassigned          unassigned                 148                 147   \n",
       "\n",
       "   ...  JD33306C1_0m_3  pcrblank2  EB_20161121  EB_20161228  EB_20170117  \\\n",
       "0  ...             956          0            0         1312            1   \n",
       "1  ...             358          0            0          535            0   \n",
       "2  ...             795          0            0            0           35   \n",
       "3  ...           82065          0            0            0          836   \n",
       "4  ...           14688          0            0           36          401   \n",
       "\n",
       "   pcrblank3  CB_CANON160925_1  CB_CANON160925_2  CB_CANON160925_3  ArtComm2  \n",
       "0          0               197                 3               210        44  \n",
       "1          0               118               162               139      5976  \n",
       "2          0                 1                 0                 1         8  \n",
       "3          0                 1                 0                 0         7  \n",
       "4          0                 0                 4                11         7  \n",
       "\n",
       "[5 rows x 79 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## ASV taxa table\n",
    "\n",
    "taxa = pd.read_csv('Filtered_ASV_taxa_table_all.csv')\n",
    "print(taxa.shape)\n",
    "print(taxa.columns)\n",
    "taxa.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It looks like this shows the number of reads for each ASV detected for each FilterID (columns), plus the matched taxa information for each ASV, if available. Blank cells have been filled with the word 'unassigned'. \n",
    "\n",
    "There are some column names that do not correspond to FilterIDs. They are the taxonomy columns (of course), in addition to the following:\n",
    "- EB_20161116\n",
    "- pcrblank_1\n",
    "- pcrblank2\n",
    "- EB_20161121\n",
    "- EB_20161228\n",
    "- EB_20170117\n",
    "- pcrblank3\n",
    "- CB_CANON160925_1\n",
    "- CB_CANON160925_2\n",
    "- CB_CANON160925_3\n",
    "- ArtComm2\n",
    "\n",
    "Are some of these controls? \n",
    "\n",
    "**Yes - here are the possible control types:**\n",
    "- CB = collection blank which was taken on the ship. Water that should be \"clean\" was passed through the filter, so there should be no reads in this sample, although it tends to be the \"dirtiest\" control. It checks for contamination during filtration, I think.\n",
    "- EB = extraction blank which was taken in the lab but contains no DNA (i.e. a negative control), so there should be no reads in this sample. It checks if contamination occurred during DNA extraction (lab work pre-PCR).\n",
    "- pcr blank = A sample that went through PCR but contained no input DNA (i.e. a negative control), so there should be no reads in this sample. It checks if contamination occurred during PCR.\n",
    "- ArtComm = artificial community that went through PCR and contained DNA from species that should not be in the study system (i.e. a positive control), so you know what results you expect.\n",
    "\n",
    "**Note** that there is some bioinformatic post-processing you can do to adjust for things that popped up in your controls.\n",
    "\n",
    "```python\n",
    "for col in taxa.columns:\n",
    "    if col not in plate['FilterID'].unique():\n",
    "        print(col)\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(60, 52)\n",
      "Index(['SequenceID', 'sample_name', 'order', 'tag_sequence',\n",
      "       'primer_sequence_F', 'primer_sequence_R', 'library_tag_combo',\n",
      "       'library', 'date_PCR', 'sample_type', 'locus', 'tag_number', 'R1', 'R2',\n",
      "       'SAMPLING_cruise', 'SAMPLING_station_number', 'SAMPLING_bottle',\n",
      "       'depth', 'SAMPLING_station', 'SAMPLING_project', 'SAMPLING_platform',\n",
      "       'SAMPLING_platform_type', 'SAMPLING_dec_lat', 'SAMPLING_dec_lon',\n",
      "       'temp', 'salinity', 'chlorophyll', 'pressure_dbar', 'nitrate',\n",
      "       'diss_oxygen', 'SAMPLING_real_depth', 'SAMPLING_transmiss_%',\n",
      "       'SAMPLING_sig_t', 'SAMPLING_fluor', 'SAMPLING_date_time', 'description',\n",
      "       'SAMPLING_PI', 'SAMPLING_institute', 'env_biome', 'env_feature',\n",
      "       'env_material', 'samp_collection_device', 'project_name',\n",
      "       'samp_vol_we_dna_ext', 'samp_filter_size_ext', 'samp_filter_ext_type',\n",
      "       'samp_store_temp', 'seq_meth', 'sequencing_facility', 'geo_loc_name',\n",
      "       'investigation_type', 'FilterID'],\n",
      "      dtype='object')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SequenceID</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>order</th>\n",
       "      <th>tag_sequence</th>\n",
       "      <th>primer_sequence_F</th>\n",
       "      <th>primer_sequence_R</th>\n",
       "      <th>library_tag_combo</th>\n",
       "      <th>library</th>\n",
       "      <th>date_PCR</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>...</th>\n",
       "      <th>project_name</th>\n",
       "      <th>samp_vol_we_dna_ext</th>\n",
       "      <th>samp_filter_size_ext</th>\n",
       "      <th>samp_filter_ext_type</th>\n",
       "      <th>samp_store_temp</th>\n",
       "      <th>seq_meth</th>\n",
       "      <th>sequencing_facility</th>\n",
       "      <th>geo_loc_name</th>\n",
       "      <th>investigation_type</th>\n",
       "      <th>FilterID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>14213c01_12_edna_1_S</td>\n",
       "      <td>14213c01_12_edna_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ACGAGACTGATT</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>S1_ACGAGACTGATT</td>\n",
       "      <td>S1</td>\n",
       "      <td>20170124</td>\n",
       "      <td>environmental</td>\n",
       "      <td>...</td>\n",
       "      <td>MBON</td>\n",
       "      <td>100ml</td>\n",
       "      <td>0.22um</td>\n",
       "      <td>Poretics</td>\n",
       "      <td>-80C</td>\n",
       "      <td>NGS Illumina Miseq</td>\n",
       "      <td>Stanford</td>\n",
       "      <td>USA:California:Monterey Bay</td>\n",
       "      <td>eukaryote</td>\n",
       "      <td>14213c01_12_edna_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>14213c01_12_edna_2_S</td>\n",
       "      <td>14213c01_12_edna_2</td>\n",
       "      <td>2</td>\n",
       "      <td>GAATACCAAGTC</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>S2_GAATACCAAGTC</td>\n",
       "      <td>S2</td>\n",
       "      <td>20170124</td>\n",
       "      <td>environmental</td>\n",
       "      <td>...</td>\n",
       "      <td>MBON</td>\n",
       "      <td>100ml</td>\n",
       "      <td>0.22um</td>\n",
       "      <td>Poretics</td>\n",
       "      <td>-80C</td>\n",
       "      <td>NGS Illumina Miseq</td>\n",
       "      <td>Stanford</td>\n",
       "      <td>USA:California:Monterey Bay</td>\n",
       "      <td>eukaryote</td>\n",
       "      <td>14213c01_12_edna_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>14213c01_12_edna_3_S</td>\n",
       "      <td>14213c01_12_edna_3</td>\n",
       "      <td>3</td>\n",
       "      <td>CGAGGGAAAGTC</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>S3_CGAGGGAAAGTC</td>\n",
       "      <td>S3</td>\n",
       "      <td>20170124</td>\n",
       "      <td>environmental</td>\n",
       "      <td>...</td>\n",
       "      <td>MBON</td>\n",
       "      <td>100ml</td>\n",
       "      <td>0.22um</td>\n",
       "      <td>Poretics</td>\n",
       "      <td>-80C</td>\n",
       "      <td>NGS Illumina Miseq</td>\n",
       "      <td>Stanford</td>\n",
       "      <td>USA:California:Monterey Bay</td>\n",
       "      <td>eukaryote</td>\n",
       "      <td>14213c01_12_edna_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>22013c01_12_edna_1_S</td>\n",
       "      <td>22013c01_12_edna_1</td>\n",
       "      <td>4</td>\n",
       "      <td>GAACACTTTGGA</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>S4_GAACACTTTGGA</td>\n",
       "      <td>S4</td>\n",
       "      <td>20170124</td>\n",
       "      <td>environmental</td>\n",
       "      <td>...</td>\n",
       "      <td>MBON</td>\n",
       "      <td>100ml</td>\n",
       "      <td>0.22um</td>\n",
       "      <td>Poretics</td>\n",
       "      <td>-80C</td>\n",
       "      <td>NGS Illumina Miseq</td>\n",
       "      <td>Stanford</td>\n",
       "      <td>USA:California:Monterey Bay</td>\n",
       "      <td>eukaryote</td>\n",
       "      <td>22013c01_12_edna_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>22013c01_12_edna_2_S</td>\n",
       "      <td>22013c01_12_edna_2</td>\n",
       "      <td>5</td>\n",
       "      <td>ACTCACAGGAAT</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>S5_ACTCACAGGAAT</td>\n",
       "      <td>S5</td>\n",
       "      <td>20170124</td>\n",
       "      <td>environmental</td>\n",
       "      <td>...</td>\n",
       "      <td>MBON</td>\n",
       "      <td>100ml</td>\n",
       "      <td>0.22um</td>\n",
       "      <td>Poretics</td>\n",
       "      <td>-80C</td>\n",
       "      <td>NGS Illumina Miseq</td>\n",
       "      <td>Stanford</td>\n",
       "      <td>USA:California:Monterey Bay</td>\n",
       "      <td>eukaryote</td>\n",
       "      <td>22013c01_12_edna_2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 52 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             SequenceID         sample_name  order  tag_sequence  \\\n",
       "0  14213c01_12_edna_1_S  14213c01_12_edna_1      1  ACGAGACTGATT   \n",
       "1  14213c01_12_edna_2_S  14213c01_12_edna_2      2  GAATACCAAGTC   \n",
       "2  14213c01_12_edna_3_S  14213c01_12_edna_3      3  CGAGGGAAAGTC   \n",
       "3  22013c01_12_edna_1_S  22013c01_12_edna_1      4  GAACACTTTGGA   \n",
       "4  22013c01_12_edna_2_S  22013c01_12_edna_2      5  ACTCACAGGAAT   \n",
       "\n",
       "  primer_sequence_F         primer_sequence_R library_tag_combo library  \\\n",
       "0  GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC   S1_ACGAGACTGATT      S1   \n",
       "1  GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC   S2_GAATACCAAGTC      S2   \n",
       "2  GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC   S3_CGAGGGAAAGTC      S3   \n",
       "3  GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC   S4_GAACACTTTGGA      S4   \n",
       "4  GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC   S5_ACTCACAGGAAT      S5   \n",
       "\n",
       "   date_PCR    sample_type  ... project_name  samp_vol_we_dna_ext  \\\n",
       "0  20170124  environmental  ...         MBON                100ml   \n",
       "1  20170124  environmental  ...         MBON                100ml   \n",
       "2  20170124  environmental  ...         MBON                100ml   \n",
       "3  20170124  environmental  ...         MBON                100ml   \n",
       "4  20170124  environmental  ...         MBON                100ml   \n",
       "\n",
       "  samp_filter_size_ext samp_filter_ext_type samp_store_temp  \\\n",
       "0               0.22um             Poretics            -80C   \n",
       "1               0.22um             Poretics            -80C   \n",
       "2               0.22um             Poretics            -80C   \n",
       "3               0.22um             Poretics            -80C   \n",
       "4               0.22um             Poretics            -80C   \n",
       "\n",
       "             seq_meth  sequencing_facility                 geo_loc_name  \\\n",
       "0  NGS Illumina Miseq             Stanford  USA:California:Monterey Bay   \n",
       "1  NGS Illumina Miseq             Stanford  USA:California:Monterey Bay   \n",
       "2  NGS Illumina Miseq             Stanford  USA:California:Monterey Bay   \n",
       "3  NGS Illumina Miseq             Stanford  USA:California:Monterey Bay   \n",
       "4  NGS Illumina Miseq             Stanford  USA:California:Monterey Bay   \n",
       "\n",
       "  investigation_type            FilterID  \n",
       "0          eukaryote  14213c01_12_edna_1  \n",
       "1          eukaryote  14213c01_12_edna_2  \n",
       "2          eukaryote  14213c01_12_edna_3  \n",
       "3          eukaryote  22013c01_12_edna_1  \n",
       "4          eukaryote  22013c01_12_edna_2  \n",
       "\n",
       "[5 rows x 52 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Plate metadata\n",
    "\n",
    "plate_meta = pd.read_csv('Plate_S_meta_OBIS_data.csv')\n",
    "print(plate_meta.shape)\n",
    "print(plate_meta.columns)\n",
    "plate_meta.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This shows some metadata associated with each SequenceID and FilterID. This metadata **seems mostly redundant with that in MB_20200317_1326_18S_analysis_metadata.csv**, so for now I'll just work with it.\n",
    "\n",
    "## Conversion Plan\n",
    "\n",
    "According to GBIF's new guide to publishing sequence-derived biodiversity data, my occurrence file should include the following:\n",
    "\n",
    "<span style=\"color:blue\">eventID</span> - Highly recommended, **SequenceID in plate and plate_meta** <br>\n",
    "<span style=\"color:blue\">eventDate</span> - Required, **SAMPLING_date_time in plate_meta** <br>\n",
    "<span style=\"color:blue\">decimalLatitude</span> - Highly recommended, **SAMPLING_dec_lat in plate_meta** <br>\n",
    "<span style=\"color:blue\">decimalLongitude</span> - Highly recommended, **SAMPLING_dec_lon in plate_meta**  <br>\n",
    "<span style=\"color:blue\">env_broad_scale</span> - Recommended, equivalent to env_biome in MIxS, the major environmental system your sample or specimen came from, use [subclasses of ENVO´s biome class](http://www.ontobee.org/ontology/ENVO?iri=http://purl.obolibrary.org/obo/ENVO_00000428), **env_biome in plate_meta** <br>\n",
    "<span style=\"color:blue\">env_local_scale</span> - Recommended, equivalent to env_feature in MIxS, the entity or entities which are in your sample or specimen´s local vicinity and which you believe have significant causal influences on your sample or specimen, use terms that are present in ENVO and which are of smaller spatial grain than your entry for env_broad_scale, **env_feature in plate_meta** <br>\n",
    "<span style=\"color:blue\">env_medium</span> - Recommended, equivalent to env_material in MIxS, environmental material that immediately surrounded your sample or specimen prior to sampling, use [subclasses of ENVO´s environmental material class](http://www.ontobee.org/ontology/ENVO?iri=http://purl.obolibrary.org/obo/ENVO_00010483), **env_material in plate_meta**<br>\n",
    "<span style=\"color:blue\">sop</span> - Recommended, standard operating procedures used in assembly and/or annotation of metagenomes or a reference to a well documented protocol (e.g. using protocols.io). **Does something like this exist? If not, what information might be important to include here?** <br>\n",
    "<span style=\"color:blue\">lib_layout</span> - Recommended, equivalent to lib_const_meth in MIxS, whether to expect single, paired, or other configuration of reads. **Is this relevant here?** <br>\n",
    "<span style=\"color:blue\">target_gene</span> - Highly recommended, targeted gene or marker name for marker-based studies (e.g. 16S rRNA), **locus? Does there need to be more info included here?** <br>\n",
    "<span style=\"color:blue\">target_subfragment</span> - Highly recommended, name of subfragment of a gene or marker (e.g. V6). **Is this relevant here? Should the tag_sequence fit in somewhere?** <br>\n",
    "<span style=\"color:blue\">pcr_primer_name_forward</span> - Highly recommended, name of forward primer. **Do these primers have names?** <br>\n",
    "<span style=\"color:blue\">pcr_primer_forward</span> - Highly recommended, sequence of the forward primer, **primer_sequence_F in plate_meta** <br>\n",
    "<span style=\"color:blue\">pcr_primer_name_reverse</span> - Highly recommended, name of reverse primer <br>\n",
    "<span style=\"color:blue\">pcr_primer_revers</span> - Highly recommended, sequence of the reverse primer, **primer_sequence_R in plate_meta** <br>\n",
    "<span style=\"color:blue\">pcr_primer_reference</span> - Highly recommended, reference for primers (e.g. a DOI to a paper). **Is there a reference for primers?** <br>\n",
    "<span style=\"color:blue\">DNA_sequence</span> - Highly recommended, the actual DNA sequence of the ASV. TaxonID is highly recommended if DNA_sequence is not provided, **ASV in plate** <br>\n",
    "<span style=\"color:blue\">scientificName</span> - Required, Latin name of the closest known taxon or an OTU identifier from BOLD or UNITE, **a combination of Genus and Species from plate, or the lowest available taxon. This will have to be looked into more, especially for cases where no traditional taxonomic classification is available.** <br>\n",
    "<span style=\"color:blue\">kingdom</span> - Highly recommended, **Kingdom in plate** <br>\n",
    "<span style=\"color:blue\">phylum</span> - Recommended, **Phylum in plate** <br>\n",
    "<span style=\"color:blue\">class</span> - Recommended, **Class in plate** <br>\n",
    "<span style=\"color:blue\">order</span> - Recommended, **Order in plate** <br>\n",
    "<span style=\"color:blue\">family</span> - Recommended, **Family in plate** <br>\n",
    "<span style=\"color:blue\">genus</span> - Recommended, **Genus in plate** <br>\n",
    "<span style=\"color:blue\">basisOfRecord</span> - Required, MaterialSample <br>\n",
    "<span style=\"color:blue\">materialSampleID</span> - Highly recommended, an identifier for the MaterialSample, use the biosample ID if one was obtained from a nucleotide archive otherwise construct a globally unique identifier. **Is this Sequence_ID again? FilterID? Something else?** <br>\n",
    "<span style=\"color:blue\">identificationRemarks</span> - recommended, specification of taxonomic identification process ideally including data on applied algorithm and reference database as well as on level of confidence in the resulting identification. **Is this information available somewhere?** <br>\n",
    "<span style=\"color:blue\">identificationReferences</span> - recommended, link to protocol or code. **Is this information available somewhere?** <br>\n",
    "<span style=\"color:blue\">organismQuantity</span> - Highly recommended, number of reads, **Reads in plate** <br>\n",
    "<span style=\"color:blue\">organismQuantityType</span> - Highly recommended, DNA sequence reads <br>\n",
    "<span style=\"color:blue\">sampleSizeValue</span> - Highly recommended, total number of reads in the sample for calculating the relative abundance of sequence variants. **Is it accurate to just sum all the reads by Sequence_ID?** <br>\n",
    "<span style=\"color:blue\">sampleSizeUnit</span> - Highly recommended, DNA sequence reads <br>\n",
    "<span style=\"color:blue\">associatedSequences</span> - recommended, list of identifiers linking to archived (raw) sequence reads. **Are these sequences already archived?** <br>\n",
    "\n",
    "## Conversion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(280440, 1)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID\n",
       "0  05114c01_12_edna_1_S\n",
       "1  05114c01_12_edna_2_S\n",
       "2  05114c01_12_edna_3_S\n",
       "3  11216c01_12_edna_1_S\n",
       "4  11216c01_12_edna_2_S"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## eventID\n",
    "\n",
    "occ = pd.DataFrame({'eventID':plate['Sequence_ID']})\n",
    "print(occ.shape)\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>pcr_primer_forward</th>\n",
       "      <th>pcr_primer_reverse</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20 15:33</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20 15:33</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20 15:33</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21 14:39</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21 14:39</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID         eventDate  decimalLatitude  decimalLongitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20 15:33          36.7958          -121.848   \n",
       "1  05114c01_12_edna_2_S  2014-02-20 15:33          36.7958          -121.848   \n",
       "2  05114c01_12_edna_3_S  2014-02-20 15:33          36.7958          -121.848   \n",
       "3  11216c01_12_edna_1_S  2016-04-21 14:39          36.7962          -121.846   \n",
       "4  11216c01_12_edna_2_S  2016-04-21 14:39          36.7962          -121.846   \n",
       "\n",
       "                 env_broad_scale                env_local_scale  \\\n",
       "0  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   \n",
       "1  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   \n",
       "2  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   \n",
       "3  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   \n",
       "4  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  pcr_primer_forward        pcr_primer_reverse  \n",
       "0   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  \n",
       "1   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  \n",
       "2   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  \n",
       "3   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  \n",
       "4   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Merge with plate_meta to obtain eventDate, decimal Lat and Lon\n",
    "\n",
    "occ = occ.merge(plate_meta[['SequenceID', 'SAMPLING_date_time', 'SAMPLING_dec_lat', 'SAMPLING_dec_lon',\n",
    "                            'env_biome', 'env_feature', 'env_material', 'locus', 'primer_sequence_F',\n",
    "                            'primer_sequence_R']], how='left', left_on='eventID', right_on='SequenceID')\n",
    "occ.drop(columns='SequenceID', inplace=True)\n",
    "occ.columns = ['eventID', 'eventDate', 'decimalLatitude', 'decimalLongitude', 'env_broad_scale', 'env_local_scale', 'env_medium',\n",
    "               'target_gene', 'pcr_primer_forward', 'pcr_primer_reverse']\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Need to add sop? lib_layout? target_subfragment? primer names? primer references?**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>pcr_primer_forward</th>\n",
       "      <th>pcr_primer_reverse</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude                env_broad_scale  \\\n",
       "0          -121.848  marine biome  (ENVO:00000447)   \n",
       "1          -121.848  marine biome  (ENVO:00000447)   \n",
       "2          -121.848  marine biome  (ENVO:00000447)   \n",
       "3          -121.846  marine biome  (ENVO:00000447)   \n",
       "4          -121.846  marine biome  (ENVO:00000447)   \n",
       "\n",
       "                 env_local_scale  \\\n",
       "0  coastal water (ENVO:00001250)   \n",
       "1  coastal water (ENVO:00001250)   \n",
       "2  coastal water (ENVO:00001250)   \n",
       "3  coastal water (ENVO:00001250)   \n",
       "4  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  pcr_primer_forward        pcr_primer_reverse  \n",
       "0   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  \n",
       "1   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  \n",
       "2   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  \n",
       "3   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  \n",
       "4   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  "
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Format eventDate\n",
    "\n",
    "pst = pytz.timezone('America/Los_Angeles')\n",
    "eventDate = [pst.localize(datetime.strptime(dt, '%Y-%m-%d %H:%M')).isoformat() for dt in occ['eventDate']]\n",
    "occ['eventDate'] = eventDate\n",
    "\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>pcr_primer_forward</th>\n",
       "      <th>pcr_primer_reverse</th>\n",
       "      <th>DNA_sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude                env_broad_scale  \\\n",
       "0          -121.848  marine biome  (ENVO:00000447)   \n",
       "1          -121.848  marine biome  (ENVO:00000447)   \n",
       "2          -121.848  marine biome  (ENVO:00000447)   \n",
       "3          -121.846  marine biome  (ENVO:00000447)   \n",
       "4          -121.846  marine biome  (ENVO:00000447)   \n",
       "\n",
       "                 env_local_scale  \\\n",
       "0  coastal water (ENVO:00001250)   \n",
       "1  coastal water (ENVO:00001250)   \n",
       "2  coastal water (ENVO:00001250)   \n",
       "3  coastal water (ENVO:00001250)   \n",
       "4  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  pcr_primer_forward        pcr_primer_reverse  \\\n",
       "0   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC   \n",
       "1   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC   \n",
       "2   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC   \n",
       "3   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC   \n",
       "4   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC   \n",
       "\n",
       "                                        DNA_sequence  \n",
       "0  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  \n",
       "1  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  \n",
       "2  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  \n",
       "3  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  \n",
       "4  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  "
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Add DNA_sequence\n",
    "\n",
    "occ['DNA_sequence'] = plate['ASV']\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Before adding taxonomic information, I'll have to clean up the Genus and Species columns in the plate df. **Weird values include:**\n",
    "- 'unassigned' has been used when there's no Genus data; also when there's no Species data\n",
    "- 'g_', **I assume** also signifies no Genus; 's_' seems to be the equivalent in Species\n",
    "- 'unknown' also has been used\n",
    "- 'no_hit' in both Genus and Species\n",
    "- 'Herdmania &lt;dinoflagellates>' - The Species entry for this seems to give both Genus and Species (Herdmania litoralis)\n",
    "    - Side note: **How can a Phylum be unknown, but then Class, Order, Family, etc. be known?**\n",
    "- 'Halofilum &lt;green algae>' - Similarly, Species is Halofilum ramosum\n",
    "- 'Candida &lt;clade Candida/Lodderomyces clade>' - This has Species unassigned, Family = Debaryomycetaceae\n",
    "\n",
    "Only in Species:\n",
    "- 'uncultured marine eukaryote'\n",
    "- 'eukaryote clone OLI11007'\n",
    "- 'Dinophyceae sp. UDMS0803'\n",
    "- 'uncultured marine picoeukaryote'\n",
    "- 'Chaetoceros sp. UNC1415'\n",
    "- 'bacterium'\n",
    "\n",
    "**Note** there are loads of terms like those given above in the Species column. **What do they mean and where do they come from?** Maybe some of those numbers correspond to BOLD or UNITE OTU ID's. **Also note** that the Species column often contains both Genus and Species - the proper scientific name - if available. I'm not sure if this always is the case, but maybe I can just use the Species column for scientificName rather than combining Genus and Species."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>pcr_primer_forward</th>\n",
       "      <th>pcr_primer_reverse</th>\n",
       "      <th>DNA_sequence</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude                env_broad_scale  \\\n",
       "0          -121.848  marine biome  (ENVO:00000447)   \n",
       "1          -121.848  marine biome  (ENVO:00000447)   \n",
       "2          -121.848  marine biome  (ENVO:00000447)   \n",
       "3          -121.846  marine biome  (ENVO:00000447)   \n",
       "4          -121.846  marine biome  (ENVO:00000447)   \n",
       "\n",
       "                 env_local_scale  \\\n",
       "0  coastal water (ENVO:00001250)   \n",
       "1  coastal water (ENVO:00001250)   \n",
       "2  coastal water (ENVO:00001250)   \n",
       "3  coastal water (ENVO:00001250)   \n",
       "4  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  pcr_primer_forward        pcr_primer_reverse  \\\n",
       "0   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC   \n",
       "1   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC   \n",
       "2   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC   \n",
       "3   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC   \n",
       "4   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC   \n",
       "\n",
       "                                        DNA_sequence scientificName  \\\n",
       "0  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...     unassigned   \n",
       "1  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...     unassigned   \n",
       "2  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...     unassigned   \n",
       "3  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...     unassigned   \n",
       "4  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...     unassigned   \n",
       "\n",
       "     kingdom      phylum        class      order         family        genus  \n",
       "0  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus  \n",
       "1  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus  \n",
       "2  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus  \n",
       "3  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus  \n",
       "4  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## scientificName, taxonomic info\n",
    "\n",
    "occ['scientificName'] = plate['Species']\n",
    "occ['kingdom'] = plate['Kingdom']\n",
    "occ['phylum'] = plate['Phylum']\n",
    "occ['class'] = plate['Class']\n",
    "occ['order'] = plate['Order']\n",
    "occ['family'] = plate['Family']\n",
    "occ['genus'] = plate['Genus']\n",
    "\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**So we don't have to do the WoRMS thing for eDNA data?**\n",
    "\n",
    "**Note** that:\n",
    "- 63600 (out of a total 280440 records) have scientificName = 'unassigned'\n",
    "- 163380 have scientificName = 's_'\n",
    "- 0 have scientificName = 'unknown'\n",
    "- 12540 have scientificName = 'no_hit'\n",
    "\n",
    "It seems like for the unassigned and s_ species, other taxonomic levels of classification (e.g. Family) are known (although I haven't checked this in all cases). When the species is 'no_hit', though, it seems like it's 'no_hit' across the board.\n",
    "\n",
    "**Let me replace these values with NaNs to make things easier to work with.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Clean scientificName and taxonomy columns\n",
    "\n",
    "cols = ['scientificName', 'kingdom', 'phylum', 'class', 'order', 'family', 'genus']\n",
    "occ[cols] = occ[cols].replace({'unassigned':np.nan,\n",
    "                              's_':np.nan,\n",
    "                              'g_':np.nan,\n",
    "                              'unknown':np.nan,\n",
    "                              'no_hit':np.nan})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Ok, so now we can see that:\n",
    "- 9300 records have full taxonomic info (all columns are not NaN)\n",
    "- A total of 239520 do not have a species designation (i.e. scientificName = NaN). This is as expected based on the above numbers.\n",
    "- Of these, 9840 are only missing species\n",
    "- 120 are only missing genus\n",
    "- 300 are only missing family\n",
    "- 1020 are only missing order\n",
    "- 780 are only missing class\n",
    "- 10860 are only missing phylum\n",
    "- 0 are only missing kingdom\n",
    "- This leaves 216600 records with more than one field missing\n",
    "\n",
    "**How do I handle these? Based on the standard information, it seems like I should fill NaNs in the species column with the lowest known taxonomic rank. If all the rows are NaN, the record cannot be submitted to OBIS. *But*, before deleting these, I need to look into the non-Linnaean name options.**\n",
    "\n",
    "**Note** that 33360 rows have ALL NaNs in ALL taxonomic columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Fill missing values in the species column with the lowest available taxonomic rank\n",
    "\n",
    "occ['scientificName'] = occ['scientificName'].combine_first(occ['genus'])\n",
    "occ['scientificName'] = occ['scientificName'].combine_first(occ['family'])\n",
    "occ['scientificName'] = occ['scientificName'].combine_first(occ['order'])\n",
    "occ['scientificName'] = occ['scientificName'].combine_first(occ['class'])\n",
    "occ['scientificName'] = occ['scientificName'].combine_first(occ['phylum'])\n",
    "occ['scientificName'] = occ['scientificName'].combine_first(occ['kingdom'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Ok, so now, **what do the weird species designations mean?**\n",
    "\n",
    "**[Barcode of Life Barcode Index Numbers (BINs)](http://www.boldsystems.org/index.php)** are of the format: BOLD:AAA0000 (where A can be any letter A-Z, and 0 can be any number 0-9). These are based on CO1 sequences (animals), rRNA ITS sequences (fungi), or rbcL and matK sequences (plants). <br>\n",
    "**[UNITE Species Hypotheses (SHs)](https://unite.ut.ee/index.php)** are of the format: SH1566366.08FU. These are based on the rRNA ITS region.\n",
    "\n",
    "None of the species designations seem to match either of these formats.\n",
    "\n",
    "**Can we search for BINs and/or SHs using the DNA_sequence (and can I automate this process using an API)? Or do we have to throw out anything that's not a valid Linnean name? To check for validity, would I still need to run things through WoRMS?**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Url didn't work, check name:  Eukaryota\n",
      "Url didn't work for uncultured marine eukaryote checking:  uncultured\n",
      "Url didn't work, check name:  uncultured\n",
      "Url didn't work for eukaryote clone OLI11007 checking:  eukaryote\n",
      "Url didn't work, check name:  eukaryote\n",
      "Url didn't work for Dinophyceae sp. UDMS0803 checking:  Dinophyceae\n",
      "Url didn't work, check name:  Pelargonium\n",
      "Url didn't work, check name:  Hemistasiidae\n",
      "Url didn't work for Polykrikos geminatum checking:  Polykrikos\n",
      "Url didn't work for uncultured marine picoeukaryote checking:  uncultured\n",
      "Url didn't work, check name:  uncultured\n",
      "Url didn't work for Strombidium caudispina checking:  Strombidium\n",
      "Url didn't work for Halodinium verrucatum checking:  Halodinium\n",
      "Url didn't work for Chaetoceros sp. UNC1415 checking:  Chaetoceros\n",
      "Url didn't work for Euduboscquella cachoni checking:  Euduboscquella\n",
      "Url didn't work, check name:  Bicosoecidae\n",
      "Url didn't work for Takayama cf. pulchellum checking:  Takayama\n",
      "Url didn't work, check name:  Bryophryidae\n",
      "Url didn't work for eukaryote clone OLI11072 checking:  eukaryote\n",
      "Url didn't work, check name:  eukaryote\n",
      "Url didn't work, check name:  Rhogostomidae\n",
      "Url didn't work for phototrophic eukaryote checking:  phototrophic\n",
      "Url didn't work, check name:  phototrophic\n",
      "Url didn't work, check name:  Crustomastigaceae\n",
      "Url didn't work for Bicosoeca vacillans checking:  Bicosoeca\n",
      "Url didn't work for Acantharian sp. 6201 checking:  Acantharian\n",
      "Url didn't work, check name:  Acantharian\n",
      "Url didn't work for Pelagophyceae sp. CCMP3390 checking:  Pelagophyceae\n",
      "Url didn't work, check name:  Diplonemidae\n",
      "Url didn't work for Choanoflagellida sp. SL163 checking:  Choanoflagellida\n",
      "Url didn't work for Tetraselmis sp. KB-CR05 checking:  Tetraselmis\n",
      "Url didn't work for Minorisa minuta checking:  Minorisa\n",
      "Url didn't work, check name:  Minorisa\n",
      "Url didn't work, check name:  Pterocystidae\n",
      "Url didn't work for Strombidium sp. SNB99-2 checking:  Strombidium\n",
      "Url didn't work for Hemistasia phaeocysticola checking:  Hemistasia\n",
      "Url didn't work for Coccoid pelagophyte CCMP1395 checking:  Coccoid\n",
      "Url didn't work, check name:  Coccoid\n",
      "Url didn't work, check name:  Myzocytiopsidaceae\n",
      "Url didn't work for Acanthamoeba mauritaniensis checking:  Acanthamoeba\n",
      "Url didn't work, check name:  Amphidiniopsidaceae\n",
      "Url didn't work, check name:  Duboscquellidae\n",
      "Url didn't work, check name:  Minorisa\n",
      "Url didn't work for Amphibelone anomala checking:  Amphibelone\n",
      "Url didn't work, check name:  Amphibelone\n",
      "Url didn't work for Pilea microphylla checking:  Pilea\n",
      "Url didn't work, check name:  Ichthyophonida\n",
      "Url didn't work for Thecate dinoflagellate UDTSW0701 checking:  Thecate\n",
      "Url didn't work, check name:  Thecate\n",
      "Url didn't work, check name:  Veneroida\n",
      "Url didn't work for Crustomastix sp. MBIC10709 checking:  Crustomastix\n",
      "Url didn't work for Chaetoceros sp. UNC1414 checking:  Chaetoceros\n",
      "Url didn't work for Chaetoceros sp. NIES-3971 checking:  Chaetoceros\n",
      "Url didn't work for Strombidinopsis sp. NSMS0601 checking:  Strombidinopsis\n",
      "Url didn't work for Cryptophyta sp. CR-MAL03 checking:  Cryptophyta\n",
      "Url didn't work for Oltmannsiellopsis sp. CCMP1240 checking:  Oltmannsiellopsis\n",
      "Url didn't work for Chaetoceros sp. TG07-C28 checking:  Chaetoceros\n",
      "Url didn't work, check name:  Polycystinea\n",
      "Url didn't work for Lineidae sp. 3DS checking:  Lineidae\n",
      "Url didn't work for Acanthoeca sp. SIOpierAcanth1 checking:  Acanthoeca\n",
      "Url didn't work for Malassezia globosa checking:  Malassezia\n",
      "Url didn't work, check name:  Malassezia\n",
      "Url didn't work for Minorisa sp. SRT710 checking:  Minorisa\n",
      "Url didn't work, check name:  Minorisa\n",
      "Url didn't work, check name:  Colpodellaceae\n",
      "Url didn't work for Uncultured marine alveolate Group II DH145-EKD20 checking:  Uncultured\n",
      "Url didn't work, check name:  Uncultured\n",
      "Url didn't work for Ciliate sp. NCMS0601 checking:  Ciliate\n",
      "Url didn't work, check name:  Ciliate\n",
      "Url didn't work for Sagitta sp. DP-2006 checking:  Sagitta\n",
      "Url didn't work, check name:  Ventrifissuridae\n",
      "Url didn't work for Eutintinnus cf. apertus checking:  Eutintinnus\n",
      "Url didn't work for Trachyrhizium urniformis checking:  Trachyrhizium\n",
      "Url didn't work, check name:  Trachyrhizium\n",
      "Url didn't work for Incisomonas marina checking:  Incisomonas\n",
      "Url didn't work, check name:  Incisomonas\n",
      "Url didn't work for eukaryote clone OLI11305 checking:  eukaryote\n",
      "Url didn't work, check name:  eukaryote\n",
      "Url didn't work for Cryothecomonas sp. APCC MC5-1Cryo checking:  Cryothecomonas\n",
      "Url didn't work for Warnowia sp. BSL-2009a checking:  Warnowia\n",
      "Url didn't work for Strombidium sp. JG-2011a checking:  Strombidium\n",
      "Url didn't work, check name:  Malassezia\n",
      "Url didn't work for Chytriodinium sp. Atlantic checking:  Chytriodinium\n",
      "Url didn't work, check name:  Melanthiaceae\n",
      "Url didn't work for Ventrifissura sp. SRT224 checking:  Ventrifissura\n",
      "Url didn't work, check name:  Ventrifissura\n",
      "Url didn't work for Pedinella sp. squamata checking:  Pedinella\n",
      "Url didn't work, check name:  Incisomonas\n",
      "Url didn't work for Eleutherascus lectardii checking:  Eleutherascus\n",
      "Url didn't work, check name:  Eleutherascus\n",
      "Url didn't work, check name:  Ceratoperidiniaeceae\n",
      "Url didn't work, check name:  Cryptomycota\n",
      "Url didn't work for Chlorophyta symbiont of Spongaster tetra checking:  Chlorophyta\n",
      "Url didn't work for eukaryote clone OLI11055 checking:  eukaryote\n",
      "Url didn't work, check name:  eukaryote\n",
      "Url didn't work for Oithona sp. Roadstead of Brest checking:  Oithona\n",
      "Url didn't work for Haptophyceae sp. EE-2014 checking:  Haptophyceae\n",
      "Url didn't work, check name:  Haptophyceae\n",
      "Url didn't work for Wallemia ichthyophaga checking:  Wallemia\n",
      "Url didn't work, check name:  Wallemia\n",
      "Url didn't work for Chaetoceros sp. Na26B1 checking:  Chaetoceros\n",
      "Url didn't work for eukaryote clone OLI11261 checking:  eukaryote\n",
      "Url didn't work, check name:  eukaryote\n",
      "Url didn't work, check name:  Selaginellaceae\n",
      "Url didn't work for Azadinium concinnum checking:  Azadinium\n",
      "Url didn't work for Halosphaera sp. M1670 checking:  Halosphaera\n",
      "Url didn't work for prasinophyte sp. MBIC10622 checking:  prasinophyte\n",
      "Url didn't work, check name:  prasinophyte\n",
      "Url didn't work for Thaumatomastix sp. CC002-Boundary Bay checking:  Thaumatomastix\n",
      "Url didn't work for Crypthecodinium sp. CAAE-CL2 checking:  Crypthecodinium\n",
      "Url didn't work for Euduboscquella sp. ex Favella arcuata checking:  Euduboscquella\n",
      "Url didn't work, check name:  Chlamydophryidae\n",
      "Url didn't work, check name:  Eudubosquellidae\n",
      "Url didn't work for Proterythropsis sp. BSL-2009a checking:  Proterythropsis\n",
      "Url didn't work for Amastigomonas sp. 2 Bamfield checking:  Amastigomonas\n",
      "Url didn't work, check name:  Paralagenidium\n",
      "Url didn't work for Labyrinthulochytrium arktikum checking:  Labyrinthulochytrium\n",
      "Url didn't work, check name:  Labyrinthulochytrium\n",
      "Url didn't work, check name:  Bicosoecida\n",
      "Url didn't work, check name:  Myoida\n",
      "Url didn't work for Hyphozyma variabilis checking:  Hyphozyma\n",
      "Url didn't work, check name:  Hyphozyma\n",
      "Url didn't work, check name:  Ebriacea\n",
      "Url didn't work, check name:  Scytomonadidae\n",
      "Url didn't work for Amphora sp. 38 checking:  Amphora\n",
      "Url didn't work for Tilletiopsis sp. B-TB2005 checking:  Tilletiopsis\n",
      "Url didn't work, check name:  Tilletiopsis\n",
      "Url didn't work, check name:  Phaffomycetaceae\n",
      "Url didn't work for Tintinnopsis sp. JG-2011a checking:  Tintinnopsis\n",
      "Url didn't work for Udeniomyces megalosporus checking:  Udeniomyces\n",
      "Url didn't work, check name:  Udeniomyces\n",
      "Url didn't work, check name:  Amphosoma\n",
      "Url didn't work, check name:  Trizona\n",
      "Url didn't work, check name:  Funariaceae\n",
      "Url didn't work for Pelagodinium sp. JD-2012 checking:  Pelagodinium\n",
      "Url didn't work, check name:  Segregatospumella\n",
      "Url didn't work for Oblongichytrium sp. BAFCcult 3519 checking:  Oblongichytrium\n",
      "Url didn't work for Lysilla sp. THS-2012 checking:  Lysilla\n",
      "Url didn't work for Coscinodiscus sp. GGM-2004 checking:  Coscinodiscus\n",
      "Url didn't work for Massisteria sp. diva checking:  Massisteria\n",
      "Url didn't work for Gigartacon muelleri checking:  Gigartacon\n",
      "Url didn't work, check name:  Gigartacon\n",
      "Url didn't work for Rhogostoma schuessleri checking:  Rhogostoma\n",
      "Url didn't work, check name:  Rhogostoma\n",
      "Url didn't work for Coccidia sp. Lutjanus/HAW36/TMW-2003 checking:  Coccidia\n",
      "Url didn't work, check name:  Coccidia\n",
      "Url didn't work, check name:  Hexaconus\n",
      "Url didn't work for Vexillifera sp. K9 checking:  Vexillifera\n",
      "Url didn't work for Pseudotontonia sp. JG-2011a checking:  Pseudotontonia\n",
      "Url didn't work for Halofilum ramosum checking:  Halofilum\n",
      "Url didn't work, check name:  Pinus\n",
      "Url didn't work, check name:  Cenangiaceae\n",
      "Url didn't work for Navicula sp. AT-145.08 checking:  Navicula\n",
      "Url didn't work for Candida <clade Candida/Lodderomyces clade> checking:  Candida\n",
      "Url didn't work for Uncultured marine alveolate Group II DH147-EKD6 checking:  Uncultured\n",
      "Url didn't work, check name:  Uncultured\n",
      "Url didn't work, check name:  Amphilithiidae\n",
      "Url didn't work for Protophysarum sp. PV64 checking:  Protophysarum\n",
      "Url didn't work, check name:  Protophysarum\n",
      "Url didn't work for Peziza echinospora checking:  Peziza\n",
      "Url didn't work for Dinophyceae sp. CCMP1878 checking:  Dinophyceae\n",
      "Url didn't work for Trieres sp. TN-2014 checking:  Trieres\n",
      "Url didn't work for Actinochloris sp. BCP-LG3VF20 checking:  Actinochloris\n",
      "Url didn't work for Fuscoporia ferrea checking:  Fuscoporia\n",
      "Url didn't work, check name:  Fuscoporia\n",
      "Url didn't work for Planomonas micra checking:  Planomonas\n",
      "Url didn't work, check name:  Planomonas\n",
      "Url didn't work for Diplonemida sp. 21sb RG-2016 checking:  Diplonemida\n",
      "Url didn't work for Carinoma sp. SA-2011 checking:  Carinoma\n",
      "Url didn't work, check name:  Phytoseiidae\n",
      "Url didn't work, check name:  Tulasnella\n",
      "Url didn't work, check name:  Balamuthiidae\n",
      "Url didn't work for Ceratiomyxella tahitiensis checking:  Ceratiomyxella\n",
      "Url didn't work, check name:  Catenophlyctis\n",
      "Url didn't work for Flamella arnhemensis checking:  Flamella\n",
      "Url didn't work for Chaetoceros sp. UNC1416 checking:  Chaetoceros\n",
      "Url didn't work for Polykrikos sp. EP-2016a checking:  Polykrikos\n",
      "Url didn't work for Amoebophrya sp. ex Ceratium tripos checking:  Amoebophrya\n",
      "Url didn't work for Creseis sp. GG-2005 checking:  Creseis\n",
      "Url didn't work for Amoebophrya sp. ex Prorocentrum micans checking:  Amoebophrya\n",
      "Url didn't work for Picobiliphyte sp. MS584-11 checking:  Picobiliphyte\n",
      "Url didn't work, check name:  Picobiliphyte\n",
      "Url didn't work for Nematodinium sp. BSL-2009a checking:  Nematodinium\n",
      "Url didn't work for Trichocerca elongata checking:  Trichocerca\n",
      "Url didn't work for Cryothecomonas longipes checking:  Cryothecomonas\n",
      "Url didn't work, check name:  Pterocystis\n",
      "Url didn't work for Chrysochromulina sp. Hp34EHU checking:  Chrysochromulina\n",
      "Url didn't work, check name:  Skeletonemataceae\n",
      "Url didn't work for eukaryote clone OLI11005 checking:  eukaryote\n",
      "Url didn't work, check name:  eukaryote\n",
      "Url didn't work for Pleurosigma sp. LM-2002 checking:  Pleurosigma\n",
      "Url didn't work for Minorisa sp. SRT705 checking:  Minorisa\n",
      "Url didn't work, check name:  Minorisa\n",
      "Url didn't work for Choanoflagellida sp. SL117#1 checking:  Choanoflagellida\n",
      "Url didn't work for Katablepharis japonica checking:  Katablepharis\n",
      "Url didn't work, check name:  Dactylodinium\n",
      "Url didn't work for Biddulphia sp. MACC 6 checking:  Biddulphia\n",
      "Url didn't work for Ventrifissura sp. SRT122 checking:  Ventrifissura\n",
      "Url didn't work, check name:  Ventrifissura\n",
      "Url didn't work for Oithonidae sp. DZMB624 checking:  Oithonidae\n"
     ]
    }
   ],
   "source": [
    "## What does WoRMS throw back?\n",
    "\n",
    "names = occ['scientificName'].unique()\n",
    "names = names[~pd.isnull(names)] # remove NaN\n",
    "name_id_dict, name_name_dict, name_taxid_dict, name_class_dict = WoRMS.run_get_worms_from_scientific_name(names, verbose_flag=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "816\n",
      "730\n"
     ]
    }
   ],
   "source": [
    "print(len(names))\n",
    "print(len(name_name_dict)) # OK, so at least most of these are matching, even if it's only the first word (e.g. Oithonidae sp. DZMB624 matching to Oithonidae)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**I'm going to complete the conversion for now**, ignoring the weird species designations. I can dig into that more after talking with Katie and better understanding what they mean."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>pcr_primer_forward</th>\n",
       "      <th>pcr_primer_reverse</th>\n",
       "      <th>DNA_sequence</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>basisOfRecord</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>MaterialSample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>MaterialSample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>MaterialSample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>MaterialSample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>MaterialSample</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude                env_broad_scale  \\\n",
       "0          -121.848  marine biome  (ENVO:00000447)   \n",
       "1          -121.848  marine biome  (ENVO:00000447)   \n",
       "2          -121.848  marine biome  (ENVO:00000447)   \n",
       "3          -121.846  marine biome  (ENVO:00000447)   \n",
       "4          -121.846  marine biome  (ENVO:00000447)   \n",
       "\n",
       "                 env_local_scale  \\\n",
       "0  coastal water (ENVO:00001250)   \n",
       "1  coastal water (ENVO:00001250)   \n",
       "2  coastal water (ENVO:00001250)   \n",
       "3  coastal water (ENVO:00001250)   \n",
       "4  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  pcr_primer_forward        pcr_primer_reverse  \\\n",
       "0   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC   \n",
       "1   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC   \n",
       "2   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC   \n",
       "3   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC   \n",
       "4   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC   \n",
       "\n",
       "                                        DNA_sequence scientificName  \\\n",
       "0  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...    Paracalanus   \n",
       "1  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...    Paracalanus   \n",
       "2  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...    Paracalanus   \n",
       "3  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...    Paracalanus   \n",
       "4  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...    Paracalanus   \n",
       "\n",
       "     kingdom      phylum        class      order         family        genus  \\\n",
       "0  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   \n",
       "1  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   \n",
       "2  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   \n",
       "3  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   \n",
       "4  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   \n",
       "\n",
       "    basisOfRecord  \n",
       "0  MaterialSample  \n",
       "1  MaterialSample  \n",
       "2  MaterialSample  \n",
       "3  MaterialSample  \n",
       "4  MaterialSample  "
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## basisOfRecord\n",
    "\n",
    "occ['basisOfRecord'] = 'MaterialSample'\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**materialSampleID** should be added. **Have these sequences been archived?** If so, use the biosample id from the archive. Otherwise, it's just a unique designation for the actual material sample - perhaps the same as Sequence_ID.\n",
    "\n",
    "**Also need to look into:**\n",
    "- identificationRemarks - recommended, specification of taxonomic identification process ideally including data on applied algorithm and reference database as well as on level of confidence in the resulting identification. Is this information available somewhere?\n",
    "- identificationReferences - recommended, link to protocol or code. Is this information available somewhere?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>pcr_primer_forward</th>\n",
       "      <th>pcr_primer_reverse</th>\n",
       "      <th>...</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>basisOfRecord</th>\n",
       "      <th>organismQuantity</th>\n",
       "      <th>organismQuantityType</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>14825</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>16094</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>22459</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>19312</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>16491</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude                env_broad_scale  \\\n",
       "0          -121.848  marine biome  (ENVO:00000447)   \n",
       "1          -121.848  marine biome  (ENVO:00000447)   \n",
       "2          -121.848  marine biome  (ENVO:00000447)   \n",
       "3          -121.846  marine biome  (ENVO:00000447)   \n",
       "4          -121.846  marine biome  (ENVO:00000447)   \n",
       "\n",
       "                 env_local_scale  \\\n",
       "0  coastal water (ENVO:00001250)   \n",
       "1  coastal water (ENVO:00001250)   \n",
       "2  coastal water (ENVO:00001250)   \n",
       "3  coastal water (ENVO:00001250)   \n",
       "4  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  pcr_primer_forward        pcr_primer_reverse  ... scientificName    kingdom  \\\n",
       "0   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...    Paracalanus  Eukaryota   \n",
       "1   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...    Paracalanus  Eukaryota   \n",
       "2   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...    Paracalanus  Eukaryota   \n",
       "3   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...    Paracalanus  Eukaryota   \n",
       "4   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...    Paracalanus  Eukaryota   \n",
       "\n",
       "       phylum        class      order         family        genus  \\\n",
       "0  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   \n",
       "1  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   \n",
       "2  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   \n",
       "3  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   \n",
       "4  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   \n",
       "\n",
       "    basisOfRecord organismQuantity  organismQuantityType  \n",
       "0  MaterialSample            14825    DNA sequence reads  \n",
       "1  MaterialSample            16094    DNA sequence reads  \n",
       "2  MaterialSample            22459    DNA sequence reads  \n",
       "3  MaterialSample            19312    DNA sequence reads  \n",
       "4  MaterialSample            16491    DNA sequence reads  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## organismQuantity (number of reads)\n",
    "\n",
    "occ['organismQuantity'] = plate['Reads']\n",
    "occ['organismQuantityType'] = 'DNA sequence reads'\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are 215537 rows where the number of reads is 0. **What does this mean? Is that an absence record??**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(280440, 22)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>pcr_primer_forward</th>\n",
       "      <th>pcr_primer_reverse</th>\n",
       "      <th>...</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>basisOfRecord</th>\n",
       "      <th>organismQuantity</th>\n",
       "      <th>organismQuantityType</th>\n",
       "      <th>sampleSizeValue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>14825</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>85600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>16094</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>90702</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>22459</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>130275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>19312</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>147220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>16491</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>121419</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude                env_broad_scale  \\\n",
       "0          -121.848  marine biome  (ENVO:00000447)   \n",
       "1          -121.848  marine biome  (ENVO:00000447)   \n",
       "2          -121.848  marine biome  (ENVO:00000447)   \n",
       "3          -121.846  marine biome  (ENVO:00000447)   \n",
       "4          -121.846  marine biome  (ENVO:00000447)   \n",
       "\n",
       "                 env_local_scale  \\\n",
       "0  coastal water (ENVO:00001250)   \n",
       "1  coastal water (ENVO:00001250)   \n",
       "2  coastal water (ENVO:00001250)   \n",
       "3  coastal water (ENVO:00001250)   \n",
       "4  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  pcr_primer_forward        pcr_primer_reverse  ...    kingdom      phylum  \\\n",
       "0   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Eukaryota  Arthropoda   \n",
       "1   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Eukaryota  Arthropoda   \n",
       "2   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Eukaryota  Arthropoda   \n",
       "3   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Eukaryota  Arthropoda   \n",
       "4   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Eukaryota  Arthropoda   \n",
       "\n",
       "         class      order         family        genus   basisOfRecord  \\\n",
       "0  Hexanauplia  Calanoida  Paracalanidae  Paracalanus  MaterialSample   \n",
       "1  Hexanauplia  Calanoida  Paracalanidae  Paracalanus  MaterialSample   \n",
       "2  Hexanauplia  Calanoida  Paracalanidae  Paracalanus  MaterialSample   \n",
       "3  Hexanauplia  Calanoida  Paracalanidae  Paracalanus  MaterialSample   \n",
       "4  Hexanauplia  Calanoida  Paracalanidae  Paracalanus  MaterialSample   \n",
       "\n",
       "  organismQuantity organismQuantityType  sampleSizeValue  \n",
       "0            14825   DNA sequence reads            85600  \n",
       "1            16094   DNA sequence reads            90702  \n",
       "2            22459   DNA sequence reads           130275  \n",
       "3            19312   DNA sequence reads           147220  \n",
       "4            16491   DNA sequence reads           121419  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## sampleSizeValue\n",
    "\n",
    "count_by_seq = plate.groupby('Sequence_ID', as_index=False)['Reads'].sum()\n",
    "occ = occ.merge(count_by_seq, how='left', left_on='eventID', right_on='Sequence_ID')\n",
    "occ.drop(columns='Sequence_ID', inplace=True)\n",
    "occ.rename(columns={'Reads':'sampleSizeValue'}, inplace=True)\n",
    "print(occ.shape)\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "sampleSizeValue is supposed to be the total number of reads in the sample for calculating the relative abundance of sequence variants. **Is it accurate to just sum all the reads by Sequence_ID?**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>pcr_primer_forward</th>\n",
       "      <th>pcr_primer_reverse</th>\n",
       "      <th>...</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>basisOfRecord</th>\n",
       "      <th>organismQuantity</th>\n",
       "      <th>organismQuantityType</th>\n",
       "      <th>sampleSizeValue</th>\n",
       "      <th>sampleSizeUnit</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>14825</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>85600</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>16094</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>90702</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>22459</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>130275</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>19312</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>147220</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>16491</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>121419</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude                env_broad_scale  \\\n",
       "0          -121.848  marine biome  (ENVO:00000447)   \n",
       "1          -121.848  marine biome  (ENVO:00000447)   \n",
       "2          -121.848  marine biome  (ENVO:00000447)   \n",
       "3          -121.846  marine biome  (ENVO:00000447)   \n",
       "4          -121.846  marine biome  (ENVO:00000447)   \n",
       "\n",
       "                 env_local_scale  \\\n",
       "0  coastal water (ENVO:00001250)   \n",
       "1  coastal water (ENVO:00001250)   \n",
       "2  coastal water (ENVO:00001250)   \n",
       "3  coastal water (ENVO:00001250)   \n",
       "4  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  pcr_primer_forward        pcr_primer_reverse  ...      phylum        class  \\\n",
       "0   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Arthropoda  Hexanauplia   \n",
       "1   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Arthropoda  Hexanauplia   \n",
       "2   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Arthropoda  Hexanauplia   \n",
       "3   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Arthropoda  Hexanauplia   \n",
       "4   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Arthropoda  Hexanauplia   \n",
       "\n",
       "       order         family        genus   basisOfRecord organismQuantity  \\\n",
       "0  Calanoida  Paracalanidae  Paracalanus  MaterialSample            14825   \n",
       "1  Calanoida  Paracalanidae  Paracalanus  MaterialSample            16094   \n",
       "2  Calanoida  Paracalanidae  Paracalanus  MaterialSample            22459   \n",
       "3  Calanoida  Paracalanidae  Paracalanus  MaterialSample            19312   \n",
       "4  Calanoida  Paracalanidae  Paracalanus  MaterialSample            16491   \n",
       "\n",
       "  organismQuantityType sampleSizeValue      sampleSizeUnit  \n",
       "0   DNA sequence reads           85600  DNA sequence reads  \n",
       "1   DNA sequence reads           90702  DNA sequence reads  \n",
       "2   DNA sequence reads          130275  DNA sequence reads  \n",
       "3   DNA sequence reads          147220  DNA sequence reads  \n",
       "4   DNA sequence reads          121419  DNA sequence reads  \n",
       "\n",
       "[5 rows x 23 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## sampleSizeUnit\n",
    "\n",
    "occ['sampleSizeUnit'] = 'DNA sequence reads'\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Are there any dois to associatedSquences for these data?** Should be a list of identifiers linking to archived (raw) sequence reads.\n",
    "\n",
    "## Save"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Save\n",
    "\n",
    "occ.to_csv('eDNA_practice_plate_occ_20201021.csv', index=False, na_rep='NaN')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Questions\n",
    "\n",
    "1. FilterID and Sequence_ID appear to be more or less the same. I assume that's how it's supposed to be? **It's like that in this example because Katie only sent me one plate. In general, multiple plates can be sequenced from the same filter, and so filterID and sequenceID would be more different in that case. Plates are labeled with letters from the alphabet; this example is plate S.** ID seems to be composed of: SAMPLING_cruise (cruise where the sample was collected) + c + SAMPLING_station_number (these data were collected on a CTD, so cast number) + _ + SAMPLING_bottle (bottle number - sometimes multiple sampels are taken from each bottle, which will be replicates a, b, c, etc.) + _ + edna (edna means that the sample was processed using the standard edna pipeline, i.e. a standard PBDF filter was used. There are other options, inlcuding hplc) + _ + 1, 2 or 3 (these numbers indicate PCR replicates from the same filter)\n",
    "2. There are a number of FilterID's (column labels) in the taxa table that do not appear in the plate data. What do these correspond to? Maybe controls? **Yes, these correspond to controls (see the following possible control categories). Is there some way to include this information in the DwC formatted data?** **<span style='color:red'> Since these are not occurrences, they would have to be incorporated into a MoF file somehow. </span>**\n",
    "    - CB = collection blank which was taken on the ship. Water that should be \"clean\" was passed through the filter, so there should be no reads in this sample, although it tends to be the \"dirtiest\" control. It checks for contamination during filtration, I think.\n",
    "    - EB = extraction blank which was taken in the lab but contains no DNA (i.e. a negative control), so there should be no reads in this sample. It checks if contamination occurred during DNA extraction (lab work pre-PCR).\n",
    "    - pcr blank = A sample that went through PCR but contained no input DNA (i.e. a negative control), so there should be no reads in this sample. It checks if contamination occurred during PCR.\n",
    "    - ArtComm = artificial community that went through PCR and contained DNA from species that should not be in the study system (i.e. a positive control), so you know what results you expect.\n",
    "3. The two metadata files seem to contain similar information. Is there something I'm missing here? **The MB_20200317_1326_18S_analysis_metadata file is the raw output from the bioinformatic pipeline; Katie reorganized this information into Plate_S_meta_OBIS_data. So yes, they're basically redundant.**\n",
    "5. I've identified target_gene to 18S (same as locus column). Is this correct? **Yes.** Is there more information that should be added? **If I want to put the full designation, I could say 18S rRNA, but I don't know if that's necessary.** Note that when I get more data, there will also be CO1 (mitochondrial cytochrome c oxidase subunit 1), 12S (section of mitochondrial DNA that codes for 12S rRNA), and some 16S (section of mitochondrial DNA that codes for 16S rRNA).\n",
    "4. Does information exist to add the following terms:\n",
    "    - sop: standard operating procedures used in assembly and/or annotation of metagenomes or a reference to a well documented protocol **Yes, there are protocols on protocol.io, and processing code on github - Katie can get links.**\n",
    "    - lib_layout: whether to expect single, paired, or other configuration of reads. **The reads are always paired, but are of different lengths depending on which locus was used; 18S is shorter, while CO1 and 12S are longer. I'm not sure if this information needs to be/can be included.**\n",
    "    - target_subfragment: name of subfragment of a gene or marker (e.g. V6). **For 18S, the subfragment will be V9. Katie doesn't think there's something analogous for CO1 and 12S, but there will be for 16S.**\n",
    "    - primer names - **Katie has this (should be on protocols.io).**\n",
    "    - primer references - **Katie has this (should be on protocols.io).**\n",
    "5. In the taxa table, there appear to be multiple things used to indicate missing data. Some of them include: 'unassigned', 'no_hit', 'unknown', etc. Can I assume all of these are NaN? **Yes, but just for my own information, here are the meanings:**\n",
    "    - unknown = [Genbank](https://www.ncbi.nlm.nih.gov/genbank/) doesn't have a scientifically-agreed-upon name for some taxonomic levels. This is how a species can be given, but not a genus, for example. It just means the Genus name either doesn't exist, or there isn't enough scientific consensus to give a name.\n",
    "    - no_hit = no [BLAST](https://blast.ncbi.nlm.nih.gov/Blast.cgi) hits for the ASV.\n",
    "    - unassigned = The ASV got BLAST hits, but the post-processing program [Megan6](https://uni-tuebingen.de/fakultaeten/mathematisch-naturwissenschaftliche-fakultaet/fachbereiche/informatik/lehrstuehle/algorithms-in-bioinformatics/software/megan6/) didn't assign the ASV to any taxonomic group.\n",
    "    - g_ or s_ = After Megan there's an additional filtering step where Genus or Species designations are only left if they were assigned with a higher confidence. If this threshold is not met, a g_ or s_ is given.\n",
    "6. The scientificName column needs to contain a Linnaean scientific name, or an OTU identifier from BOLD or UNITE. I don't know if there are other options, but will investigate. **The non-Linnaean species designations in these data come from GenBank. Will OBIS accept a GenBankID? Possibly not, since GenBank is not curated, but it would be ideal for us if they would. Using megan is one way to get around the possibility of bad sequences on GenBank.** Note that only 86 taxa (out of about 816 unique taxa) have no match on WoRMS at any taxonomic level, and therefore no Linnaean name option. **<span style='color:red'> OBIS cannot accept a record that has no taxonomic identification except for GenBank ID. These records would need to be excluded. Is there some other way to include this information? E.g. in the MoF? Is there a column where the GenBank ID can at least be noted? </span>**\n",
    "7. Have these sequences been archived? Will they be archived? If so where, and what kind of identifying information will be likely to come along with it? Are there DOIs? **If the data are already published, the sequences have been archived with the [Sequence Read Archive (SRA)](https://www.ncbi.nlm.nih.gov/sra). It will be up to Francisco if he's interested in making data public and archiving sequences pre-publication. \n",
    "8. Can these be added:\n",
    "    - identificationRemarks - recommended, specification of taxonomic identification process ideally including data on applied algorithm and reference database as well as on level of confidence in the resulting identification. Is this information available somewhere? **This information might include the GenBank version used, the fact that BLAST and Megan6 were used, and the BLAST confidence level.**\n",
    "    - identificationReferences - recommended, link to protocol or code. Is this information available somewhere? **This can link to any published papers and github.**\n",
    "9. In most of these data, Reads = 0. What does this indicate? **The ASV was not present in another sample on the same plate. Can be dropped.**\n",
    "10. sampleSizeValue is supposed to be the total number of reads in the sample for calculating the relative abundance of sequence variants. Is it accurate to just sum all the reads by Sequence_ID? **Yes, this is fine. If we run our decontamination script before submitting, some reads might be removed (e.g. from Bacteria). Does OBIS want the decontaminated version? Probably yes.**\n",
    "\n",
    "**<span style='color:red'> Aside from the above notes in red, I'm just waiting for a new version of these data from Katie with the additional metadata fields I discussed with her populated. </span>**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "venv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
