{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Darwin Core Conversion of eDNA Sequence Data Using the DNA Derived Data Extension\n",
    "\n",
    "**Author:** Diana LaScala-Gruenewald\n",
    "\n",
    "**Last Updated:** 07-Jul-2021\n",
    "\n",
    "**Requirements:**\n",
    "- Python 3\n",
    "- Python 3 packages:\n",
    "    - datetime\n",
    "    - random\n",
    "    - sys\n",
    "- External packages:\n",
    "    - numpy\n",
    "    - pandas\n",
    "    - pytz\n",
    "- Custom modules:\n",
    "    - WoRMS\n",
    "\n",
    "**Resources:**\n",
    "- https://docs.gbif-uat.org/publishing-dna-derived-data/1.0/en/\n",
    "- https://tools.gbif.org/dwca-validator/extension.do?id=http://rs.gbif.org/terms/1.0/DNADerivedData"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Imports\n",
    "\n",
    "from datetime import datetime\n",
    "import os\n",
    "import random\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pytz # for handling time zones\n",
    "\n",
    "import WoRMS # custom functions for querying WoRMS API"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load data\n",
    "\n",
    "Note that in a Jupyter Notebook, the current directory is always where the .ipynb file is being run."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(280440, 11)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ASV</th>\n",
       "      <th>FilterID</th>\n",
       "      <th>Sequence_ID</th>\n",
       "      <th>Reads</th>\n",
       "      <th>Kingdom</th>\n",
       "      <th>Phylum</th>\n",
       "      <th>Class</th>\n",
       "      <th>Order</th>\n",
       "      <th>Family</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>05114c01_12_edna_1</td>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>14825</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>unassigned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>05114c01_12_edna_2</td>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>16094</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>unassigned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>05114c01_12_edna_3</td>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>22459</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>unassigned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>11216c01_12_edna_1</td>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>19312</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>unassigned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>11216c01_12_edna_2</td>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>16491</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>unassigned</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 ASV            FilterID  \\\n",
       "0  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  05114c01_12_edna_1   \n",
       "1  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  05114c01_12_edna_2   \n",
       "2  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  05114c01_12_edna_3   \n",
       "3  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  11216c01_12_edna_1   \n",
       "4  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  11216c01_12_edna_2   \n",
       "\n",
       "            Sequence_ID  Reads    Kingdom      Phylum        Class      Order  \\\n",
       "0  05114c01_12_edna_1_S  14825  Eukaryota  Arthropoda  Hexanauplia  Calanoida   \n",
       "1  05114c01_12_edna_2_S  16094  Eukaryota  Arthropoda  Hexanauplia  Calanoida   \n",
       "2  05114c01_12_edna_3_S  22459  Eukaryota  Arthropoda  Hexanauplia  Calanoida   \n",
       "3  11216c01_12_edna_1_S  19312  Eukaryota  Arthropoda  Hexanauplia  Calanoida   \n",
       "4  11216c01_12_edna_2_S  16491  Eukaryota  Arthropoda  Hexanauplia  Calanoida   \n",
       "\n",
       "          Family        Genus     Species  \n",
       "0  Paracalanidae  Paracalanus  unassigned  \n",
       "1  Paracalanidae  Paracalanus  unassigned  \n",
       "2  Paracalanidae  Paracalanus  unassigned  \n",
       "3  Paracalanidae  Paracalanus  unassigned  \n",
       "4  Paracalanidae  Paracalanus  unassigned  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Plate data\n",
    "\n",
    "filename = os.getcwd().replace('src', os.path.join('raw', 'asv_table.csv'))  \n",
    "plate = pd.read_csv(filename)\n",
    "print(plate.shape)\n",
    "plate.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Plate data contains:\n",
    "\n",
    "| Column name| Column definition                                                                                                                                                                           |\n",
    "|------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|\n",
    " ASV        | The sequence of the Amplicon Sequence Variant observed                                                                                                                                       |\n",
    "| FilterID   | A unique identifier for the filter the sample was obtained from, composed of: <br>- cruise number <br>- CTD cast number<br>- CTD bottle number <br>- filter indicator <br>- replicate number |\n",
    "| Sequence_ID| The FilterID plus a letter indicating which plate the sample was on when sequenced                                                                                                           |\n",
    "| Reads      | The number of reads for the ASV                                                                                                                                                             |\n",
    "| Kingdom    | The Kingdom of the taxonomic identity assigned to the ASV, if known                                                                                                                          |\n",
    "| Phylum     | The Phylum of the taxonomic identity assigned to the ASV, if known                                                                                                                           |\n",
    "| Class      | The Class of the taxonomic identity assigned to the ASV, if known                                                                                                                            |\n",
    "| Order      | The Order of the taxonomic identity assigned to the ASV, if known                                                                                                                            |\n",
    "| Family     | The Family of the taxonomic identity assigned to the ASV, if known                                                                                                                           |\n",
    "| Genus      | The Genus of the taxonomic identity assigned to the ASV, if known                                                                                                                            |\n",
    "| Species    | The Species of the taxonomic identity assigned to the ASV, if known                                                                                                                          |\n",
    "\n",
    "Additionally, taxonomic columns may include the following designations:\n",
    "- **unknown** = GenBank couldn't give a scientifically-agreed-upon name for a given taxonomic rank. I.e., either the name doesn't exist, or there isn't enough scientific consensus to give a name.\n",
    "- **no_hit** = BLAST did not find any hits for the ASV.\n",
    "- **unassigned** = The ASV got BLAST hits, but the post-processing program Megan6 didn't assign the ASV to any taxonomic group.\n",
    "- **g_** or **s_** = Megan6 assigned the ASV to a genus or species, but not with high enough confidence to include it. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(60, 67)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>library</th>\n",
       "      <th>tag_sequence</th>\n",
       "      <th>primer_sequence_forward</th>\n",
       "      <th>primer_sequence_reverse</th>\n",
       "      <th>R1</th>\n",
       "      <th>R2</th>\n",
       "      <th>PlateID</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>...</th>\n",
       "      <th>pcr_primer_name_forward</th>\n",
       "      <th>pcr_primer_name_reverse</th>\n",
       "      <th>pcr_primer_reference</th>\n",
       "      <th>seq_meth</th>\n",
       "      <th>sequencing_facility</th>\n",
       "      <th>seqID</th>\n",
       "      <th>identificationRemarks</th>\n",
       "      <th>identificationReferences</th>\n",
       "      <th>FilterID</th>\n",
       "      <th>associatedSequences</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>14213c01_12_eDNA_1</td>\n",
       "      <td>S1</td>\n",
       "      <td>ACGAGACTGATT</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>14213c01_12_edna_1_S1_L001_R1_001.fastq.gz</td>\n",
       "      <td>14213c01_12_edna_1_S1_L001_R2_001.fastq.gz</td>\n",
       "      <td>S</td>\n",
       "      <td>environmental</td>\n",
       "      <td>18S</td>\n",
       "      <td>...</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>Amaral-Zettler LA, McCliment EA, Ducklow HW, H...</td>\n",
       "      <td>NGS Illumina Miseq</td>\n",
       "      <td>Stanford</td>\n",
       "      <td>14213c01_12_edna_1_S</td>\n",
       "      <td>Genbank nr Release 221 September 20 2017</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>14213c01_12_edna</td>\n",
       "      <td>NCBI BioProject accession number PRJNA433203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>14213c01_12_eDNA_2</td>\n",
       "      <td>S2</td>\n",
       "      <td>GAATACCAAGTC</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>14213c01_12_edna_2_S2_L001_R1_001.fastq.gz</td>\n",
       "      <td>14213c01_12_edna_2_S2_L001_R2_001.fastq.gz</td>\n",
       "      <td>S</td>\n",
       "      <td>environmental</td>\n",
       "      <td>18S</td>\n",
       "      <td>...</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>Amaral-Zettler LA, McCliment EA, Ducklow HW, H...</td>\n",
       "      <td>NGS Illumina Miseq</td>\n",
       "      <td>Stanford</td>\n",
       "      <td>14213c01_12_edna_2_S</td>\n",
       "      <td>Genbank nr Release 221 September 20 2017</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>14213c01_12_edna</td>\n",
       "      <td>NCBI BioProject accession number PRJNA433203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>14213c01_12_eDNA_3</td>\n",
       "      <td>S3</td>\n",
       "      <td>CGAGGGAAAGTC</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>14213c01_12_edna_3_S3_L001_R1_001.fastq.gz</td>\n",
       "      <td>14213c01_12_edna_3_S3_L001_R2_001.fastq.gz</td>\n",
       "      <td>S</td>\n",
       "      <td>environmental</td>\n",
       "      <td>18S</td>\n",
       "      <td>...</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>Amaral-Zettler LA, McCliment EA, Ducklow HW, H...</td>\n",
       "      <td>NGS Illumina Miseq</td>\n",
       "      <td>Stanford</td>\n",
       "      <td>14213c01_12_edna_3_S</td>\n",
       "      <td>Genbank nr Release 221 September 20 2017</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>14213c01_12_edna</td>\n",
       "      <td>NCBI BioProject accession number PRJNA433203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>22013c01_12_eDNA_1</td>\n",
       "      <td>S4</td>\n",
       "      <td>GAACACTTTGGA</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>22013c01_12_edna_1_S4_L001_R1_001.fastq.gz</td>\n",
       "      <td>22013c01_12_edna_1_S4_L001_R2_001.fastq.gz</td>\n",
       "      <td>S</td>\n",
       "      <td>environmental</td>\n",
       "      <td>18S</td>\n",
       "      <td>...</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>Amaral-Zettler LA, McCliment EA, Ducklow HW, H...</td>\n",
       "      <td>NGS Illumina Miseq</td>\n",
       "      <td>Stanford</td>\n",
       "      <td>22013c01_12_edna_1_S</td>\n",
       "      <td>Genbank nr Release 221 September 20 2017</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>22013c01_12_edna</td>\n",
       "      <td>NCBI BioProject accession number PRJNA433203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>22013c01_12_eDNA_2</td>\n",
       "      <td>S5</td>\n",
       "      <td>ACTCACAGGAAT</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>22013c01_12_edna_2_S5_L001_R1_001.fastq.gz</td>\n",
       "      <td>22013c01_12_edna_2_S5_L001_R2_001.fastq.gz</td>\n",
       "      <td>S</td>\n",
       "      <td>environmental</td>\n",
       "      <td>18S</td>\n",
       "      <td>...</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>Amaral-Zettler LA, McCliment EA, Ducklow HW, H...</td>\n",
       "      <td>NGS Illumina Miseq</td>\n",
       "      <td>Stanford</td>\n",
       "      <td>22013c01_12_edna_2_S</td>\n",
       "      <td>Genbank nr Release 221 September 20 2017</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>22013c01_12_edna</td>\n",
       "      <td>NCBI BioProject accession number PRJNA433203</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 67 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          sample_name library  tag_sequence primer_sequence_forward  \\\n",
       "0  14213c01_12_eDNA_1      S1  ACGAGACTGATT        GTACACACCGCCCGTC   \n",
       "1  14213c01_12_eDNA_2      S2  GAATACCAAGTC        GTACACACCGCCCGTC   \n",
       "2  14213c01_12_eDNA_3      S3  CGAGGGAAAGTC        GTACACACCGCCCGTC   \n",
       "3  22013c01_12_eDNA_1      S4  GAACACTTTGGA        GTACACACCGCCCGTC   \n",
       "4  22013c01_12_eDNA_2      S5  ACTCACAGGAAT        GTACACACCGCCCGTC   \n",
       "\n",
       "    primer_sequence_reverse                                          R1  \\\n",
       "0  TGATCCTTCTGCAGGTTCACCTAC  14213c01_12_edna_1_S1_L001_R1_001.fastq.gz   \n",
       "1  TGATCCTTCTGCAGGTTCACCTAC  14213c01_12_edna_2_S2_L001_R1_001.fastq.gz   \n",
       "2  TGATCCTTCTGCAGGTTCACCTAC  14213c01_12_edna_3_S3_L001_R1_001.fastq.gz   \n",
       "3  TGATCCTTCTGCAGGTTCACCTAC  22013c01_12_edna_1_S4_L001_R1_001.fastq.gz   \n",
       "4  TGATCCTTCTGCAGGTTCACCTAC  22013c01_12_edna_2_S5_L001_R1_001.fastq.gz   \n",
       "\n",
       "                                           R2 PlateID    sample_type  \\\n",
       "0  14213c01_12_edna_1_S1_L001_R2_001.fastq.gz       S  environmental   \n",
       "1  14213c01_12_edna_2_S2_L001_R2_001.fastq.gz       S  environmental   \n",
       "2  14213c01_12_edna_3_S3_L001_R2_001.fastq.gz       S  environmental   \n",
       "3  22013c01_12_edna_1_S4_L001_R2_001.fastq.gz       S  environmental   \n",
       "4  22013c01_12_edna_2_S5_L001_R2_001.fastq.gz       S  environmental   \n",
       "\n",
       "  target_gene  ...  pcr_primer_name_forward pcr_primer_name_reverse  \\\n",
       "0         18S  ...                    1391f                   EukBr   \n",
       "1         18S  ...                    1391f                   EukBr   \n",
       "2         18S  ...                    1391f                   EukBr   \n",
       "3         18S  ...                    1391f                   EukBr   \n",
       "4         18S  ...                    1391f                   EukBr   \n",
       "\n",
       "                                pcr_primer_reference            seq_meth  \\\n",
       "0  Amaral-Zettler LA, McCliment EA, Ducklow HW, H...  NGS Illumina Miseq   \n",
       "1  Amaral-Zettler LA, McCliment EA, Ducklow HW, H...  NGS Illumina Miseq   \n",
       "2  Amaral-Zettler LA, McCliment EA, Ducklow HW, H...  NGS Illumina Miseq   \n",
       "3  Amaral-Zettler LA, McCliment EA, Ducklow HW, H...  NGS Illumina Miseq   \n",
       "4  Amaral-Zettler LA, McCliment EA, Ducklow HW, H...  NGS Illumina Miseq   \n",
       "\n",
       "   sequencing_facility                 seqID  \\\n",
       "0             Stanford  14213c01_12_edna_1_S   \n",
       "1             Stanford  14213c01_12_edna_2_S   \n",
       "2             Stanford  14213c01_12_edna_3_S   \n",
       "3             Stanford  22013c01_12_edna_1_S   \n",
       "4             Stanford  22013c01_12_edna_2_S   \n",
       "\n",
       "                      identificationRemarks  \\\n",
       "0  Genbank nr Release 221 September 20 2017   \n",
       "1  Genbank nr Release 221 September 20 2017   \n",
       "2  Genbank nr Release 221 September 20 2017   \n",
       "3  Genbank nr Release 221 September 20 2017   \n",
       "4  Genbank nr Release 221 September 20 2017   \n",
       "\n",
       "                            identificationReferences          FilterID  \\\n",
       "0  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...  14213c01_12_edna   \n",
       "1  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...  14213c01_12_edna   \n",
       "2  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...  14213c01_12_edna   \n",
       "3  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...  22013c01_12_edna   \n",
       "4  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...  22013c01_12_edna   \n",
       "\n",
       "                            associatedSequences  \n",
       "0  NCBI BioProject accession number PRJNA433203  \n",
       "1  NCBI BioProject accession number PRJNA433203  \n",
       "2  NCBI BioProject accession number PRJNA433203  \n",
       "3  NCBI BioProject accession number PRJNA433203  \n",
       "4  NCBI BioProject accession number PRJNA433203  \n",
       "\n",
       "[5 rows x 67 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Plate metadata\n",
    "\n",
    "filename = os.getcwd().replace('src', os.path.join('raw', 'metadata_table.csv'))  \n",
    "meta = pd.read_csv(filename)\n",
    "print(meta.shape)\n",
    "meta.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['sample_name', 'library', 'tag_sequence', 'primer_sequence_forward',\n",
       "       'primer_sequence_reverse', 'R1', 'R2', 'PlateID', 'sample_type',\n",
       "       'target_gene', 'replicate', 'eventDate', 'SAMPLING_cruise',\n",
       "       'SAMPLING_station_number', 'SAMPLING_bottle', 'SAMPLING_station',\n",
       "       'depth', 'decimalLatitude', 'decimalLongitude', 'SAMPLING_PI',\n",
       "       'SAMPLING_institute', 'SAMPLING_campaign', 'project_name',\n",
       "       'SAMPLING_platform', 'SAMPLING_platform_type', 'samp_collect_device',\n",
       "       'geo_loc_name', 'fluor', 'SAMPLING_rdepth', 'density', 'pressure',\n",
       "       'nitrate', 'chlorophyll', 'diss_oxygen', 'salinity', 'temp',\n",
       "       'samp_filter_ext_type', 'samp_filter_size_ext', 'env_broad_scale',\n",
       "       'env_local_scale', 'env_medium', 'ESP', 'SC', 'minimumDepthInMeters',\n",
       "       'maximumDepthInMeters', 'start_GMT', 'end_GMT', 'deployment_ID',\n",
       "       'date_PCR', 'DNA_concentration', 'PCR_settings', 'samp_vol_we_dna_ext',\n",
       "       'extraction_date', 'nucl_acid_ext', 'nucl_acid_amp', 'samp_store_temp',\n",
       "       'sop', 'pcr_primer_name_forward', 'pcr_primer_name_reverse',\n",
       "       'pcr_primer_reference', 'seq_meth', 'sequencing_facility', 'seqID',\n",
       "       'identificationRemarks', 'identificationReferences', 'FilterID',\n",
       "       'associatedSequences'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Metadata contains:\n",
    "\n",
    "| Column name             | Column definition                                                                       |\n",
    "|-------------------------|-----------------------------------------------------------------------------------------|\n",
    "| primer_sequence_forward | The sequence of the forward primer used during PCR                                      |\n",
    "| primer_sequence_reverse | The sequence of the reverse primer used during PCR                                      |\n",
    "| target_gene             | The gene being targeted for amplification during PCR                                    |\n",
    "| eventDate               | The date (and time, if available) the water sample was collected                        |\n",
    "| decimalLatitude         | The latitude in decimal degrees where the water sample was collected (WGS84)            |\n",
    "| decimalLongitude        | The longitude in decimal degrees where the water sample was collected (WGS84)           |\n",
    "| env_broad_scale         | The most broad descriptor of the environment from which the water sample was collected  |\n",
    "| env_local_scale         | A more specific descriptor of the environment from which the water sample was collected |\n",
    "| env_medium              | A descriptor of the medium from which the DNA was collected                             |\n",
    "| sop                     | Links or references to standard operating protocols used to obtain the data             |\n",
    "| pcr_primer_name_forward | Name of the forward primer used during PCR                                              |\n",
    "| pcr_primer_name_reverse | Name of the reverse primer used during PCR                                              |\n",
    "| pcr_primer_reference    | Reference for PCR primers                                                               |\n",
    "|                         |                                                                                         |\n",
    "\n",
    "Note that only columns used in the following processing script are described."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create occurrence file\n",
    "\n",
    "For this data set, an `event` is a filtered water sample that was sequenced and an `occurrence` is an ASV observed within a water sample. Since there are no event-level measurements (i.e., measurements that are associated with the water sample but not the ASV), a separate event file is not required. We will assemble an occurrence file complying with the Occurrence Core and a DNA derived data (ddd) file complying with the DNA derived data extension."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(280440, 1)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID\n",
       "0  05114c01_12_edna_1_S\n",
       "1  05114c01_12_edna_2_S\n",
       "2  05114c01_12_edna_3_S\n",
       "3  11216c01_12_edna_1_S\n",
       "4  11216c01_12_edna_2_S"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## eventID - the Sequence_ID column in the plate dataframe uniquely identifies a water sample\n",
    "\n",
    "occ = pd.DataFrame({'eventID':plate['Sequence_ID']})\n",
    "print(occ.shape)\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2/20/14 15:33</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2/20/14 15:33</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2/20/14 15:33</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>4/21/16 14:39</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>4/21/16 14:39</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID      eventDate  decimalLatitude  decimalLongitude\n",
       "0  05114c01_12_edna_1_S  2/20/14 15:33          36.7958          -121.848\n",
       "1  05114c01_12_edna_2_S  2/20/14 15:33          36.7958          -121.848\n",
       "2  05114c01_12_edna_3_S  2/20/14 15:33          36.7958          -121.848\n",
       "3  11216c01_12_edna_1_S  4/21/16 14:39          36.7962          -121.846\n",
       "4  11216c01_12_edna_2_S  4/21/16 14:39          36.7962          -121.846"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Merge with plate_meta to obtain eventDate, decimalLatitude, and decimalLongitude\n",
    "\n",
    "metadata_cols = [\n",
    "    'seqID',\n",
    "    'eventDate', \n",
    "    'decimalLatitude', \n",
    "    'decimalLongitude',\n",
    "]\n",
    "\n",
    "dwc_cols = metadata_cols.copy()\n",
    "dwc_cols[0] = 'eventID'\n",
    "\n",
    "occ = occ.merge(meta[metadata_cols], how='left', left_on='eventID', right_on='seqID')\n",
    "occ.drop(columns='seqID', inplace=True)\n",
    "occ.columns = dwc_cols\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude  \n",
       "0          -121.848  \n",
       "1          -121.848  \n",
       "2          -121.848  \n",
       "3          -121.846  \n",
       "4          -121.846  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Format eventDate\n",
    "\n",
    "pst = pytz.timezone('America/Los_Angeles')\n",
    "eventDate = [pst.localize(datetime.strptime(dt, '%m/%d/%y %H:%M')).isoformat() for dt in occ['eventDate']]\n",
    "occ['eventDate'] = eventDate\n",
    "\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>occurrenceID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>05114c01_12_edna_1_S_occ1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>05114c01_12_edna_2_S_occ1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>05114c01_12_edna_3_S_occ1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>11216c01_12_edna_1_S_occ1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>11216c01_12_edna_2_S_occ1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude               occurrenceID  \n",
       "0          -121.848  05114c01_12_edna_1_S_occ1  \n",
       "1          -121.848  05114c01_12_edna_2_S_occ1  \n",
       "2          -121.848  05114c01_12_edna_3_S_occ1  \n",
       "3          -121.846  11216c01_12_edna_1_S_occ1  \n",
       "4          -121.846  11216c01_12_edna_2_S_occ1  "
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Add an occurrenceID that will uniquely identify an ASV within a water sample\n",
    "\n",
    "occ['occurrenceID'] = plate.groupby('Sequence_ID')['ASV'].cumcount()+1\n",
    "occ['occurrenceID'] = occ['eventID'] + '_occ' + occ['occurrenceID'].astype(str)\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>occurrenceID</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>05114c01_12_edna_1_S_occ1</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>05114c01_12_edna_2_S_occ1</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>05114c01_12_edna_3_S_occ1</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>11216c01_12_edna_1_S_occ1</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>11216c01_12_edna_2_S_occ1</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude               occurrenceID scientificName    kingdom  \\\n",
       "0          -121.848  05114c01_12_edna_1_S_occ1     unassigned  Eukaryota   \n",
       "1          -121.848  05114c01_12_edna_2_S_occ1     unassigned  Eukaryota   \n",
       "2          -121.848  05114c01_12_edna_3_S_occ1     unassigned  Eukaryota   \n",
       "3          -121.846  11216c01_12_edna_1_S_occ1     unassigned  Eukaryota   \n",
       "4          -121.846  11216c01_12_edna_2_S_occ1     unassigned  Eukaryota   \n",
       "\n",
       "       phylum        class      order         family        genus  \n",
       "0  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus  \n",
       "1  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus  \n",
       "2  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus  \n",
       "3  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus  \n",
       "4  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## scientificName, taxonomic info\n",
    "\n",
    "occ['scientificName'] = plate['Species']\n",
    "occ['kingdom'] = plate['Kingdom']\n",
    "occ['phylum'] = plate['Phylum']\n",
    "occ['class'] = plate['Class']\n",
    "occ['order'] = plate['Order']\n",
    "occ['family'] = plate['Family']\n",
    "occ['genus'] = plate['Genus']\n",
    "\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For the purpose of submitting data to OBIS, all the variations on missing data (e.g. \"unknown,\" \"no_hit,\" etc.) do not add information. We can replace these with NaN, which is easy to work with in pandas."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>occurrenceID</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>05114c01_12_edna_1_S_occ1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>05114c01_12_edna_2_S_occ1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>05114c01_12_edna_3_S_occ1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>11216c01_12_edna_1_S_occ1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>11216c01_12_edna_2_S_occ1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude               occurrenceID scientificName    kingdom  \\\n",
       "0          -121.848  05114c01_12_edna_1_S_occ1            NaN  Eukaryota   \n",
       "1          -121.848  05114c01_12_edna_2_S_occ1            NaN  Eukaryota   \n",
       "2          -121.848  05114c01_12_edna_3_S_occ1            NaN  Eukaryota   \n",
       "3          -121.846  11216c01_12_edna_1_S_occ1            NaN  Eukaryota   \n",
       "4          -121.846  11216c01_12_edna_2_S_occ1            NaN  Eukaryota   \n",
       "\n",
       "       phylum        class      order         family        genus  \n",
       "0  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus  \n",
       "1  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus  \n",
       "2  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus  \n",
       "3  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus  \n",
       "4  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Replace 'unknown', 'unassigned', etc. in scientificName and taxonomy columns with NaN\n",
    "\n",
    "cols = ['scientificName', 'kingdom', 'phylum', 'class', 'order', 'family', 'genus']\n",
    "occ[cols] = occ[cols].replace({'unassigned':np.nan,\n",
    "                              's_':np.nan,\n",
    "                              'g_':np.nan,\n",
    "                              'unknown':np.nan,\n",
    "                              'no_hit':np.nan})\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "323\n"
     ]
    }
   ],
   "source": [
    "## Get unique species names\n",
    "\n",
    "names = occ['scientificName'].unique()\n",
    "names = names[~pd.isnull(names)]  # remove NaN\n",
    "print(len(names))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "These names **do** have to be WoRMS-approved in order to show up as valid occurrences OBIS. But there are a number of entries in the `scientificName` column, like \"uncultured marine eukaryote,\" \"eukaryote clone OLI11007,\" and \"Acantharian sp. 6201,\" that are **not proper Latin species names**. Since these essentially indicate that a more precise name is unknown, it seemed reasonable to replace these with NaN as well. **I used a simple rule to filter them out, but it's important to check and see if any true species names are being removed.**\n",
    "\n",
    "To check names:\n",
    "```python\n",
    "names = occ['scientificName'].unique()\n",
    "names = names[~pd.isnull(names)]  # remove NaN\n",
    "for name in names:\n",
    "    words_in_name = name.split(' ')\n",
    "    if len(words_in_name) > 2:\n",
    "        print(name)\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'dx.doi.org/10.17504/protocols.io.xjufknw | dx.doi.org/10.17504/protocols.io.n2vdge6 | https://github.com/MBARI-BOG/BOG-Banzai-Dada2-Pipeline'"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Clean sop\n",
    "\n",
    "occ['sop'] = occ['sop'].str.replace('|', ' | ', regex=False)\n",
    "occ['sop'].iloc[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>pcr_primer_forward</th>\n",
       "      <th>pcr_primer_reverse</th>\n",
       "      <th>pcr_primer_name_forward</th>\n",
       "      <th>pcr_primer_name_reverse</th>\n",
       "      <th>pcr_primer_reference</th>\n",
       "      <th>sop</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2/20/14 15:33</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>Amaral-Zettler LA, McCliment EA, Ducklow HW, H...</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw | dx....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2/20/14 15:33</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>Amaral-Zettler LA, McCliment EA, Ducklow HW, H...</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw | dx....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2/20/14 15:33</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>Amaral-Zettler LA, McCliment EA, Ducklow HW, H...</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw | dx....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>4/21/16 14:39</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>Amaral-Zettler LA, McCliment EA, Ducklow HW, H...</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw | dx....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>4/21/16 14:39</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>Amaral-Zettler LA, McCliment EA, Ducklow HW, H...</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw | dx....</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID      eventDate  decimalLatitude  decimalLongitude  \\\n",
       "0  05114c01_12_edna_1_S  2/20/14 15:33          36.7958          -121.848   \n",
       "1  05114c01_12_edna_2_S  2/20/14 15:33          36.7958          -121.848   \n",
       "2  05114c01_12_edna_3_S  2/20/14 15:33          36.7958          -121.848   \n",
       "3  11216c01_12_edna_1_S  4/21/16 14:39          36.7962          -121.846   \n",
       "4  11216c01_12_edna_2_S  4/21/16 14:39          36.7962          -121.846   \n",
       "\n",
       "                 env_broad_scale                env_local_scale  \\\n",
       "0  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   \n",
       "1  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   \n",
       "2  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   \n",
       "3  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   \n",
       "4  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  pcr_primer_forward        pcr_primer_reverse pcr_primer_name_forward  \\\n",
       "0   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC                   1391f   \n",
       "1   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC                   1391f   \n",
       "2   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC                   1391f   \n",
       "3   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC                   1391f   \n",
       "4   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC                   1391f   \n",
       "\n",
       "  pcr_primer_name_reverse                               pcr_primer_reference  \\\n",
       "0                   EukBr  Amaral-Zettler LA, McCliment EA, Ducklow HW, H...   \n",
       "1                   EukBr  Amaral-Zettler LA, McCliment EA, Ducklow HW, H...   \n",
       "2                   EukBr  Amaral-Zettler LA, McCliment EA, Ducklow HW, H...   \n",
       "3                   EukBr  Amaral-Zettler LA, McCliment EA, Ducklow HW, H...   \n",
       "4                   EukBr  Amaral-Zettler LA, McCliment EA, Ducklow HW, H...   \n",
       "\n",
       "                                                 sop  \n",
       "0  dx.doi.org/10.17504/protocols.io.xjufknw | dx....  \n",
       "1  dx.doi.org/10.17504/protocols.io.xjufknw | dx....  \n",
       "2  dx.doi.org/10.17504/protocols.io.xjufknw | dx....  \n",
       "3  dx.doi.org/10.17504/protocols.io.xjufknw | dx....  \n",
       "4  dx.doi.org/10.17504/protocols.io.xjufknw | dx....  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Change column names as needed\n",
    "\n",
    "occ = occ.rename(columns = {'primer_sequence_forward':'pcr_primer_forward',\n",
    "                           'primer_sequence_reverse':'pcr_primer_reverse'})\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>primer_sequence_forward</th>\n",
       "      <th>primer_sequence_reverse</th>\n",
       "      <th>pcr_primer_name_forward</th>\n",
       "      <th>pcr_primer_name_reverse</th>\n",
       "      <th>pcr_primer_reference</th>\n",
       "      <th>sop</th>\n",
       "      <th>DNA_sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>Amaral-Zettler LA, McCliment EA, Ducklow HW, H...</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw | dx....</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>Amaral-Zettler LA, McCliment EA, Ducklow HW, H...</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw | dx....</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>Amaral-Zettler LA, McCliment EA, Ducklow HW, H...</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw | dx....</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>Amaral-Zettler LA, McCliment EA, Ducklow HW, H...</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw | dx....</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>Amaral-Zettler LA, McCliment EA, Ducklow HW, H...</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw | dx....</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude                env_broad_scale  \\\n",
       "0          -121.848  marine biome  (ENVO:00000447)   \n",
       "1          -121.848  marine biome  (ENVO:00000447)   \n",
       "2          -121.848  marine biome  (ENVO:00000447)   \n",
       "3          -121.846  marine biome  (ENVO:00000447)   \n",
       "4          -121.846  marine biome  (ENVO:00000447)   \n",
       "\n",
       "                 env_local_scale  \\\n",
       "0  coastal water (ENVO:00001250)   \n",
       "1  coastal water (ENVO:00001250)   \n",
       "2  coastal water (ENVO:00001250)   \n",
       "3  coastal water (ENVO:00001250)   \n",
       "4  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  primer_sequence_forward   primer_sequence_reverse pcr_primer_name_forward  \\\n",
       "0        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC                   1391f   \n",
       "1        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC                   1391f   \n",
       "2        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC                   1391f   \n",
       "3        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC                   1391f   \n",
       "4        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC                   1391f   \n",
       "\n",
       "  pcr_primer_name_reverse                               pcr_primer_reference  \\\n",
       "0                   EukBr  Amaral-Zettler LA, McCliment EA, Ducklow HW, H...   \n",
       "1                   EukBr  Amaral-Zettler LA, McCliment EA, Ducklow HW, H...   \n",
       "2                   EukBr  Amaral-Zettler LA, McCliment EA, Ducklow HW, H...   \n",
       "3                   EukBr  Amaral-Zettler LA, McCliment EA, Ducklow HW, H...   \n",
       "4                   EukBr  Amaral-Zettler LA, McCliment EA, Ducklow HW, H...   \n",
       "\n",
       "                                                 sop  \\\n",
       "0  dx.doi.org/10.17504/protocols.io.xjufknw | dx....   \n",
       "1  dx.doi.org/10.17504/protocols.io.xjufknw | dx....   \n",
       "2  dx.doi.org/10.17504/protocols.io.xjufknw | dx....   \n",
       "3  dx.doi.org/10.17504/protocols.io.xjufknw | dx....   \n",
       "4  dx.doi.org/10.17504/protocols.io.xjufknw | dx....   \n",
       "\n",
       "                                        DNA_sequence  \n",
       "0  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  \n",
       "1  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  \n",
       "2  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  \n",
       "3  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  \n",
       "4  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Add DNA_sequence\n",
    "\n",
    "occ['DNA_sequence'] = plate['ASV']\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Replace non-Latin species names with NaN\n",
    "\n",
    "# Get non-Latin names\n",
    "non_latin_names = []\n",
    "for name in names:\n",
    "    words_in_name = name.split(' ')\n",
    "    if len(words_in_name) > 2:\n",
    "        non_latin_names.append(name)\n",
    "non_latin_names_dict = {i:np.nan for i in non_latin_names}\n",
    "\n",
    "# Add any names that didn't get caught in the simple filter\n",
    "non_latin_names_dict['phototrophic eukaryote'] = np.nan\n",
    "non_latin_names_dict['Candida <clade Candida/Lodderomyces clade>'] = np.nan\n",
    "\n",
    "# Replace\n",
    "occ['scientificName'].replace(non_latin_names_dict, inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In addition, 52380 records **only give \"Eukaryota\" as the scientific name** (i.e. Eukaryota is in the kingdom field, and there is no more taxonomic information). These should be replaced with Biota (http://marinespecies.org/aphia.php?p=taxdetails&id=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Replace entries where kingdom = 'Eukaryota' with the WoRMS-approved 'Biota'\n",
    "\n",
    "occ.loc[occ['kingdom'] == 'Eukaryota', 'kingdom'] = 'Biota'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally **GenBank and WoRMS seem to disagree on their taxonomy**. Two examples:\n",
    "\n",
    "1. 120 records have \"Pelargonium\" as the genus. [On GenBank](https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?mode=Info&id=4030&lvl=3&lin=f&keep=1&srchmode=1&unlock) the full taxonomy is given as:\n",
    "    - superkingdom Eukaryota\n",
    "    - kingdom Viridiplantae (subkingdom on WoRMS)\n",
    "    - phylum Streptophyta (infrakingdom on WoRMS, phylum listed as Tracheophyta)\n",
    "    - subphylum Streptophytina (Spermatophytina on WoRMS)\n",
    "    - class Magnoliopsida (matches on WoRMS)\n",
    "        - this was indicated as 'unknown' in original data, i.e. GenBank couldn't give a scientifically-agreed-upon name for a given taxonomic level. **This doesn't make sense to me.**\n",
    "    - order Geraniales (matches on WoRMS)\n",
    "    - family Geraniaceae ([matches on WoRMS](http://marinespecies.org/aphia.php?p=taxdetails&id=382461))\n",
    "    - genus Pelargonium (does NOT match on WoRMS)\n",
    "\n",
    "\n",
    "2. 600 records have family Hemistasiidae (and no further taxonomic information). [On GenBank](https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?mode=Info&id=2015512&lvl=3&lin=f&keep=1&srchmode=1&unlock), the full taxonomy is given as:\n",
    "    - superkingdom Eukaryota\n",
    "    - phylum Euglenozoa (matches on WoRMS)\n",
    "    - order Diplonemea ([matches on WoRMS](http://marinespecies.org/aphia.php?p=taxdetails&id=582176))\n",
    "    - family Hemistasiidae (does NOT match on WoRMS)\n",
    "\n",
    "Interestingly, there are 60 records where species is Hemistasia phaeocysticola. These records also have family Hemistasiidae. The genus Hemistasia [matches on WoRMS](http://marinespecies.org/aphia.php?p=taxdetails&id=146165), but with a totally different taxonomy than [on GenBank](https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?mode=Undef&id=1503927&lvl=3&srchmode=1&keep=1&unlock):\n",
    "- kingdom Protozoa\n",
    "- phylum Euglenozoa\n",
    "- class Kinetoplastea\n",
    "- order not listed\n",
    "- family not listed\n",
    "\n",
    "**There are ongoing discussions about this issue. IOOS is working on a crosswalk between WoRMS IDs and NCBI taxonomy IDs, but it will only work for named species. See [this](https://github.com/iobis/Project-team-Genetic-Data/issues/5) GitHub issue. Either way, this is a problem that is bigger than I can solve without going case-by-case.**\n",
    "\n",
    "**In both examples above**, the lowest taxonomic rank name given by the dataset does not match on WoRMS. I therefore need a system for searching through the higher taxonomic ranks given, finding the lowest one that will match on WoRMS, and putting that name in the scientificName column. The following few code blocks do this."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Functions for finding the lowest available taxonomic rank that will match on WoRMS\n",
    "\n",
    "def fill_lowest_taxon(df, cols):\n",
    "    \"\"\" Takes the occurrence pandas data frame and fills missing values in scientificName with values from the first non-missing taxonomic rank column, listed in cols. \"\"\"\n",
    "    \n",
    "    cols.reverse()\n",
    "    \n",
    "    for col in cols[:-1]:\n",
    "        df['scientificName'] = df['scientificName'].combine_first(df[col])\n",
    "    \n",
    "    cols.reverse()\n",
    "    \n",
    "    return(df)\n",
    "\n",
    "def find_not_matched(df, name_dict):\n",
    "    \"\"\" Takes the occurrence pandas data frame and name_dict matching scientificName values with names on WoRMS and returns a list of names that did not match on WoRMS. \"\"\"\n",
    "    \n",
    "    not_matched = []\n",
    "    \n",
    "    for name in df['scientificName'].unique():\n",
    "        if name not in name_dict.keys():\n",
    "            not_matched.append(name)\n",
    "    \n",
    "    try:\n",
    "        not_matched.remove(np.nan)\n",
    "    except ValueError:\n",
    "        pass\n",
    "            \n",
    "    return(not_matched)\n",
    "\n",
    "def replace_not_matched(df, not_matched, cols):\n",
    "    \"\"\" Takes the occurrence pandas data frame and a list of scientificName values that did not match on WoRMS and replaces those values with NaN in the columns specified by cols. \"\"\"\n",
    "    \n",
    "    df[cols] = df[cols].replace(not_matched, np.nan)\n",
    "    \n",
    "    return(df)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Matching 756 names on WoRMS.\n",
      "Length of name_name_dict: 695\n",
      "Number of names not matched: 61\n",
      "Matching 35 names on WoRMS.\n",
      "Length of name_name_dict: 715\n",
      "Number of names not matched: 15\n",
      "Matching 11 names on WoRMS.\n",
      "Length of name_name_dict: 720\n",
      "Number of names not matched: 6\n",
      "Matching 3 names on WoRMS.\n",
      "Length of name_name_dict: 721\n",
      "Number of names not matched: 2\n",
      "Matching 1 names on WoRMS.\n",
      "Length of name_name_dict: 721\n",
      "Number of names not matched: 1\n",
      "Matching 0 names on WoRMS.\n",
      "Length of name_name_dict: 721\n",
      "Number of names not matched: 0\n"
     ]
    }
   ],
   "source": [
    "## Iterate to match lowest possible taxonomic rank on WoRMS (takes ~8 minutes when starting with ~750 names)\n",
    "\n",
    "# Note that cols (list of taxonomic column names) was defined when cleaning them above\n",
    "\n",
    "# Initialize dictionaries\n",
    "name_name_dict = {}\n",
    "name_id_dict = {}\n",
    "name_taxid_dict = {}\n",
    "name_class_dict = {}\n",
    "\n",
    "# Initialize not_matched\n",
    "not_matched = [1]\n",
    "\n",
    "# Iterate\n",
    "while len(not_matched) > 0:\n",
    "    \n",
    "    # Step 1 - fill\n",
    "    occ = fill_lowest_taxon(occ, cols)\n",
    "\n",
    "    # Step 2 - get names to match\n",
    "    to_match = find_not_matched(occ, name_name_dict)\n",
    "\n",
    "    # Step 3 - match\n",
    "    print('Matching {num} names on WoRMS.'.format(num = len(to_match)))\n",
    "    name_id, name_name, name_taxid, name_class = WoRMS.run_get_worms_from_scientific_name(to_match, verbose_flag=False)\n",
    "    name_id_dict = {**name_id_dict, **name_id}\n",
    "    name_name_dict = {**name_name_dict, **name_name}\n",
    "    name_taxid_dict = {**name_taxid_dict, **name_taxid}\n",
    "    name_class_dict = {**name_class_dict, **name_class}\n",
    "    print('Length of name_name_dict: {length}'.format(length = len(name_name_dict)))\n",
    "    \n",
    "    ## ---- Add in a progress bar? Also, a better/additional stopping criterion? What if not all names can be matched?\n",
    "    ## ---- Also, update this code to use pyworms instead of my functions? (Could consider cloning pyworms and adding functionality)\n",
    "\n",
    "    # Step 4 - get names that didn't match\n",
    "    not_matched = find_not_matched(occ, name_name_dict)\n",
    "    print('Number of names not matched: {num}'.format(num = len(not_matched)))\n",
    "\n",
    "    # Step 5 - replace these values with NaN\n",
    "    occ = replace_not_matched(occ, not_matched, cols)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Note** that there are 33360 records (~12% of the data) where no taxonomic information was obtained at all (i.e. scientificName is still NaN). I've used 'Biota' for these records.\n",
    "\n",
    "```python\n",
    "occ[occ['scientificName'].isna() == True].shape\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(33360, 22)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(0, 22)"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Change scientificName to Biota in cases where all taxonomic information is missing\n",
    "\n",
    "print(occ[occ['scientificName'].isna() == True].shape)\n",
    "occ.loc[occ['scientificName'].isna() == True, 'scientificName'] = 'Biota'\n",
    "occ[occ['scientificName'].isna() == True].shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally, during the above process, **I totally messed up the taxonomy columns in order to obtain the best possible scientificName column**. I'll fix that below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Fix taxonomy columns\n",
    "\n",
    "# Replace with original data\n",
    "occ[cols[1:]] = plate[['Kingdom', 'Phylum', 'Class', 'Order', 'Family', 'Genus']].copy()\n",
    "\n",
    "# Replace missing data indicators in original data with empty strings ('')\n",
    "occ[cols[1:]] = occ[cols[1:]].replace({\n",
    "    'unassigned':'',\n",
    "    's_':'',\n",
    "    'g_':'',\n",
    "    'unknown':'',\n",
    "    'no_hit':''})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check using:\n",
    "\n",
    "```python\n",
    "occ[occ['scientificName'] == 'Diplonemea'].head()\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>primer_sequence_forward</th>\n",
       "      <th>primer_sequence_reverse</th>\n",
       "      <th>...</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>taxonID</th>\n",
       "      <th>nameAccordingTo</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude                env_broad_scale  \\\n",
       "0          -121.848  marine biome  (ENVO:00000447)   \n",
       "1          -121.848  marine biome  (ENVO:00000447)   \n",
       "2          -121.848  marine biome  (ENVO:00000447)   \n",
       "3          -121.846  marine biome  (ENVO:00000447)   \n",
       "4          -121.846  marine biome  (ENVO:00000447)   \n",
       "\n",
       "                 env_local_scale  \\\n",
       "0  coastal water (ENVO:00001250)   \n",
       "1  coastal water (ENVO:00001250)   \n",
       "2  coastal water (ENVO:00001250)   \n",
       "3  coastal water (ENVO:00001250)   \n",
       "4  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  primer_sequence_forward   primer_sequence_reverse  ... scientificName  \\\n",
       "0        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...    Paracalanus   \n",
       "1        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...    Paracalanus   \n",
       "2        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...    Paracalanus   \n",
       "3        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...    Paracalanus   \n",
       "4        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...    Paracalanus   \n",
       "\n",
       "     kingdom      phylum        class      order         family        genus  \\\n",
       "0  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   \n",
       "1  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   \n",
       "2  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   \n",
       "3  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   \n",
       "4  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   \n",
       "\n",
       "                            scientificNameID taxonID nameAccordingTo  \n",
       "0  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS  \n",
       "1  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS  \n",
       "2  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS  \n",
       "3  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS  \n",
       "4  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS  \n",
       "\n",
       "[5 rows x 25 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Add scientific name-related columns\n",
    "\n",
    "occ['scientificNameID'] = occ['scientificName']\n",
    "occ['scientificNameID'].replace(name_id_dict, inplace=True)\n",
    "\n",
    "occ['taxonID'] = occ['scientificName']\n",
    "occ['taxonID'].replace(name_taxid_dict, inplace=True)\n",
    "\n",
    "occ['scientificName'].replace(name_name_dict, inplace=True)\n",
    "\n",
    "occ['nameAccordingTo'] = 'WoRMS'\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['pubmed', 'protein', 'nuccore', 'ipg', 'nucleotide', 'structure', 'genome', 'annotinfo', 'assembly', 'bioproject', 'biosample', 'blastdbinfo', 'books', 'cdd', 'clinvar', 'gap', 'gapplus', 'grasp', 'dbvar', 'gene', 'gds', 'geoprofiles', 'homologene', 'medgen', 'mesh', 'ncbisearch', 'nlmcatalog', 'omim', 'orgtrack', 'pmc', 'popset', 'proteinclusters', 'pcassay', 'protfam', 'biosystems', 'pccompound', 'pcsubstance', 'seqannot', 'snp', 'sra', 'taxonomy', 'biocollections', 'gtr']"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Get set up to query NCBI taxonomy \n",
    "\n",
    "from Bio import Entrez\n",
    "Entrez.email = 'dianalg@mbari.org'\n",
    "\n",
    "# Get list of all databases available through this tool\n",
    "record = Entrez.read(Entrez.einfo())\n",
    "all_dbs = record['DbList']\n",
    "all_dbs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Get NCBI taxIDs for each name in dataset ---- TAKES ~ 2 MINUTES FOR 300 RECORDS\n",
    "\n",
    "name_ncbiid_dict = {}\n",
    "\n",
    "for name in names:\n",
    "    handle = Entrez.esearch(db='taxonomy', retmax=10, term=name)\n",
    "    record = Entrez.read(handle)\n",
    "    name_ncbiid_dict[name] = record['IdList'][0]\n",
    "    handle.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Note** that this code will throw an IndexError (IndexError: list index out of range) if a term is not found."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>primer_sequence_forward</th>\n",
       "      <th>primer_sequence_reverse</th>\n",
       "      <th>...</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>taxonID</th>\n",
       "      <th>nameAccordingTo</th>\n",
       "      <th>taxonConceptID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude                env_broad_scale  \\\n",
       "0          -121.848  marine biome  (ENVO:00000447)   \n",
       "1          -121.848  marine biome  (ENVO:00000447)   \n",
       "2          -121.848  marine biome  (ENVO:00000447)   \n",
       "3          -121.846  marine biome  (ENVO:00000447)   \n",
       "4          -121.846  marine biome  (ENVO:00000447)   \n",
       "\n",
       "                 env_local_scale  \\\n",
       "0  coastal water (ENVO:00001250)   \n",
       "1  coastal water (ENVO:00001250)   \n",
       "2  coastal water (ENVO:00001250)   \n",
       "3  coastal water (ENVO:00001250)   \n",
       "4  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  primer_sequence_forward   primer_sequence_reverse  ...    kingdom  \\\n",
       "0        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Eukaryota   \n",
       "1        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Eukaryota   \n",
       "2        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Eukaryota   \n",
       "3        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Eukaryota   \n",
       "4        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Eukaryota   \n",
       "\n",
       "       phylum        class      order         family        genus  \\\n",
       "0  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   \n",
       "1  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   \n",
       "2  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   \n",
       "3  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   \n",
       "4  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   \n",
       "\n",
       "                            scientificNameID taxonID nameAccordingTo  \\\n",
       "0  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   \n",
       "1  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   \n",
       "2  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   \n",
       "3  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   \n",
       "4  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   \n",
       "\n",
       "  taxonConceptID  \n",
       "0                 \n",
       "1                 \n",
       "2                 \n",
       "3                 \n",
       "4                 \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Add NCBI taxonomy IDs under taxonConceptID\n",
    "\n",
    "# Map indicators that say no taxonomy was assigned to empty strings\n",
    "name_ncbiid_dict['unassigned'], name_ncbiid_dict['s_'], name_ncbiid_dict['no_hit'], name_ncbiid_dict['unknown'], name_ncbiid_dict['g_'] = '', '', '', '', ''\n",
    "\n",
    "# Create column\n",
    "occ['taxonConceptID']  = plate['Species'].copy()\n",
    "occ['taxonConceptID'].replace(name_ncbiid_dict, inplace=True)\n",
    "\n",
    "# Add remainder of text and clean\n",
    "occ['taxonConceptID'] = 'NCBI:txid' + occ['taxonConceptID']\n",
    "occ['taxonConceptID'].replace('NCBI:txid', '', inplace=True)\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>primer_sequence_forward</th>\n",
       "      <th>primer_sequence_reverse</th>\n",
       "      <th>...</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>taxonID</th>\n",
       "      <th>nameAccordingTo</th>\n",
       "      <th>taxonConceptID</th>\n",
       "      <th>identificationRemarks</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude                env_broad_scale  \\\n",
       "0          -121.848  marine biome  (ENVO:00000447)   \n",
       "1          -121.848  marine biome  (ENVO:00000447)   \n",
       "2          -121.848  marine biome  (ENVO:00000447)   \n",
       "3          -121.846  marine biome  (ENVO:00000447)   \n",
       "4          -121.846  marine biome  (ENVO:00000447)   \n",
       "\n",
       "                 env_local_scale  \\\n",
       "0  coastal water (ENVO:00001250)   \n",
       "1  coastal water (ENVO:00001250)   \n",
       "2  coastal water (ENVO:00001250)   \n",
       "3  coastal water (ENVO:00001250)   \n",
       "4  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  primer_sequence_forward   primer_sequence_reverse  ...      phylum  \\\n",
       "0        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Arthropoda   \n",
       "1        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Arthropoda   \n",
       "2        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Arthropoda   \n",
       "3        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Arthropoda   \n",
       "4        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Arthropoda   \n",
       "\n",
       "         class      order         family        genus  \\\n",
       "0  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   \n",
       "1  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   \n",
       "2  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   \n",
       "3  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   \n",
       "4  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   \n",
       "\n",
       "                            scientificNameID taxonID nameAccordingTo  \\\n",
       "0  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   \n",
       "1  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   \n",
       "2  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   \n",
       "3  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   \n",
       "4  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   \n",
       "\n",
       "  taxonConceptID                              identificationRemarks  \n",
       "0                 unassigned, Genbank nr Release 221 September 2...  \n",
       "1                 unassigned, Genbank nr Release 221 September 2...  \n",
       "2                 unassigned, Genbank nr Release 221 September 2...  \n",
       "3                 unassigned, Genbank nr Release 221 September 2...  \n",
       "4                 unassigned, Genbank nr Release 221 September 2...  \n",
       "\n",
       "[5 rows x 27 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## identificationRemarks\n",
    "\n",
    "# Get identificationRemarks\n",
    "occ = occ.merge(plate_meta[['seqID', 'identificationRemarks']], how='left', left_on='eventID', right_on='seqID')\n",
    "occ.drop(columns='seqID', inplace=True)\n",
    "\n",
    "# Add name that matched in GenBank\n",
    "occ['identificationRemarks'] = plate['Species'].copy() + ', ' + occ['identificationRemarks']\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>primer_sequence_forward</th>\n",
       "      <th>primer_sequence_reverse</th>\n",
       "      <th>...</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>taxonID</th>\n",
       "      <th>nameAccordingTo</th>\n",
       "      <th>taxonConceptID</th>\n",
       "      <th>identificationRemarks</th>\n",
       "      <th>basisOfRecord</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "      <td>MaterialSample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "      <td>MaterialSample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "      <td>MaterialSample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "      <td>MaterialSample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "      <td>MaterialSample</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude                env_broad_scale  \\\n",
       "0          -121.848  marine biome  (ENVO:00000447)   \n",
       "1          -121.848  marine biome  (ENVO:00000447)   \n",
       "2          -121.848  marine biome  (ENVO:00000447)   \n",
       "3          -121.846  marine biome  (ENVO:00000447)   \n",
       "4          -121.846  marine biome  (ENVO:00000447)   \n",
       "\n",
       "                 env_local_scale  \\\n",
       "0  coastal water (ENVO:00001250)   \n",
       "1  coastal water (ENVO:00001250)   \n",
       "2  coastal water (ENVO:00001250)   \n",
       "3  coastal water (ENVO:00001250)   \n",
       "4  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  primer_sequence_forward   primer_sequence_reverse  ...        class  \\\n",
       "0        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Hexanauplia   \n",
       "1        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Hexanauplia   \n",
       "2        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Hexanauplia   \n",
       "3        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Hexanauplia   \n",
       "4        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Hexanauplia   \n",
       "\n",
       "       order         family        genus  \\\n",
       "0  Calanoida  Paracalanidae  Paracalanus   \n",
       "1  Calanoida  Paracalanidae  Paracalanus   \n",
       "2  Calanoida  Paracalanidae  Paracalanus   \n",
       "3  Calanoida  Paracalanidae  Paracalanus   \n",
       "4  Calanoida  Paracalanidae  Paracalanus   \n",
       "\n",
       "                            scientificNameID taxonID nameAccordingTo  \\\n",
       "0  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   \n",
       "1  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   \n",
       "2  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   \n",
       "3  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   \n",
       "4  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   \n",
       "\n",
       "  taxonConceptID                              identificationRemarks  \\\n",
       "0                 unassigned, Genbank nr Release 221 September 2...   \n",
       "1                 unassigned, Genbank nr Release 221 September 2...   \n",
       "2                 unassigned, Genbank nr Release 221 September 2...   \n",
       "3                 unassigned, Genbank nr Release 221 September 2...   \n",
       "4                 unassigned, Genbank nr Release 221 September 2...   \n",
       "\n",
       "    basisOfRecord  \n",
       "0  MaterialSample  \n",
       "1  MaterialSample  \n",
       "2  MaterialSample  \n",
       "3  MaterialSample  \n",
       "4  MaterialSample  \n",
       "\n",
       "[5 rows x 28 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## basisOfRecord\n",
    "\n",
    "occ['basisOfRecord'] = 'MaterialSample'\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>primer_sequence_forward</th>\n",
       "      <th>primer_sequence_reverse</th>\n",
       "      <th>...</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>taxonID</th>\n",
       "      <th>nameAccordingTo</th>\n",
       "      <th>taxonConceptID</th>\n",
       "      <th>identificationRemarks</th>\n",
       "      <th>basisOfRecord</th>\n",
       "      <th>identificationReferences</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude                env_broad_scale  \\\n",
       "0          -121.848  marine biome  (ENVO:00000447)   \n",
       "1          -121.848  marine biome  (ENVO:00000447)   \n",
       "2          -121.848  marine biome  (ENVO:00000447)   \n",
       "3          -121.846  marine biome  (ENVO:00000447)   \n",
       "4          -121.846  marine biome  (ENVO:00000447)   \n",
       "\n",
       "                 env_local_scale  \\\n",
       "0  coastal water (ENVO:00001250)   \n",
       "1  coastal water (ENVO:00001250)   \n",
       "2  coastal water (ENVO:00001250)   \n",
       "3  coastal water (ENVO:00001250)   \n",
       "4  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  primer_sequence_forward   primer_sequence_reverse  ...      order  \\\n",
       "0        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Calanoida   \n",
       "1        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Calanoida   \n",
       "2        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Calanoida   \n",
       "3        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Calanoida   \n",
       "4        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Calanoida   \n",
       "\n",
       "          family        genus                           scientificNameID  \\\n",
       "0  Paracalanidae  Paracalanus  urn:lsid:marinespecies.org:taxname:104196   \n",
       "1  Paracalanidae  Paracalanus  urn:lsid:marinespecies.org:taxname:104196   \n",
       "2  Paracalanidae  Paracalanus  urn:lsid:marinespecies.org:taxname:104196   \n",
       "3  Paracalanidae  Paracalanus  urn:lsid:marinespecies.org:taxname:104196   \n",
       "4  Paracalanidae  Paracalanus  urn:lsid:marinespecies.org:taxname:104196   \n",
       "\n",
       "  taxonID nameAccordingTo taxonConceptID  \\\n",
       "0  104196           WoRMS                  \n",
       "1  104196           WoRMS                  \n",
       "2  104196           WoRMS                  \n",
       "3  104196           WoRMS                  \n",
       "4  104196           WoRMS                  \n",
       "\n",
       "                               identificationRemarks   basisOfRecord  \\\n",
       "0  unassigned, Genbank nr Release 221 September 2...  MaterialSample   \n",
       "1  unassigned, Genbank nr Release 221 September 2...  MaterialSample   \n",
       "2  unassigned, Genbank nr Release 221 September 2...  MaterialSample   \n",
       "3  unassigned, Genbank nr Release 221 September 2...  MaterialSample   \n",
       "4  unassigned, Genbank nr Release 221 September 2...  MaterialSample   \n",
       "\n",
       "                            identificationReferences  \n",
       "0  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...  \n",
       "1  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...  \n",
       "2  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...  \n",
       "3  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...  \n",
       "4  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...  \n",
       "\n",
       "[5 rows x 29 columns]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Add identificationReferences \n",
    "\n",
    "occ = occ.merge(plate_meta[['seqID', 'identificationReferences']], how='left', left_on='eventID', right_on='seqID')\n",
    "occ.drop(columns='seqID', inplace=True)\n",
    "occ['identificationReferences'] = occ['identificationReferences'].str.replace('| ', ' | ', regex=False)\n",
    "\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>primer_sequence_forward</th>\n",
       "      <th>primer_sequence_reverse</th>\n",
       "      <th>...</th>\n",
       "      <th>genus</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>taxonID</th>\n",
       "      <th>nameAccordingTo</th>\n",
       "      <th>taxonConceptID</th>\n",
       "      <th>identificationRemarks</th>\n",
       "      <th>basisOfRecord</th>\n",
       "      <th>identificationReferences</th>\n",
       "      <th>organismQuantity</th>\n",
       "      <th>organismQuantityType</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>14825</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>16094</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>22459</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>19312</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>16491</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude                env_broad_scale  \\\n",
       "0          -121.848  marine biome  (ENVO:00000447)   \n",
       "1          -121.848  marine biome  (ENVO:00000447)   \n",
       "2          -121.848  marine biome  (ENVO:00000447)   \n",
       "3          -121.846  marine biome  (ENVO:00000447)   \n",
       "4          -121.846  marine biome  (ENVO:00000447)   \n",
       "\n",
       "                 env_local_scale  \\\n",
       "0  coastal water (ENVO:00001250)   \n",
       "1  coastal water (ENVO:00001250)   \n",
       "2  coastal water (ENVO:00001250)   \n",
       "3  coastal water (ENVO:00001250)   \n",
       "4  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  primer_sequence_forward   primer_sequence_reverse  ...        genus  \\\n",
       "0        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Paracalanus   \n",
       "1        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Paracalanus   \n",
       "2        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Paracalanus   \n",
       "3        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Paracalanus   \n",
       "4        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Paracalanus   \n",
       "\n",
       "                            scientificNameID taxonID nameAccordingTo  \\\n",
       "0  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   \n",
       "1  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   \n",
       "2  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   \n",
       "3  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   \n",
       "4  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   \n",
       "\n",
       "  taxonConceptID                              identificationRemarks  \\\n",
       "0                 unassigned, Genbank nr Release 221 September 2...   \n",
       "1                 unassigned, Genbank nr Release 221 September 2...   \n",
       "2                 unassigned, Genbank nr Release 221 September 2...   \n",
       "3                 unassigned, Genbank nr Release 221 September 2...   \n",
       "4                 unassigned, Genbank nr Release 221 September 2...   \n",
       "\n",
       "    basisOfRecord                           identificationReferences  \\\n",
       "0  MaterialSample  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...   \n",
       "1  MaterialSample  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...   \n",
       "2  MaterialSample  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...   \n",
       "3  MaterialSample  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...   \n",
       "4  MaterialSample  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...   \n",
       "\n",
       "  organismQuantity organismQuantityType  \n",
       "0            14825   DNA sequence reads  \n",
       "1            16094   DNA sequence reads  \n",
       "2            22459   DNA sequence reads  \n",
       "3            19312   DNA sequence reads  \n",
       "4            16491   DNA sequence reads  \n",
       "\n",
       "[5 rows x 31 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## organismQuantity (number of reads)\n",
    "\n",
    "occ['organismQuantity'] = plate['Reads']\n",
    "occ['organismQuantityType'] = 'DNA sequence reads'\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Note** that there are 215537 rows where the number of reads is 0. These will need to be dropped at the end of the script."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(280440, 32)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>primer_sequence_forward</th>\n",
       "      <th>primer_sequence_reverse</th>\n",
       "      <th>...</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>taxonID</th>\n",
       "      <th>nameAccordingTo</th>\n",
       "      <th>taxonConceptID</th>\n",
       "      <th>identificationRemarks</th>\n",
       "      <th>basisOfRecord</th>\n",
       "      <th>identificationReferences</th>\n",
       "      <th>organismQuantity</th>\n",
       "      <th>organismQuantityType</th>\n",
       "      <th>sampleSizeValue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>14825</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>85600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>16094</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>90702</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>22459</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>130275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>19312</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>147220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>16491</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>121419</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude                env_broad_scale  \\\n",
       "0          -121.848  marine biome  (ENVO:00000447)   \n",
       "1          -121.848  marine biome  (ENVO:00000447)   \n",
       "2          -121.848  marine biome  (ENVO:00000447)   \n",
       "3          -121.846  marine biome  (ENVO:00000447)   \n",
       "4          -121.846  marine biome  (ENVO:00000447)   \n",
       "\n",
       "                 env_local_scale  \\\n",
       "0  coastal water (ENVO:00001250)   \n",
       "1  coastal water (ENVO:00001250)   \n",
       "2  coastal water (ENVO:00001250)   \n",
       "3  coastal water (ENVO:00001250)   \n",
       "4  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  primer_sequence_forward   primer_sequence_reverse  ...  \\\n",
       "0        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   \n",
       "1        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   \n",
       "2        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   \n",
       "3        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   \n",
       "4        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   \n",
       "\n",
       "                            scientificNameID taxonID nameAccordingTo  \\\n",
       "0  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   \n",
       "1  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   \n",
       "2  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   \n",
       "3  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   \n",
       "4  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   \n",
       "\n",
       "  taxonConceptID                              identificationRemarks  \\\n",
       "0                 unassigned, Genbank nr Release 221 September 2...   \n",
       "1                 unassigned, Genbank nr Release 221 September 2...   \n",
       "2                 unassigned, Genbank nr Release 221 September 2...   \n",
       "3                 unassigned, Genbank nr Release 221 September 2...   \n",
       "4                 unassigned, Genbank nr Release 221 September 2...   \n",
       "\n",
       "    basisOfRecord                           identificationReferences  \\\n",
       "0  MaterialSample  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...   \n",
       "1  MaterialSample  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...   \n",
       "2  MaterialSample  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...   \n",
       "3  MaterialSample  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...   \n",
       "4  MaterialSample  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...   \n",
       "\n",
       "  organismQuantity organismQuantityType sampleSizeValue  \n",
       "0            14825   DNA sequence reads           85600  \n",
       "1            16094   DNA sequence reads           90702  \n",
       "2            22459   DNA sequence reads          130275  \n",
       "3            19312   DNA sequence reads          147220  \n",
       "4            16491   DNA sequence reads          121419  \n",
       "\n",
       "[5 rows x 32 columns]"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## sampleSizeValue\n",
    "\n",
    "count_by_seq = plate.groupby('Sequence_ID', as_index=False)['Reads'].sum()\n",
    "occ = occ.merge(count_by_seq, how='left', left_on='eventID', right_on='Sequence_ID')\n",
    "occ.drop(columns='Sequence_ID', inplace=True)\n",
    "occ.rename(columns={'Reads':'sampleSizeValue'}, inplace=True)\n",
    "print(occ.shape)\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>primer_sequence_forward</th>\n",
       "      <th>primer_sequence_reverse</th>\n",
       "      <th>...</th>\n",
       "      <th>taxonID</th>\n",
       "      <th>nameAccordingTo</th>\n",
       "      <th>taxonConceptID</th>\n",
       "      <th>identificationRemarks</th>\n",
       "      <th>basisOfRecord</th>\n",
       "      <th>identificationReferences</th>\n",
       "      <th>organismQuantity</th>\n",
       "      <th>organismQuantityType</th>\n",
       "      <th>sampleSizeValue</th>\n",
       "      <th>sampleSizeUnit</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>14825</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>85600</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>16094</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>90702</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>22459</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>130275</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>19312</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>147220</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>16491</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>121419</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude                env_broad_scale  \\\n",
       "0          -121.848  marine biome  (ENVO:00000447)   \n",
       "1          -121.848  marine biome  (ENVO:00000447)   \n",
       "2          -121.848  marine biome  (ENVO:00000447)   \n",
       "3          -121.846  marine biome  (ENVO:00000447)   \n",
       "4          -121.846  marine biome  (ENVO:00000447)   \n",
       "\n",
       "                 env_local_scale  \\\n",
       "0  coastal water (ENVO:00001250)   \n",
       "1  coastal water (ENVO:00001250)   \n",
       "2  coastal water (ENVO:00001250)   \n",
       "3  coastal water (ENVO:00001250)   \n",
       "4  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  primer_sequence_forward   primer_sequence_reverse  ... taxonID  \\\n",
       "0        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  104196   \n",
       "1        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  104196   \n",
       "2        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  104196   \n",
       "3        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  104196   \n",
       "4        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  104196   \n",
       "\n",
       "  nameAccordingTo taxonConceptID  \\\n",
       "0           WoRMS                  \n",
       "1           WoRMS                  \n",
       "2           WoRMS                  \n",
       "3           WoRMS                  \n",
       "4           WoRMS                  \n",
       "\n",
       "                               identificationRemarks   basisOfRecord  \\\n",
       "0  unassigned, Genbank nr Release 221 September 2...  MaterialSample   \n",
       "1  unassigned, Genbank nr Release 221 September 2...  MaterialSample   \n",
       "2  unassigned, Genbank nr Release 221 September 2...  MaterialSample   \n",
       "3  unassigned, Genbank nr Release 221 September 2...  MaterialSample   \n",
       "4  unassigned, Genbank nr Release 221 September 2...  MaterialSample   \n",
       "\n",
       "                            identificationReferences organismQuantity  \\\n",
       "0  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...            14825   \n",
       "1  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...            16094   \n",
       "2  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...            22459   \n",
       "3  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...            19312   \n",
       "4  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...            16491   \n",
       "\n",
       "  organismQuantityType sampleSizeValue      sampleSizeUnit  \n",
       "0   DNA sequence reads           85600  DNA sequence reads  \n",
       "1   DNA sequence reads           90702  DNA sequence reads  \n",
       "2   DNA sequence reads          130275  DNA sequence reads  \n",
       "3   DNA sequence reads          147220  DNA sequence reads  \n",
       "4   DNA sequence reads          121419  DNA sequence reads  \n",
       "\n",
       "[5 rows x 33 columns]"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## sampleSizeUnit\n",
    "\n",
    "occ['sampleSizeUnit'] = 'DNA sequence reads'\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>primer_sequence_forward</th>\n",
       "      <th>primer_sequence_reverse</th>\n",
       "      <th>...</th>\n",
       "      <th>nameAccordingTo</th>\n",
       "      <th>taxonConceptID</th>\n",
       "      <th>identificationRemarks</th>\n",
       "      <th>basisOfRecord</th>\n",
       "      <th>identificationReferences</th>\n",
       "      <th>organismQuantity</th>\n",
       "      <th>organismQuantityType</th>\n",
       "      <th>sampleSizeValue</th>\n",
       "      <th>sampleSizeUnit</th>\n",
       "      <th>associatedSequences</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>14825</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>85600</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>NCBI BioProject accession number PRJNA433203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>16094</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>90702</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>NCBI BioProject accession number PRJNA433203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>22459</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>130275</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>NCBI BioProject accession number PRJNA433203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>19312</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>147220</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>NCBI BioProject accession number PRJNA433203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>16491</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>121419</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>NCBI BioProject accession number PRJNA433203</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude                env_broad_scale  \\\n",
       "0          -121.848  marine biome  (ENVO:00000447)   \n",
       "1          -121.848  marine biome  (ENVO:00000447)   \n",
       "2          -121.848  marine biome  (ENVO:00000447)   \n",
       "3          -121.846  marine biome  (ENVO:00000447)   \n",
       "4          -121.846  marine biome  (ENVO:00000447)   \n",
       "\n",
       "                 env_local_scale  \\\n",
       "0  coastal water (ENVO:00001250)   \n",
       "1  coastal water (ENVO:00001250)   \n",
       "2  coastal water (ENVO:00001250)   \n",
       "3  coastal water (ENVO:00001250)   \n",
       "4  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  primer_sequence_forward   primer_sequence_reverse  ... nameAccordingTo  \\\n",
       "0        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...           WoRMS   \n",
       "1        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...           WoRMS   \n",
       "2        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...           WoRMS   \n",
       "3        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...           WoRMS   \n",
       "4        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...           WoRMS   \n",
       "\n",
       "  taxonConceptID                              identificationRemarks  \\\n",
       "0                 unassigned, Genbank nr Release 221 September 2...   \n",
       "1                 unassigned, Genbank nr Release 221 September 2...   \n",
       "2                 unassigned, Genbank nr Release 221 September 2...   \n",
       "3                 unassigned, Genbank nr Release 221 September 2...   \n",
       "4                 unassigned, Genbank nr Release 221 September 2...   \n",
       "\n",
       "    basisOfRecord                           identificationReferences  \\\n",
       "0  MaterialSample  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...   \n",
       "1  MaterialSample  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...   \n",
       "2  MaterialSample  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...   \n",
       "3  MaterialSample  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...   \n",
       "4  MaterialSample  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...   \n",
       "\n",
       "  organismQuantity organismQuantityType sampleSizeValue      sampleSizeUnit  \\\n",
       "0            14825   DNA sequence reads           85600  DNA sequence reads   \n",
       "1            16094   DNA sequence reads           90702  DNA sequence reads   \n",
       "2            22459   DNA sequence reads          130275  DNA sequence reads   \n",
       "3            19312   DNA sequence reads          147220  DNA sequence reads   \n",
       "4            16491   DNA sequence reads          121419  DNA sequence reads   \n",
       "\n",
       "                            associatedSequences  \n",
       "0  NCBI BioProject accession number PRJNA433203  \n",
       "1  NCBI BioProject accession number PRJNA433203  \n",
       "2  NCBI BioProject accession number PRJNA433203  \n",
       "3  NCBI BioProject accession number PRJNA433203  \n",
       "4  NCBI BioProject accession number PRJNA433203  \n",
       "\n",
       "[5 rows x 34 columns]"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## associatedSequences\n",
    "\n",
    "occ = occ.merge(plate_meta[['seqID', 'associatedSequences']], how='left', left_on='eventID', right_on='seqID')\n",
    "occ.drop(columns='seqID', inplace=True)\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(64903, 34)\n"
     ]
    }
   ],
   "source": [
    "## Drop records where organismQuantity = 0 (absences are not meaningful for this data set)\n",
    "\n",
    "occ = occ[occ['organismQuantity'] > 0]\n",
    "print(occ.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Note** that the vast majority of this data has 0 reads (215537 records out of 280440, or about 75%). **This makes sense, and is generally fine. Although Katie said something about the 0's being potentially important if the data set were to be reanalyzed due to the compositional nature of the data. We can stay in conversation about this.**\n",
    "\n",
    "Also **note** that there is one sample (34916c01_12_edna_2_S) that has no reads at all in the ASV table."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "eventID                     0\n",
       "eventDate                   0\n",
       "decimalLatitude             0\n",
       "decimalLongitude            0\n",
       "env_broad_scale             0\n",
       "env_local_scale             0\n",
       "env_medium                  0\n",
       "target_gene                 0\n",
       "primer_sequence_forward     0\n",
       "primer_sequence_reverse     0\n",
       "pcr_primer_name_forward     0\n",
       "pcr_primer_name_reverse     0\n",
       "pcr_primer_reference        0\n",
       "sop                         0\n",
       "DNA_sequence                0\n",
       "scientificName              0\n",
       "kingdom                     0\n",
       "phylum                      0\n",
       "class                       0\n",
       "order                       0\n",
       "family                      0\n",
       "genus                       0\n",
       "scientificNameID            0\n",
       "taxonID                     0\n",
       "nameAccordingTo             0\n",
       "taxonConceptID              0\n",
       "identificationRemarks       0\n",
       "basisOfRecord               0\n",
       "identificationReferences    0\n",
       "organismQuantity            0\n",
       "organismQuantityType        0\n",
       "sampleSizeValue             0\n",
       "sampleSizeUnit              0\n",
       "associatedSequences         0\n",
       "dtype: int64"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Check for NaN values in string fields\n",
    "\n",
    "occ.isna().sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Save"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Save\n",
    "\n",
    "occ.to_csv('eDNA_practice_plate_occ_20210527.csv', index=False, na_rep='NaN')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Questions\n",
    "\n",
    "1. Katie still needs to remove the clarifying text from the sop column (e.g. 'ext: ') **Done.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "standardization_env",
   "language": "python",
   "name": "standardization_env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
